{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "923483ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import plotly.graph_objects as go\n",
    "import json\n",
    "\n",
    "from copy import deepcopy\n",
    "\n",
    "from tqdm.notebook import tqdm\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "39efc0dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_excel(\"../supplement/SupplementalTable7_top1000_pSNVs.xlsx\", skiprows=1)\n",
    "data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2b061866",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Check the number of TERT, RALY and CDC20 mutations.\n",
    "data[data[\"GENE\"] == \"TERT\"].value_counts([\"POS\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57ab1372",
   "metadata": {},
   "outputs": [],
   "source": [
    "data[data[\"GENE\"] == \"RALY\"].value_counts([\"POS\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "806a1f32",
   "metadata": {},
   "outputs": [],
   "source": [
    "data[data[\"GENE\"] == \"CDC20\"].value_counts([\"POS\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b2622533",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Original function to add REMIND-Cancer scores based on the scoring weights.\n",
    "def _add_remind_cancer_score(\n",
    "    dataframe: pd.DataFrame,\n",
    "    remind_cancer_scoring_weights: dict,\n",
    "    ge_column: str\n",
    "):\n",
    "    \"\"\"Add the REMIND-Cancer score to the DataFrame.\n",
    "\n",
    "    Args:\n",
    "        dataframe (pd.DataFrame): Pandas DataFrame containing the data to score.\n",
    "        remind_cancer_scoring_weights (dict): Dictionary containing the REMIND-Cancer scoring weights.\n",
    "            Example can be found in REMIND-Cancer/examples/results/configuration_file.json, particularly the last key `REMIND-Cancer_scoring_weights`.\n",
    "        ge_column (str): Name of the column containing gene expression data.\n",
    "\n",
    "    Returns:\n",
    "        pd.DataFrame: DataFrame with the REMIND-Cancer score added.\n",
    "    \"\"\"\n",
    "    # Iterate through each row of the DataFrame and calculate the REMIND-Cancer score.\n",
    "    for idx, row in dataframe.iterrows():\n",
    "        # (1) Genomic.\n",
    "        # Transcription Factors\n",
    "        created_tf_weight = np.min([\n",
    "            float(row[\"num_created_tfs_passing_tf_expression_threshold\"]) *\n",
    "            remind_cancer_scoring_weights[\"genomic\"][\"tfbs\"][\"creation_weight_per_tfbs\"],\n",
    "            remind_cancer_scoring_weights[\"genomic\"][\"tfbs\"][\"creation_weight_maximum\"]\n",
    "        ])\n",
    "\n",
    "        destroyed_tf_weight = np.min([\n",
    "            float(row[\"num_destroyed_tfs_passing_tf_expression_threshold\"]) *\n",
    "            remind_cancer_scoring_weights[\"genomic\"][\"tfbs\"][\"destruction_weight_per_tfbs\"],\n",
    "            remind_cancer_scoring_weights[\"genomic\"][\"tfbs\"][\"destruction_weight_maximum\"]\n",
    "        ])\n",
    "\n",
    "        # Recurrence\n",
    "        recurrence_weight = np.min([\n",
    "            float(row[\"num_recurrent_mutations\"]) *\n",
    "            remind_cancer_scoring_weights[\"genomic\"][\"recurrence\"][\"weight_per_recurrent_mutation\"],\n",
    "            remind_cancer_scoring_weights[\"genomic\"][\"recurrence\"][\"weight_maximum\"]\n",
    "        ])\n",
    "\n",
    "        # Purity\n",
    "        try:\n",
    "            purity_weight = remind_cancer_scoring_weights[\"genomic\"][\"purity\"][\"purity_weight\"] if float(\n",
    "                row[\"ICGC_Estimated_Purity\"]) >= remind_cancer_scoring_weights[\"genomic\"][\"purity\"][\"purity_threshold_for_weight\"] else 0\n",
    "        except:\n",
    "            # print(\n",
    "            #     f\"Purity of {row['ICGC_Estimated_Purity']} does not work. Setting score to 0.\")\n",
    "            purity_weight = 0\n",
    "\n",
    "        # Allele Frequency\n",
    "        try:\n",
    "            af_weight = remind_cancer_scoring_weights[\"genomic\"][\"allele_frequency\"][\"af_weight\"] if float(\n",
    "                row[\"allele_frequency\"]) >= remind_cancer_scoring_weights[\"genomic\"][\"allele_frequency\"][\"af_threshold_for_weight\"] else 0\n",
    "        except:\n",
    "            # print(\n",
    "            #     f\"Allele frequency of {row['allele_frequency']} does not work. Setting score to 0.\")\n",
    "            af_weight = 0\n",
    "\n",
    "        # (2) Transcriptomic\n",
    "        # Gene Expression\n",
    "        gene_expression_weight = np.min([\n",
    "            float(row[ge_column]) *\n",
    "            remind_cancer_scoring_weights[\"transcriptomic\"][\"gene_expression\"][\"weight_per_unit_of_expression\"],\n",
    "            remind_cancer_scoring_weights[\"transcriptomic\"][\"gene_expression\"][\"weight_maximum\"]\n",
    "        ])\n",
    "\n",
    "        # (3) Annotations\n",
    "        # CGC\n",
    "        cgc_weight = bool(row[\"within_cgc_list\"]) * \\\n",
    "            remind_cancer_scoring_weights[\"annotations\"][\"cgc\"][\"weight\"]\n",
    "\n",
    "        # Open Chromatin\n",
    "        open_chromatin_weight = bool(\n",
    "            row[\"open_chromatin\"]) * remind_cancer_scoring_weights[\"annotations\"][\"open_chromatin\"][\"weight\"]\n",
    "\n",
    "        # Create the final score by summing all the weights.\n",
    "        # Round the score to 2 decimal places.\n",
    "        # Assign the score to the \"score_v2\" column in the DataFrame.\n",
    "        dataframe.loc[idx, \"score_v2\"] = np.round(\n",
    "            np.sum([\n",
    "                created_tf_weight,\n",
    "                destroyed_tf_weight,\n",
    "                recurrence_weight,\n",
    "                purity_weight,\n",
    "                af_weight,\n",
    "                gene_expression_weight,\n",
    "                cgc_weight,\n",
    "                open_chromatin_weight\n",
    "            ]), 2)\n",
    "\n",
    "    # Sort the DataFrame by the \"score_v2\" column in descending order.\n",
    "    dataframe = dataframe.sort_values(\"score_v2\", ascending=False)\n",
    "\n",
    "    return dataframe\n",
    "\n",
    "with open(\"../../examples/results/configuration_file.json\", \"r\") as f:\n",
    "    scoring_weights = json.load(f)\n",
    "    scoring_weights = scoring_weights[\"REMIND-Cancer_scoring_weights\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9946b842",
   "metadata": {},
   "source": [
    "# Get the median, mean, min, and max ranks for each gene and position\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "02518649",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define the weight mapping for each scoring attribute.\n",
    "# The main key (e.g. Recurrence Weight) is the name of the weight\n",
    "# The `attribute` is a list of the keys/subkeys to access the weight\n",
    "#      Ex: scoring_weights[\"genomic\"][\"recurrence\"][\"weight_per_recurrent_mutation\"] = [\"genomic\", \"recurrence\", \"weight_per_recurrent_mutation\"]\n",
    "# The `range` is the range of values to iterate through for the weight\n",
    "#      Ex: np.linspace(1, 100, 100) will create a range from 1 to 100 with 100 points\n",
    "# The `original_weight` is the original weight value for the attribute\n",
    "\n",
    "weight_mapping = {\n",
    "    \"Recurrence Weight\": {\n",
    "        \"attribute\": [\"genomic\", \"recurrence\", \"weight_per_recurrent_mutation\"],\n",
    "        \"range\": np.linspace(1, 100, 100),\n",
    "        \"original_weight\": scoring_weights[\"genomic\"][\"recurrence\"][\"weight_per_recurrent_mutation\"],\n",
    "    },\n",
    "    \"Gene Expression Weight\": {\n",
    "        \"attribute\": [\"transcriptomic\", \"gene_expression\", \"weight_per_unit_of_expression\"],\n",
    "        \"range\": np.linspace(1, 100, 100),\n",
    "        \"original_weight\": scoring_weights[\"transcriptomic\"][\"gene_expression\"][\"weight_per_unit_of_expression\"],\n",
    "    },\n",
    "    \"Purity Weight\": {\n",
    "        \"attribute\": [\"genomic\", \"purity\", \"purity_weight\"],\n",
    "        \"range\": np.linspace(1, 100, 100),\n",
    "        \"original_weight\": scoring_weights[\"genomic\"][\"purity\"][\"purity_weight\"],\n",
    "    },\n",
    "    \"Transcription Factor Creation Weight\": {\n",
    "        \"attribute\": [\"genomic\", \"tfbs\", \"creation_weight_per_tfbs\"],\n",
    "        \"range\": np.linspace(1, 20, 100),\n",
    "        \"original_weight\": scoring_weights[\"genomic\"][\"tfbs\"][\"creation_weight_per_tfbs\"],\n",
    "    },\n",
    "    \"Transcription Factor Destruction Weight\": {\n",
    "        \"attribute\": [\"genomic\", \"tfbs\", \"destruction_weight_per_tfbs\"],\n",
    "        \"range\": np.linspace(1, 20, 100),\n",
    "        \"original_weight\": scoring_weights[\"genomic\"][\"tfbs\"][\"destruction_weight_per_tfbs\"],\n",
    "    },\n",
    "    \"Allele Frequency Weight\": {\n",
    "        \"attribute\": [\"genomic\", \"allele_frequency\", \"af_weight\"],\n",
    "        \"range\": np.linspace(1, 50, 100),\n",
    "        \"original_weight\": scoring_weights[\"genomic\"][\"allele_frequency\"][\"af_weight\"],\n",
    "    },\n",
    "    \"CGC Weight\": {\n",
    "        \"attribute\": [\"annotations\", \"cgc\", \"weight\"],\n",
    "        \"range\": np.linspace(1, 50, 100),\n",
    "        \"original_weight\": scoring_weights[\"annotations\"][\"cgc\"][\"weight\"],\n",
    "    },\n",
    "    \"Open Chromatin Weight\": {\n",
    "        \"attribute\": [\"annotations\", \"open_chromatin\", \"weight\"],\n",
    "        \"range\": np.linspace(1, 50, 100),\n",
    "        \"original_weight\": scoring_weights[\"annotations\"][\"open_chromatin\"][\"weight\"],\n",
    "    }\n",
    " }\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "74a769aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# For each gene and position, plot the four graphs in a plotly subplot.\n",
    "import plotly.subplots as sp\n",
    "\n",
    "def create_subplots(\n",
    "    medians: dict, \n",
    "    means: dict, \n",
    "    mins: dict, \n",
    "    maxes: dict, \n",
    "    name: str, \n",
    "    gene: str, \n",
    "    position: int, \n",
    "    output_dir=\"./plots\"\n",
    "):\n",
    "    \"\"\"Create subplots for the given means, medians, mins, and maxes.\n",
    "\n",
    "    Args:\n",
    "        medians (dict): A dictionary containing the median ranks for each gene and position.\n",
    "        means (dict): A dictionary containing the mean ranks for each gene and position.\n",
    "        mins (dict): A dictionary containing the minimum ranks for each gene and position.\n",
    "        maxes (dict): A dictionary containing the maximum ranks for each gene and position.\n",
    "        name (str): The name of the weight being varied.\n",
    "        gene (str): The gene being analyzed.\n",
    "        position (int): The position of the gene being analyzed.\n",
    "        output_dir (str, optional): The directory to save the plots. Defaults to \"./plots\".\n",
    "    \"\"\"\n",
    "    # Ensure the output directory exists.\n",
    "    if not os.path.exists(output_dir):\n",
    "        os.makedirs(output_dir, exist_ok=True)\n",
    "        print(f\"Created output directory: {os.path.abspath(output_dir)}\")\n",
    "\n",
    "    # Define the 2x2 grid for the subplots.\n",
    "    fig = sp.make_subplots(\n",
    "        rows=2, cols=2,\n",
    "        subplot_titles=[\n",
    "            \"Mean Rank\", \"Median Rank\",\n",
    "            \"Min Rank\", \"Max Rank\"\n",
    "        ],\n",
    "    )\n",
    "\n",
    "    # Plot the mean rank at position (1,1)\n",
    "    fig.add_trace(\n",
    "        go.Scatter(\n",
    "            x=weight_mapping[name][\"range\"],\n",
    "            y=means[gene][position],\n",
    "            mode=\"lines\",\n",
    "            name=f\"{gene} {position} Mean Rank\",\n",
    "            marker=dict(size=10, color=\"blue\")\n",
    "        ),\n",
    "        row=1, col=1\n",
    "    )\n",
    "    # Also, add a vertical line for the original weight.\n",
    "    fig.add_vline(\n",
    "        x=weight_mapping[name][\"original_weight\"],\n",
    "        line=dict(color=\"black\", width=2, dash=\"dash\"),\n",
    "    )\n",
    "\n",
    "    # Plot the median rank at position (1,2)\n",
    "    fig.add_trace(\n",
    "        go.Scatter(\n",
    "            x=weight_mapping[name][\"range\"],\n",
    "            y=medians[gene][position],\n",
    "            mode=\"lines\",\n",
    "            name=f\"{gene} {position} Median Rank\",\n",
    "            marker=dict(size=10, color=\"orange\")\n",
    "        ),\n",
    "        row=1, col=2\n",
    "    )\n",
    "    # Also, add a vertical line for the original weight.\n",
    "    fig.add_vline(\n",
    "        x=weight_mapping[name][\"original_weight\"],\n",
    "        line=dict(color=\"black\", width=2, dash=\"dash\"),\n",
    "    )\n",
    "\n",
    "    # Plot the min rank at position (2,1).\n",
    "    fig.add_trace(\n",
    "        go.Scatter(\n",
    "            x=weight_mapping[name][\"range\"],\n",
    "            y=mins[gene][position],\n",
    "            mode=\"lines\",\n",
    "            name=f\"{gene} {position} Min Rank\",\n",
    "            marker=dict(size=10, color=\"green\")\n",
    "        ),\n",
    "        row=2, col=1\n",
    "    )\n",
    "    # Also, add a vertical line for the original weight.\n",
    "    fig.add_vline(\n",
    "        x=weight_mapping[name][\"original_weight\"],\n",
    "        line=dict(color=\"black\", width=2, dash=\"dash\"),\n",
    "    )\n",
    "\n",
    "    # Plot the max rank at position (2,2).\n",
    "    fig.add_trace(\n",
    "        go.Scatter(\n",
    "            x=weight_mapping[name][\"range\"],\n",
    "            y=maxes[gene][position],\n",
    "            mode=\"lines\",\n",
    "            name=f\"{gene} {position} Max Rank\",\n",
    "            marker=dict(size=10, color=\"red\")\n",
    "        ),\n",
    "        row=2, col=2\n",
    "    )\n",
    "    # Also, add a vertical line for the original weight.\n",
    "    fig.add_vline(\n",
    "        x=weight_mapping[name][\"original_weight\"],\n",
    "        line=dict(color=\"black\", width=2, dash=\"dash\"),\n",
    "    )\n",
    "    fig.update_layout(\n",
    "        title=f\"Iterrating through <b>{name}</b> Scoring for {gene} at position {position} (n={data[(data['GENE'] == gene) & (data['POS'] == position)].shape[0]})\",\n",
    "        showlegend=True\n",
    "    )\n",
    "    # Add a common x-axis label.\n",
    "    fig.update_xaxes(title_text=name, row=2, col=1)\n",
    "    fig.update_xaxes(title_text=name, row=2, col=2)\n",
    "    # Add a common y-axis label.\n",
    "    fig.update_yaxes(title_text=\"Rank\", row=1, col=1)\n",
    "    fig.update_yaxes(title_text=\"Rank\", row=2, col=1)\n",
    "\n",
    "\n",
    "    fig.write_html(os.path.join(output_dir, f\"{name.replace(' ', '_')}_{gene}_{position}_score_iteration.html\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8c94002",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Iterate through the recurrence scoring.\n",
    "\n",
    "# Load in the original scoring weights. These will be changed for each iteration.\n",
    "with open(\"../../examples/results/configuration_file.json\", \"r\") as f:\n",
    "    scoring_weights = json.load(f)\n",
    "    scoring_weights = scoring_weights[\"REMIND-Cancer_scoring_weights\"]\n",
    "\n",
    "# Iterate through the weight mapping and create subplots for each weight.\n",
    "for name in weight_mapping:\n",
    "    # Initialize the figure for the subplots.\n",
    "    fig = go.Figure()\n",
    "\n",
    "    # Initialize dictionaries to store means, medians, mins, and maxes for each gene and position.\n",
    "    means = {}\n",
    "    medians = {}\n",
    "    mins = {}\n",
    "    maxes = {}\n",
    "\n",
    "    # Iterate through the genes and positions of interest.\n",
    "    for gene, position, color in [\n",
    "        (\"TERT\", 1295228, \"red\"), \n",
    "        (\"TERT\", 1295250, \"blue\"), \n",
    "        (\"RALY\", 32580927, \"green\"), \n",
    "        (\"CDC20\", 43824529, \"yellow\")\n",
    "    ]:\n",
    "        # Initialize the dictionaries for each gene and position.\n",
    "        if gene not in means:\n",
    "            means[gene] = {}\n",
    "            medians[gene] = {}\n",
    "            mins[gene] = {}\n",
    "            maxes[gene] = {}\n",
    "        \n",
    "        if position not in means[gene]:\n",
    "            means[gene][position] = []\n",
    "            medians[gene][position] = []\n",
    "            mins[gene][position] = []\n",
    "            maxes[gene][position] = []\n",
    "\n",
    "        # Initialize the x-axis range, which is just the range of the weight.\n",
    "        x_axis_scoring_range = weight_mapping[name][\"range\"]\n",
    "\n",
    "        # Initialize the y-axis score list.\n",
    "        y_axis_score = []\n",
    "\n",
    "        # Iterate through the x-axis scoring range and calculate the scores.\n",
    "        for x in tqdm(x_axis_scoring_range, desc=f\"{name}: {gene} {position}\"):\n",
    "            # Create a temporary copy of the scoring weights to modify.\n",
    "            scoring_weights_temp = deepcopy(scoring_weights)\n",
    "\n",
    "            # Update the scoring weights with the current x value.\n",
    "            # Ex: scoring_weights_temp[\"genomic\"][\"recurrence\"][\"weight_per_recurrent_mutation\"] = 1\n",
    "            scoring_weights_temp[\n",
    "                weight_mapping[name][\"attribute\"][0]\n",
    "            ][\n",
    "                weight_mapping[name][\"attribute\"][1]\n",
    "            ][\n",
    "                weight_mapping[name][\"attribute\"][2]\n",
    "            ] = x\n",
    "\n",
    "            # Calculate the scores using the modified scoring weights.\n",
    "            data_temp = _add_remind_cancer_score(\n",
    "                dataframe=data.copy(),\n",
    "                remind_cancer_scoring_weights=scoring_weights_temp,\n",
    "                ge_column=\"FPKM_Z_score\"\n",
    "            )\n",
    "\n",
    "            # Since data_temp now has the new scores as `score_v2`, sort it.\n",
    "            data_temp = data_temp.sort_values(\"score_v2\", ascending=False, inplace=False)\n",
    "\n",
    "            # Reset the index of the data_temp DataFrame to have the ranks.\n",
    "            data_temp = data_temp.reset_index(drop=True)\n",
    "\n",
    "            # Get the median rank by filtering the data_temp DataFrame for the current gene and position and \n",
    "            # then getting the index, which represents the rank.\n",
    "            median = np.median(\n",
    "                list(\n",
    "                    data_temp[(data_temp[\"GENE\"] == gene) & (data_temp[\"POS\"] == position)].index\n",
    "                )\n",
    "            )\n",
    "\n",
    "            # Get the mean, min, and max ranks for the current gene and position with the same idea.\n",
    "            mean = np.mean(list(data_temp[(data_temp[\"GENE\"] == gene) & (data_temp[\"POS\"] == position)].index))\n",
    "            min = np.min(list(data_temp[(data_temp[\"GENE\"] == gene) & (data_temp[\"POS\"] == position)].index))\n",
    "            max = np.max(list(data_temp[(data_temp[\"GENE\"] == gene) & (data_temp[\"POS\"] == position)].index))\n",
    "\n",
    "            # Append the mean, median, min, and max ranks to the respective lists.\n",
    "            means[gene][position].append(mean)\n",
    "            medians[gene][position].append(median)\n",
    "            mins[gene][position].append(min)\n",
    "            maxes[gene][position].append(max)\n",
    "\n",
    "        # Call the create_subplots function to create the subplots for the current gene and position.\n",
    "        # These should be saved within the `plots` directory.\n",
    "        create_subplots(medians, means, mins, maxes, name, gene, position, output_dir=\"./plots\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ecabbf6",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "maptropolis-api (3.12.10)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
