{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Why did the TERT C228T mutations drop out?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>score</th>\n",
       "      <th>pid</th>\n",
       "      <th>cohort</th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>GENE</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>strand</th>\n",
       "      <th>...</th>\n",
       "      <th>num_promoter_mutations</th>\n",
       "      <th>num_recurrent_mutations</th>\n",
       "      <th>expression_traces</th>\n",
       "      <th>tf_summary</th>\n",
       "      <th>ncbi_gene_summary</th>\n",
       "      <th>ncbi_url</th>\n",
       "      <th>in_old_results</th>\n",
       "      <th>cn_score</th>\n",
       "      <th>cn_file_of_score</th>\n",
       "      <th>ICGC_Estimated_Purity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>93.45</td>\n",
       "      <td>45a7949d-e63f-4956-866c-df51257032de</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>2</td>\n",
       "      <td>71204529</td>\n",
       "      <td>ANKRD53</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>284</td>\n",
       "      <td>5.0</td>\n",
       "      <td>{'ANKRD53': {'BLCA-US': {'raw': [0.13674067793...</td>\n",
       "      <td>{'REL': {'summary': \"This gene encodes a prote...</td>\n",
       "      <td>Involved in mitotic metaphase plate congressio...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/79998</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>93.41</td>\n",
       "      <td>6c884037-49f9-41c3-b0e5-9cbcd545aeb7</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>9</td>\n",
       "      <td>91933357</td>\n",
       "      <td>SECISBP2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>879</td>\n",
       "      <td>6.0</td>\n",
       "      <td>{'SECISBP2': {'SKCM-US': {'raw': [5.0687620882...</td>\n",
       "      <td>{'ELK4': {'summary': 'This gene is a member of...</td>\n",
       "      <td>The protein encoded by this gene is one of the...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/79048</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>88.10</td>\n",
       "      <td>49dec0c2-8e75-4f44-a253-82b2ea605890</td>\n",
       "      <td>LUAD-US</td>\n",
       "      <td>2</td>\n",
       "      <td>71204529</td>\n",
       "      <td>ANKRD53</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>436</td>\n",
       "      <td>5.0</td>\n",
       "      <td>{'ANKRD53': {'LUAD-US': {'raw': [0.73980555761...</td>\n",
       "      <td>{'REL': {'summary': \"This gene encodes a prote...</td>\n",
       "      <td>Involved in mitotic metaphase plate congressio...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/79998</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>86.45</td>\n",
       "      <td>d692ecd0-2433-426b-9af2-a30c44a80f7c</td>\n",
       "      <td>THCA-US</td>\n",
       "      <td>5</td>\n",
       "      <td>1295228</td>\n",
       "      <td>TERT</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>95.0</td>\n",
       "      <td>{'TERT': {'THCA-US': {'raw': [0.0, 0.006731104...</td>\n",
       "      <td>{'ELK4': {'summary': 'This gene is a member of...</td>\n",
       "      <td>Telomerase is a ribonucleoprotein polymerase t...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/7015</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>86.30</td>\n",
       "      <td>9988eb07-01f6-4f83-8699-bb63e0525f08</td>\n",
       "      <td>HNSC-US</td>\n",
       "      <td>7</td>\n",
       "      <td>145813786</td>\n",
       "      <td>CNTNAP2</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>459</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'CNTNAP2': {'HNSC-US': {'raw': [0.38939713831...</td>\n",
       "      <td>{'ZEB1': {'summary': 'This gene encodes a zinc...</td>\n",
       "      <td>This gene encodes a member of the neurexin fam...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/26047</td>\n",
       "      <td>False</td>\n",
       "      <td>9</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.489</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 165 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  score                                   pid   cohort  #CHROM  \\\n",
       "0           0  93.45  45a7949d-e63f-4956-866c-df51257032de  BLCA-US       2   \n",
       "1           1  93.41  6c884037-49f9-41c3-b0e5-9cbcd545aeb7  SKCM-US       9   \n",
       "2           2  88.10  49dec0c2-8e75-4f44-a253-82b2ea605890  LUAD-US       2   \n",
       "3           3  86.45  d692ecd0-2433-426b-9af2-a30c44a80f7c  THCA-US       5   \n",
       "4           4  86.30  9988eb07-01f6-4f83-8699-bb63e0525f08  HNSC-US       7   \n",
       "\n",
       "         POS      GENE REF ALT strand  ...  num_promoter_mutations  \\\n",
       "0   71204529   ANKRD53   G   A      +  ...                     284   \n",
       "1   91933357  SECISBP2   G   A      +  ...                     879   \n",
       "2   71204529   ANKRD53   G   A      +  ...                     436   \n",
       "3    1295228      TERT   G   A      -  ...                      20   \n",
       "4  145813786   CNTNAP2   C   A      +  ...                     459   \n",
       "\n",
       "   num_recurrent_mutations                                  expression_traces  \\\n",
       "0                      5.0  {'ANKRD53': {'BLCA-US': {'raw': [0.13674067793...   \n",
       "1                      6.0  {'SECISBP2': {'SKCM-US': {'raw': [5.0687620882...   \n",
       "2                      5.0  {'ANKRD53': {'LUAD-US': {'raw': [0.73980555761...   \n",
       "3                     95.0  {'TERT': {'THCA-US': {'raw': [0.0, 0.006731104...   \n",
       "4                      0.0  {'CNTNAP2': {'HNSC-US': {'raw': [0.38939713831...   \n",
       "\n",
       "                                          tf_summary  \\\n",
       "0  {'REL': {'summary': \"This gene encodes a prote...   \n",
       "1  {'ELK4': {'summary': 'This gene is a member of...   \n",
       "2  {'REL': {'summary': \"This gene encodes a prote...   \n",
       "3  {'ELK4': {'summary': 'This gene is a member of...   \n",
       "4  {'ZEB1': {'summary': 'This gene encodes a zinc...   \n",
       "\n",
       "                                   ncbi_gene_summary  \\\n",
       "0  Involved in mitotic metaphase plate congressio...   \n",
       "1  The protein encoded by this gene is one of the...   \n",
       "2  Involved in mitotic metaphase plate congressio...   \n",
       "3  Telomerase is a ribonucleoprotein polymerase t...   \n",
       "4  This gene encodes a member of the neurexin fam...   \n",
       "\n",
       "                                  ncbi_url in_old_results cn_score  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/gene/79998           True        3   \n",
       "1  https://www.ncbi.nlm.nih.gov/gene/79048           True        6   \n",
       "2  https://www.ncbi.nlm.nih.gov/gene/79998           True        4   \n",
       "3   https://www.ncbi.nlm.nih.gov/gene/7015           True        3   \n",
       "4  https://www.ncbi.nlm.nih.gov/gene/26047          False        9   \n",
       "\n",
       "   cn_file_of_score ICGC_Estimated_Purity  \n",
       "0     not_available                 0.938  \n",
       "1     not_available                  0.78  \n",
       "2     not_available                   0.5  \n",
       "3     not_available                 0.589  \n",
       "4     not_available                 0.489  \n",
       "\n",
       "[5 rows x 165 columns]"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv(\"/Users/nicholasabad/Desktop/workspace/data/_DATA_FOR_PAPER_/FinalResults_long.csv\", delimiter=\"\\t\", nrows=100)\n",
    "\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Unnamed: 0',\n",
       " 'score',\n",
       " 'pid',\n",
       " 'cohort',\n",
       " '#CHROM',\n",
       " 'POS',\n",
       " 'GENE',\n",
       " 'REF',\n",
       " 'ALT',\n",
       " 'strand',\n",
       " 'FPKM',\n",
       " 'FPKM_Z_score',\n",
       " 'num_paths_of_recurrence',\n",
       " 'within_cgc_list',\n",
       " 'within_chromhmm_promoter',\n",
       " 'allele_frequency',\n",
       " 'SuRE_score',\n",
       " 'SuRE_ID',\n",
       " 'num_created_tfs_passing_tf_expression_threshold',\n",
       " 'created_tfs_passing_tf_expression_threshold',\n",
       " 'num_destroyed_tfs_passing_tf_expression_threshold',\n",
       " 'destroyed_tfs_passing_tf_expression_threshold',\n",
       " 'num_remaining_tfs',\n",
       " 'remaining_tfs',\n",
       " 'ID',\n",
       " 'QUAL',\n",
       " 'FILTER',\n",
       " 'INFO',\n",
       " '#Chromosome',\n",
       " 'Start_position',\n",
       " 'Reference_Allele',\n",
       " 'Tumor_Seq_Allele2',\n",
       " 'Tumor_Sample_Barcode',\n",
       " 'Project_Code',\n",
       " 'Donor_ID',\n",
       " 'Strand',\n",
       " 'Variant_Classification',\n",
       " 'Variant_Type',\n",
       " 'Tumor_Seq_Allele1',\n",
       " 'dbSNP_RS',\n",
       " 'dbSNP_Val_Status',\n",
       " 'Matched_Norm_Sample_Barcode',\n",
       " 'Genome_Change',\n",
       " 'ref_context',\n",
       " 'gc_content',\n",
       " 'i_1000genomes_AF',\n",
       " 'i_1000genomes_ID',\n",
       " 'i_Callers',\n",
       " 'i_GERM1000G',\n",
       " 'i_GERMOVLP',\n",
       " 'i_LOWSUPPORT',\n",
       " 'i_NORMALPANEL',\n",
       " 'i_NumCallers',\n",
       " 'i_OXOGFAIL',\n",
       " 'i_REMAPFAIL',\n",
       " 'i_SEXF',\n",
       " 'i_VAF',\n",
       " 'i_bPcr',\n",
       " 'i_bSeq',\n",
       " 'i_qual',\n",
       " 'i_repeat_masker',\n",
       " 'i_signature_N3',\n",
       " 'i_signature_R1',\n",
       " 'i_signature_R2',\n",
       " 'i_snv_near_indel',\n",
       " 't_alt_count',\n",
       " 't_ref_count',\n",
       " 'i_model_score',\n",
       " 'i_n_vaf',\n",
       " 'enhancers',\n",
       " 'gc19_pc.3utr',\n",
       " 'gc19_pc.5utr',\n",
       " 'gc19_pc.cds',\n",
       " 'gc19_pc.promCore',\n",
       " 'gc19_pc.promDomain',\n",
       " 'gc19_pc.ss',\n",
       " 'lncrna.ncrna',\n",
       " 'lncrna.promCore',\n",
       " 'lncrna.promDomain',\n",
       " 'lncrna.ss',\n",
       " 'mirna.mat',\n",
       " 'mirna.pre',\n",
       " 'mirna.prom',\n",
       " 'smallrna.ncrna',\n",
       " 'roadmap_enhancer',\n",
       " 'roadmap_promoter',\n",
       " 'ucsc_cpgi',\n",
       " 'wgEncodeRegDnaseClusteredV3',\n",
       " 'MAPABILITY',\n",
       " 'HISEQDEPTH',\n",
       " 'SIMPLE_TANDEMREPEATS',\n",
       " 'REPEAT_MASKER',\n",
       " 'DUKE_EXCLUDED',\n",
       " 'DAC_BLACKLIST',\n",
       " 'SELFCHAIN',\n",
       " 'CpGislands',\n",
       " 'CgiMountains',\n",
       " 'Enhancers',\n",
       " 'miRNAs_snoRNAs',\n",
       " 'miRBase18',\n",
       " 'miRNAtargets',\n",
       " 'phastConsElem20bp',\n",
       " 'TFBScons',\n",
       " 'COSMIC',\n",
       " 'ENCODE_DNASE',\n",
       " 'ENCODE_TFBS',\n",
       " 'PANCAN_ENCODE_TFBS',\n",
       " 'PANCAN_NC_TRANSCRIPT',\n",
       " 'PANCAN_CODING_GENES_FLAT',\n",
       " 'PANCAN_FLAT_PROMOTERS',\n",
       " 'PANCAN_LINC_RNA_FLAT',\n",
       " 'PANCAN_CALLABLE_GENOME',\n",
       " 'PANCAN_MAPPABILITY',\n",
       " 'PANCAN_MIRBASE_20_MIRNA',\n",
       " 'PANCAN_REPLICATION_TIMING',\n",
       " 'PANCAN_ROADMAP_STR_DYADICS',\n",
       " 'PANCAN_ROADMAP_STR_ENHANCERS',\n",
       " 'PANCAN_ROADMAP_STR_PROMOTERS',\n",
       " 'PANCAN_ENCODE_TFB_SITES',\n",
       " 'DBSNP',\n",
       " '1K_GENOMES',\n",
       " 'ANNOVAR_FUNCTION',\n",
       " 'EXONIC_CLASSIFICATION',\n",
       " 'ANNOVAR_TRANSCRIPTS',\n",
       " 'SEGDUP',\n",
       " 'CYTOBAND',\n",
       " 'SEQUENCE_CONTEXT',\n",
       " 'JASPAR2020_CORE_vertebrates_non_redundant',\n",
       " 'within_pcawg_whitelist',\n",
       " 'purity_median_after_esd',\n",
       " 'purity_sd_after_esd',\n",
       " 'purity_median_before_esd',\n",
       " 'path_to_purity_file',\n",
       " 'JASPAR2020_CORE_vertebrates_non_redundant(tf_name,binding_affinity,seq1,seq2,raw,zscore,log)',\n",
       " 'promoter_start',\n",
       " 'promoter_end',\n",
       " 'distance_from_gene_start',\n",
       " 'distance_from_start_of_promoter',\n",
       " 'element',\n",
       " 'genome_region',\n",
       " 'within_basic_promoter',\n",
       " 'promoter_start.1',\n",
       " 'promoter_end.1',\n",
       " 'distance_from_gene_start.1',\n",
       " 'distance_from_start_of_promoter.1',\n",
       " 'element.1',\n",
       " 'genome_region.1',\n",
       " 'paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))',\n",
       " 'number_of_pcawg_recurrent_mutations',\n",
       " 'path_to_pcawg_recurrence_dict',\n",
       " 'JASPAR2020_CORE_vertebrates_non_redundant(tf_name,binding_affinity,seq1,seq2,raw,zscore,log,tf_sequence_logo)',\n",
       " 'transcription_factor_pfms',\n",
       " 'jaspar_matrix_ids',\n",
       " 'num_original_mutations',\n",
       " 'num_final_mutations',\n",
       " 'num_promoter_mutations',\n",
       " 'num_recurrent_mutations',\n",
       " 'expression_traces',\n",
       " 'tf_summary',\n",
       " 'ncbi_gene_summary',\n",
       " 'ncbi_url',\n",
       " 'in_old_results',\n",
       " 'cn_score',\n",
       " 'cn_file_of_score',\n",
       " 'ICGC_Estimated_Purity']"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(data.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "tert_c228t = data.iloc[3]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2s/38kg81v526j7qcgd8gdq8jd00000gn/T/ipykernel_1243/437269650.py:12: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '-0.230602071237921' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.\n",
      "  tert_c228t_dataframe.loc[idx+1, \"zscore\"] = zscore\n"
     ]
    }
   ],
   "source": [
    "tert_c228t_dataframe = pd.DataFrame()\n",
    "tert_c228t_dataframe.loc[0, \"pid\"] = tert_c228t[\"pid\"]\n",
    "tert_c228t_dataframe.loc[0, \"cohort\"] = tert_c228t[\"cohort\"]\n",
    "tert_c228t_dataframe.loc[0, \"zscore\"] = tert_c228t[\"FPKM_Z_score\"]\n",
    "\n",
    "\n",
    "\n",
    "for idx, entry in enumerate(tert_c228t[\"paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))\"].split(\";\")):\n",
    "    path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af = entry.split(\",\")[:14]\n",
    "    tert_c228t_dataframe.loc[idx+1, \"pid\"] = pid\n",
    "    tert_c228t_dataframe.loc[idx+1, \"cohort\"] = cohort\n",
    "    tert_c228t_dataframe.loc[idx+1, \"zscore\"] = zscore\n",
    "    \n",
    "pids_of_tert_c228ts_passing_pipeline = list(data[(data[\"POS\"] == 1295228) & (data[\"GENE\"] == \"TERT\")][\"pid\"].unique())\n",
    "\n",
    "for idx, row in tert_c228t_dataframe.iterrows():\n",
    "    tert_c228t_dataframe.loc[idx, \"passes_pipeline\"] = row[\"pid\"] in pids_of_tert_c228ts_passing_pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "passes_pipeline\n",
       "False    73\n",
       "True     23\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tert_c228t_dataframe[\"passes_pipeline\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Are they the same?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2s/38kg81v526j7qcgd8gdq8jd00000gn/T/ipykernel_1243/2502772735.py:1: DtypeWarning: Columns (88) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  pcawg_results = pd.read_csv(\"/Users/nicholasabad/Desktop/workspace/data/_DATA_FOR_PAPER_/FinalResults_long.csv\", delimiter=\"\\t\")\n"
     ]
    }
   ],
   "source": [
    "pcawg_results = pd.read_csv(\"/Users/nicholasabad/Desktop/workspace/data/_DATA_FOR_PAPER_/FinalResults_long.csv\", delimiter=\"\\t\")\n",
    "pcawg_results.drop(\"Unnamed: 0\", axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>pid</th>\n",
       "      <th>cohort</th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>GENE</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>strand</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>...</th>\n",
       "      <th>num_promoter_mutations</th>\n",
       "      <th>num_recurrent_mutations</th>\n",
       "      <th>expression_traces</th>\n",
       "      <th>tf_summary</th>\n",
       "      <th>ncbi_gene_summary</th>\n",
       "      <th>ncbi_url</th>\n",
       "      <th>in_old_results</th>\n",
       "      <th>cn_score</th>\n",
       "      <th>cn_file_of_score</th>\n",
       "      <th>ICGC_Estimated_Purity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>93.45</td>\n",
       "      <td>45a7949d-e63f-4956-866c-df51257032de</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>2</td>\n",
       "      <td>71204529</td>\n",
       "      <td>ANKRD53</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>1.428501</td>\n",
       "      <td>...</td>\n",
       "      <td>284</td>\n",
       "      <td>5.0</td>\n",
       "      <td>{'ANKRD53': {'BLCA-US': {'raw': [0.13674067793...</td>\n",
       "      <td>{'REL': {'summary': \"This gene encodes a prote...</td>\n",
       "      <td>Involved in mitotic metaphase plate congressio...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/79998</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>93.41</td>\n",
       "      <td>6c884037-49f9-41c3-b0e5-9cbcd545aeb7</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>9</td>\n",
       "      <td>91933357</td>\n",
       "      <td>SECISBP2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>15.079190</td>\n",
       "      <td>...</td>\n",
       "      <td>879</td>\n",
       "      <td>6.0</td>\n",
       "      <td>{'SECISBP2': {'SKCM-US': {'raw': [5.0687620882...</td>\n",
       "      <td>{'ELK4': {'summary': 'This gene is a member of...</td>\n",
       "      <td>The protein encoded by this gene is one of the...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/79048</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>88.10</td>\n",
       "      <td>49dec0c2-8e75-4f44-a253-82b2ea605890</td>\n",
       "      <td>LUAD-US</td>\n",
       "      <td>2</td>\n",
       "      <td>71204529</td>\n",
       "      <td>ANKRD53</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>1.764278</td>\n",
       "      <td>...</td>\n",
       "      <td>436</td>\n",
       "      <td>5.0</td>\n",
       "      <td>{'ANKRD53': {'LUAD-US': {'raw': [0.73980555761...</td>\n",
       "      <td>{'REL': {'summary': \"This gene encodes a prote...</td>\n",
       "      <td>Involved in mitotic metaphase plate congressio...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/79998</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>86.45</td>\n",
       "      <td>d692ecd0-2433-426b-9af2-a30c44a80f7c</td>\n",
       "      <td>THCA-US</td>\n",
       "      <td>5</td>\n",
       "      <td>1295228</td>\n",
       "      <td>TERT</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>0.380236</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>95.0</td>\n",
       "      <td>{'TERT': {'THCA-US': {'raw': [0.0, 0.006731104...</td>\n",
       "      <td>{'ELK4': {'summary': 'This gene is a member of...</td>\n",
       "      <td>Telomerase is a ribonucleoprotein polymerase t...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/7015</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>86.30</td>\n",
       "      <td>9988eb07-01f6-4f83-8699-bb63e0525f08</td>\n",
       "      <td>HNSC-US</td>\n",
       "      <td>7</td>\n",
       "      <td>145813786</td>\n",
       "      <td>CNTNAP2</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>21.329213</td>\n",
       "      <td>...</td>\n",
       "      <td>459</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'CNTNAP2': {'HNSC-US': {'raw': [0.38939713831...</td>\n",
       "      <td>{'ZEB1': {'summary': 'This gene encodes a zinc...</td>\n",
       "      <td>This gene encodes a member of the neurexin fam...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/26047</td>\n",
       "      <td>False</td>\n",
       "      <td>9</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18409</th>\n",
       "      <td>2.02</td>\n",
       "      <td>6ca5c1bb-275b-4d05-948a-3c6c7d03fab9</td>\n",
       "      <td>COAD-US</td>\n",
       "      <td>19</td>\n",
       "      <td>39615618</td>\n",
       "      <td>PAK4</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>14.560673</td>\n",
       "      <td>...</td>\n",
       "      <td>6298</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'PAK4': {'COAD-US': {'raw': [23.9501634252, 4...</td>\n",
       "      <td>{'ZNF384': {'summary': \"This gene encodes a C2...</td>\n",
       "      <td>Not Available</td>\n",
       "      <td>Not Available</td>\n",
       "      <td>False</td>\n",
       "      <td>not_available</td>\n",
       "      <td>not_available</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18410</th>\n",
       "      <td>2.02</td>\n",
       "      <td>978ae91e-6ebe-4efa-97ff-cfad511ae7b3</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>17</td>\n",
       "      <td>1645475</td>\n",
       "      <td>SERPINF2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>1.000828</td>\n",
       "      <td>...</td>\n",
       "      <td>2157</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'SERPINF2': {'SKCM-US': {'raw': [0.1213962340...</td>\n",
       "      <td>{'MZF1': {'summary': 'Enables DNA-binding tran...</td>\n",
       "      <td>Not Available</td>\n",
       "      <td>Not Available</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18411</th>\n",
       "      <td>2.01</td>\n",
       "      <td>a96eddfc-3afb-4bf8-a440-c91778113fbd</td>\n",
       "      <td>LUSC-US</td>\n",
       "      <td>2</td>\n",
       "      <td>155824463</td>\n",
       "      <td>AC092625.1</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>0.110200</td>\n",
       "      <td>...</td>\n",
       "      <td>611</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'AC092625.1': {'LUSC-US': {'raw': [0.11019988...</td>\n",
       "      <td>{'ZNF384': {'summary': \"This gene encodes a C2...</td>\n",
       "      <td>Not Available</td>\n",
       "      <td>Not Available</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "      <td>not_available</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18412</th>\n",
       "      <td>2.00</td>\n",
       "      <td>bb659e7c-0b97-41b5-ac10-a66bb1f6f283</td>\n",
       "      <td>RECA-EU</td>\n",
       "      <td>14</td>\n",
       "      <td>96829898</td>\n",
       "      <td>ATG2B</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>3.022303</td>\n",
       "      <td>...</td>\n",
       "      <td>102</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'ATG2B': {'RECA-EU': {'raw': [4.19346821296, ...</td>\n",
       "      <td>{'ZBTB14': {'summary': 'Enables DNA-binding tr...</td>\n",
       "      <td>Not Available</td>\n",
       "      <td>Not Available</td>\n",
       "      <td>False</td>\n",
       "      <td>not_available</td>\n",
       "      <td>not_available</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18413</th>\n",
       "      <td>2.00</td>\n",
       "      <td>14dd86c1-dfb3-4124-a453-4d64223bc048</td>\n",
       "      <td>KIRP-US</td>\n",
       "      <td>1</td>\n",
       "      <td>202317452</td>\n",
       "      <td>PPP1R12B</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>+</td>\n",
       "      <td>1.224841</td>\n",
       "      <td>...</td>\n",
       "      <td>46</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'PPP1R12B': {'KIRP-US': {'raw': [0.9522805347...</td>\n",
       "      <td>{'GMEB2': {'summary': 'This gene is a member o...</td>\n",
       "      <td>Myosin phosphatase is a protein complex compri...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/4660</td>\n",
       "      <td>False</td>\n",
       "      <td>not_available</td>\n",
       "      <td>not_available</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18414 rows Ã— 164 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       score                                   pid   cohort #CHROM        POS  \\\n",
       "0      93.45  45a7949d-e63f-4956-866c-df51257032de  BLCA-US      2   71204529   \n",
       "1      93.41  6c884037-49f9-41c3-b0e5-9cbcd545aeb7  SKCM-US      9   91933357   \n",
       "2      88.10  49dec0c2-8e75-4f44-a253-82b2ea605890  LUAD-US      2   71204529   \n",
       "3      86.45  d692ecd0-2433-426b-9af2-a30c44a80f7c  THCA-US      5    1295228   \n",
       "4      86.30  9988eb07-01f6-4f83-8699-bb63e0525f08  HNSC-US      7  145813786   \n",
       "...      ...                                   ...      ...    ...        ...   \n",
       "18409   2.02  6ca5c1bb-275b-4d05-948a-3c6c7d03fab9  COAD-US     19   39615618   \n",
       "18410   2.02  978ae91e-6ebe-4efa-97ff-cfad511ae7b3  SKCM-US     17    1645475   \n",
       "18411   2.01  a96eddfc-3afb-4bf8-a440-c91778113fbd  LUSC-US      2  155824463   \n",
       "18412   2.00  bb659e7c-0b97-41b5-ac10-a66bb1f6f283  RECA-EU     14   96829898   \n",
       "18413   2.00  14dd86c1-dfb3-4124-a453-4d64223bc048  KIRP-US      1  202317452   \n",
       "\n",
       "             GENE REF ALT strand       FPKM  ...  num_promoter_mutations  \\\n",
       "0         ANKRD53   G   A      +   1.428501  ...                     284   \n",
       "1        SECISBP2   G   A      +  15.079190  ...                     879   \n",
       "2         ANKRD53   G   A      +   1.764278  ...                     436   \n",
       "3            TERT   G   A      -   0.380236  ...                      20   \n",
       "4         CNTNAP2   C   A      +  21.329213  ...                     459   \n",
       "...           ...  ..  ..    ...        ...  ...                     ...   \n",
       "18409        PAK4   T   A      +  14.560673  ...                    6298   \n",
       "18410    SERPINF2   G   A      +   1.000828  ...                    2157   \n",
       "18411  AC092625.1   C   A      -   0.110200  ...                     611   \n",
       "18412       ATG2B   G   T      -   3.022303  ...                     102   \n",
       "18413    PPP1R12B   T   G      +   1.224841  ...                      46   \n",
       "\n",
       "       num_recurrent_mutations  \\\n",
       "0                          5.0   \n",
       "1                          6.0   \n",
       "2                          5.0   \n",
       "3                         95.0   \n",
       "4                          0.0   \n",
       "...                        ...   \n",
       "18409                      0.0   \n",
       "18410                      0.0   \n",
       "18411                      0.0   \n",
       "18412                      0.0   \n",
       "18413                      0.0   \n",
       "\n",
       "                                       expression_traces  \\\n",
       "0      {'ANKRD53': {'BLCA-US': {'raw': [0.13674067793...   \n",
       "1      {'SECISBP2': {'SKCM-US': {'raw': [5.0687620882...   \n",
       "2      {'ANKRD53': {'LUAD-US': {'raw': [0.73980555761...   \n",
       "3      {'TERT': {'THCA-US': {'raw': [0.0, 0.006731104...   \n",
       "4      {'CNTNAP2': {'HNSC-US': {'raw': [0.38939713831...   \n",
       "...                                                  ...   \n",
       "18409  {'PAK4': {'COAD-US': {'raw': [23.9501634252, 4...   \n",
       "18410  {'SERPINF2': {'SKCM-US': {'raw': [0.1213962340...   \n",
       "18411  {'AC092625.1': {'LUSC-US': {'raw': [0.11019988...   \n",
       "18412  {'ATG2B': {'RECA-EU': {'raw': [4.19346821296, ...   \n",
       "18413  {'PPP1R12B': {'KIRP-US': {'raw': [0.9522805347...   \n",
       "\n",
       "                                              tf_summary  \\\n",
       "0      {'REL': {'summary': \"This gene encodes a prote...   \n",
       "1      {'ELK4': {'summary': 'This gene is a member of...   \n",
       "2      {'REL': {'summary': \"This gene encodes a prote...   \n",
       "3      {'ELK4': {'summary': 'This gene is a member of...   \n",
       "4      {'ZEB1': {'summary': 'This gene encodes a zinc...   \n",
       "...                                                  ...   \n",
       "18409  {'ZNF384': {'summary': \"This gene encodes a C2...   \n",
       "18410  {'MZF1': {'summary': 'Enables DNA-binding tran...   \n",
       "18411  {'ZNF384': {'summary': \"This gene encodes a C2...   \n",
       "18412  {'ZBTB14': {'summary': 'Enables DNA-binding tr...   \n",
       "18413  {'GMEB2': {'summary': 'This gene is a member o...   \n",
       "\n",
       "                                       ncbi_gene_summary  \\\n",
       "0      Involved in mitotic metaphase plate congressio...   \n",
       "1      The protein encoded by this gene is one of the...   \n",
       "2      Involved in mitotic metaphase plate congressio...   \n",
       "3      Telomerase is a ribonucleoprotein polymerase t...   \n",
       "4      This gene encodes a member of the neurexin fam...   \n",
       "...                                                  ...   \n",
       "18409                                      Not Available   \n",
       "18410                                      Not Available   \n",
       "18411                                      Not Available   \n",
       "18412                                      Not Available   \n",
       "18413  Myosin phosphatase is a protein complex compri...   \n",
       "\n",
       "                                      ncbi_url in_old_results       cn_score  \\\n",
       "0      https://www.ncbi.nlm.nih.gov/gene/79998           True              3   \n",
       "1      https://www.ncbi.nlm.nih.gov/gene/79048           True              6   \n",
       "2      https://www.ncbi.nlm.nih.gov/gene/79998           True              4   \n",
       "3       https://www.ncbi.nlm.nih.gov/gene/7015           True              3   \n",
       "4      https://www.ncbi.nlm.nih.gov/gene/26047          False              9   \n",
       "...                                        ...            ...            ...   \n",
       "18409                            Not Available          False  not_available   \n",
       "18410                            Not Available          False              4   \n",
       "18411                            Not Available          False              3   \n",
       "18412                            Not Available          False  not_available   \n",
       "18413   https://www.ncbi.nlm.nih.gov/gene/4660          False  not_available   \n",
       "\n",
       "      cn_file_of_score  ICGC_Estimated_Purity  \n",
       "0        not_available                  0.938  \n",
       "1        not_available                   0.78  \n",
       "2        not_available                    0.5  \n",
       "3        not_available                  0.589  \n",
       "4        not_available                  0.489  \n",
       "...                ...                    ...  \n",
       "18409    not_available          not_available  \n",
       "18410    not_available                  0.239  \n",
       "18411    not_available          not_available  \n",
       "18412    not_available          not_available  \n",
       "18413    not_available          not_available  \n",
       "\n",
       "[18414 rows x 164 columns]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pcawg_results = pcawg_results[(pcawg_results[\"REF\"].str.len() == 1) & (pcawg_results[\"ALT\"].str.len() == 1)].reset_index(drop=True)\n",
    "pcawg_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "pcawg_results.rename(\n",
    "    {\n",
    "        \"within_chromhmm_promoter\": \"open_chromatin\"\n",
    "    },\n",
    "    axis=1,\n",
    "    inplace=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Combine subset_long with EOPC-DE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [],
   "source": [
    "eopc_de = pd.read_csv(\"/Users/nicholasabad/Desktop/workspace/data/_DATA_FOR_PAPER_/SupplementaryTables/2_eopc_de_remaining_mutations_28Feb2024.csv\")\n",
    "eopc_de = eopc_de[(eopc_de[\"REF\"].str.len() == 1) & (eopc_de[\"ALT\"].str.len() == 1)].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: All columns were taken from EOPC-DE dataframe except for cn_score.\n",
    "\n",
    "necessary_columns = [\n",
    "    'score',\n",
    "    'pid',\n",
    "    'cohort',\n",
    "    '#CHROM',\n",
    "    'POS',\n",
    "    'GENE',\n",
    "    'SEQUENCE_CONTEXT',\n",
    "    'REF',\n",
    "    'ALT',\n",
    "    'strand',\n",
    "    'FPKM',\n",
    "    'FPKM_Z_score',\n",
    "    'within_cgc_list',\n",
    "    'open_chromatin',\n",
    "    'allele_frequency',\n",
    "    'ICGC_Estimated_Purity',\n",
    "    'num_created_tfs_passing_tf_expression_threshold',\n",
    "    'created_tfs_passing_tf_expression_threshold',\n",
    "    'num_destroyed_tfs_passing_tf_expression_threshold',\n",
    "    'destroyed_tfs_passing_tf_expression_threshold',\n",
    "    'num_remaining_tfs',\n",
    "    'remaining_tfs',\n",
    "    'JASPAR2020_CORE_vertebrates_non_redundant(tf_name,binding_affinity,seq1,seq2,raw,zscore,log,tf_sequence_logo)',\n",
    "    'paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))',\n",
    "    'num_original_mutations',\n",
    "    'num_final_mutations',\n",
    "    'num_promoter_mutations',\n",
    "    'num_recurrent_mutations',\n",
    "    'cn_score'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_pcawg_and_eopc = pd.concat([pcawg_results, eopc_de], ignore_index=True)\n",
    "combined_pcawg_and_eopc.sort_values(\"score\", ascending=False, inplace=True)\n",
    "\n",
    "combined_pcawg_and_eopc = combined_pcawg_and_eopc[necessary_columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_pcawg_and_eopc.to_csv(\n",
    "    \"/Users/nicholasabad/Desktop/workspace/data/_DATA_FOR_PAPER_/pcawg_with_eopc_de_all_mutations_18March2025.csv\",\n",
    "    sep=\",\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>pid</th>\n",
       "      <th>cohort</th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SEQUENCE_CONTEXT</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>strand</th>\n",
       "      <th>...</th>\n",
       "      <th>destroyed_tfs_passing_tf_expression_threshold</th>\n",
       "      <th>num_remaining_tfs</th>\n",
       "      <th>remaining_tfs</th>\n",
       "      <th>JASPAR2020_CORE_vertebrates_non_redundant(tf_name,binding_affinity,seq1,seq2,raw,zscore,log,tf_sequence_logo)</th>\n",
       "      <th>paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))</th>\n",
       "      <th>num_original_mutations</th>\n",
       "      <th>num_final_mutations</th>\n",
       "      <th>num_promoter_mutations</th>\n",
       "      <th>num_recurrent_mutations</th>\n",
       "      <th>cn_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>93.45</td>\n",
       "      <td>45a7949d-e63f-4956-866c-df51257032de</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>2</td>\n",
       "      <td>71204529</td>\n",
       "      <td>ANKRD53</td>\n",
       "      <td>GGGAAAGGGA,ACTCCCTTTC</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NFKB1,NFKB2</td>\n",
       "      <td>REL,13.2727,.,GGGAGTTTCC,0.298502304678,-1.254...</td>\n",
       "      <td>/omics/groups/OE0436/internal/nabad/results/16...</td>\n",
       "      <td>11050</td>\n",
       "      <td>76</td>\n",
       "      <td>284</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>93.41</td>\n",
       "      <td>6c884037-49f9-41c3-b0e5-9cbcd545aeb7</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>9</td>\n",
       "      <td>91933357</td>\n",
       "      <td>SECISBP2</td>\n",
       "      <td>CCGGCCGGAA,AAAGGGCGGG</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>ELK4,ZBTB7A</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ELK1,ELK3</td>\n",
       "      <td>ELK1,0.10067117472187072,GCCGGAAGAA,.,14.99847...</td>\n",
       "      <td>/omics/groups/OE0436/internal/nabad/results/16...</td>\n",
       "      <td>40712</td>\n",
       "      <td>223</td>\n",
       "      <td>879</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>88.10</td>\n",
       "      <td>49dec0c2-8e75-4f44-a253-82b2ea605890</td>\n",
       "      <td>LUAD-US</td>\n",
       "      <td>2</td>\n",
       "      <td>71204529</td>\n",
       "      <td>ANKRD53</td>\n",
       "      <td>GGGAAAGGGA,ACTCCCTTTC</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NFKB1,NFKB2</td>\n",
       "      <td>REL,13.2727,.,GGGAGTTTCC,1.13424410656,-0.0631...</td>\n",
       "      <td>/omics/groups/OE0436/internal/nabad/results/16...</td>\n",
       "      <td>22203</td>\n",
       "      <td>91</td>\n",
       "      <td>436</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>86.45</td>\n",
       "      <td>d692ecd0-2433-426b-9af2-a30c44a80f7c</td>\n",
       "      <td>THCA-US</td>\n",
       "      <td>5</td>\n",
       "      <td>1295228</td>\n",
       "      <td>TERT</td>\n",
       "      <td>AGGGCCCGGA,GGGGCTGGGC</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>SP8,TFAP2A,TFAP2C,CTCFL,MAZ,ZNF148</td>\n",
       "      <td>9.0</td>\n",
       "      <td>TFAP2A,ELF1,ELF4,ETV6,KLF16,ZBTB7A,TFAP2B,TFAP...</td>\n",
       "      <td>TFAP2A,0.09557122923715045,GCCCCCTCCGGGCC,.,0....</td>\n",
       "      <td>/omics/groups/OE0436/internal/nabad/results/16...</td>\n",
       "      <td>1099</td>\n",
       "      <td>5</td>\n",
       "      <td>20</td>\n",
       "      <td>95.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>86.30</td>\n",
       "      <td>9988eb07-01f6-4f83-8699-bb63e0525f08</td>\n",
       "      <td>HNSC-US</td>\n",
       "      <td>7</td>\n",
       "      <td>145813786</td>\n",
       "      <td>CNTNAP2</td>\n",
       "      <td>GCTGAGACTC,CCTGCCGCTC</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>ZNF148</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NHLH2,SNAI3</td>\n",
       "      <td>ZEB1,14.5102,.,CTCACCTGCCG,0.832158029939,-0.7...</td>\n",
       "      <td>.</td>\n",
       "      <td>23779</td>\n",
       "      <td>92</td>\n",
       "      <td>459</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18410</th>\n",
       "      <td>2.02</td>\n",
       "      <td>978ae91e-6ebe-4efa-97ff-cfad511ae7b3</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>17</td>\n",
       "      <td>1645475</td>\n",
       "      <td>SERPINF2</td>\n",
       "      <td>TCTTGGGTAG,GGAAAGAGAC</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>MZF1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MAZ</td>\n",
       "      <td>MZF1,0.08042270172024159,GTAGGGGAAA,.,1.908352...</td>\n",
       "      <td>.</td>\n",
       "      <td>110086</td>\n",
       "      <td>472</td>\n",
       "      <td>2157</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18409</th>\n",
       "      <td>2.02</td>\n",
       "      <td>6ca5c1bb-275b-4d05-948a-3c6c7d03fab9</td>\n",
       "      <td>COAD-US</td>\n",
       "      <td>19</td>\n",
       "      <td>39615618</td>\n",
       "      <td>PAK4</td>\n",
       "      <td>ATACAAAAAA,TAGCTGGGTG</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZNF384,11.5106,.,TACAAAAAAATA,7.03125452052,-1...</td>\n",
       "      <td>.</td>\n",
       "      <td>413781</td>\n",
       "      <td>1320</td>\n",
       "      <td>6298</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18411</th>\n",
       "      <td>2.01</td>\n",
       "      <td>a96eddfc-3afb-4bf8-a440-c91778113fbd</td>\n",
       "      <td>LUSC-US</td>\n",
       "      <td>2</td>\n",
       "      <td>155824463</td>\n",
       "      <td>AC092625.1</td>\n",
       "      <td>GTTGAAATTA,AAAAATGTTG</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZNF384,11.7021,.,AATTAAAAAAAT,14.9732578533,0....</td>\n",
       "      <td>.</td>\n",
       "      <td>30011</td>\n",
       "      <td>145</td>\n",
       "      <td>611</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18412</th>\n",
       "      <td>2.00</td>\n",
       "      <td>bb659e7c-0b97-41b5-ac10-a66bb1f6f283</td>\n",
       "      <td>RECA-EU</td>\n",
       "      <td>14</td>\n",
       "      <td>96829898</td>\n",
       "      <td>ATG2B</td>\n",
       "      <td>AGGACGCCGG,CGCGCAGGTG</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>ZBTB14</td>\n",
       "      <td>2.0</td>\n",
       "      <td>PAX5,KLF15</td>\n",
       "      <td>PAX5,10.5091,.,CTGCGCGACCGG,0.00464238213385,-...</td>\n",
       "      <td>.</td>\n",
       "      <td>4949</td>\n",
       "      <td>16</td>\n",
       "      <td>102</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18413</th>\n",
       "      <td>2.00</td>\n",
       "      <td>14dd86c1-dfb3-4124-a453-4d64223bc048</td>\n",
       "      <td>KIRP-US</td>\n",
       "      <td>1</td>\n",
       "      <td>202317452</td>\n",
       "      <td>PPP1R12B</td>\n",
       "      <td>GTTTAGGTAC,TAATGTTTAG</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GMEB2,11.5152,.,TTACGTAC,2.48928711784,-1.6381...</td>\n",
       "      <td>.</td>\n",
       "      <td>2737</td>\n",
       "      <td>13</td>\n",
       "      <td>46</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18468 rows Ã— 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       score                                   pid   cohort #CHROM        POS  \\\n",
       "0      93.45  45a7949d-e63f-4956-866c-df51257032de  BLCA-US      2   71204529   \n",
       "1      93.41  6c884037-49f9-41c3-b0e5-9cbcd545aeb7  SKCM-US      9   91933357   \n",
       "2      88.10  49dec0c2-8e75-4f44-a253-82b2ea605890  LUAD-US      2   71204529   \n",
       "3      86.45  d692ecd0-2433-426b-9af2-a30c44a80f7c  THCA-US      5    1295228   \n",
       "4      86.30  9988eb07-01f6-4f83-8699-bb63e0525f08  HNSC-US      7  145813786   \n",
       "...      ...                                   ...      ...    ...        ...   \n",
       "18410   2.02  978ae91e-6ebe-4efa-97ff-cfad511ae7b3  SKCM-US     17    1645475   \n",
       "18409   2.02  6ca5c1bb-275b-4d05-948a-3c6c7d03fab9  COAD-US     19   39615618   \n",
       "18411   2.01  a96eddfc-3afb-4bf8-a440-c91778113fbd  LUSC-US      2  155824463   \n",
       "18412   2.00  bb659e7c-0b97-41b5-ac10-a66bb1f6f283  RECA-EU     14   96829898   \n",
       "18413   2.00  14dd86c1-dfb3-4124-a453-4d64223bc048  KIRP-US      1  202317452   \n",
       "\n",
       "             GENE       SEQUENCE_CONTEXT REF ALT strand  ...  \\\n",
       "0         ANKRD53  GGGAAAGGGA,ACTCCCTTTC   G   A      +  ...   \n",
       "1        SECISBP2  CCGGCCGGAA,AAAGGGCGGG   G   A      +  ...   \n",
       "2         ANKRD53  GGGAAAGGGA,ACTCCCTTTC   G   A      +  ...   \n",
       "3            TERT  AGGGCCCGGA,GGGGCTGGGC   G   A      -  ...   \n",
       "4         CNTNAP2  GCTGAGACTC,CCTGCCGCTC   C   A      +  ...   \n",
       "...           ...                    ...  ..  ..    ...  ...   \n",
       "18410    SERPINF2  TCTTGGGTAG,GGAAAGAGAC   G   A      +  ...   \n",
       "18409        PAK4  ATACAAAAAA,TAGCTGGGTG   T   A      +  ...   \n",
       "18411  AC092625.1  GTTGAAATTA,AAAAATGTTG   C   A      -  ...   \n",
       "18412       ATG2B  AGGACGCCGG,CGCGCAGGTG   G   T      -  ...   \n",
       "18413    PPP1R12B  GTTTAGGTAC,TAATGTTTAG   T   G      +  ...   \n",
       "\n",
       "       destroyed_tfs_passing_tf_expression_threshold  num_remaining_tfs  \\\n",
       "0                                                NaN                2.0   \n",
       "1                                        ELK4,ZBTB7A                2.0   \n",
       "2                                                NaN                2.0   \n",
       "3                 SP8,TFAP2A,TFAP2C,CTCFL,MAZ,ZNF148                9.0   \n",
       "4                                             ZNF148                2.0   \n",
       "...                                              ...                ...   \n",
       "18410                                           MZF1                1.0   \n",
       "18409                                            NaN                0.0   \n",
       "18411                                            NaN                0.0   \n",
       "18412                                         ZBTB14                2.0   \n",
       "18413                                            NaN                0.0   \n",
       "\n",
       "                                           remaining_tfs  \\\n",
       "0                                            NFKB1,NFKB2   \n",
       "1                                              ELK1,ELK3   \n",
       "2                                            NFKB1,NFKB2   \n",
       "3      TFAP2A,ELF1,ELF4,ETV6,KLF16,ZBTB7A,TFAP2B,TFAP...   \n",
       "4                                            NHLH2,SNAI3   \n",
       "...                                                  ...   \n",
       "18410                                                MAZ   \n",
       "18409                                                NaN   \n",
       "18411                                                NaN   \n",
       "18412                                         PAX5,KLF15   \n",
       "18413                                                NaN   \n",
       "\n",
       "       JASPAR2020_CORE_vertebrates_non_redundant(tf_name,binding_affinity,seq1,seq2,raw,zscore,log,tf_sequence_logo)  \\\n",
       "0      REL,13.2727,.,GGGAGTTTCC,0.298502304678,-1.254...                                                               \n",
       "1      ELK1,0.10067117472187072,GCCGGAAGAA,.,14.99847...                                                               \n",
       "2      REL,13.2727,.,GGGAGTTTCC,1.13424410656,-0.0631...                                                               \n",
       "3      TFAP2A,0.09557122923715045,GCCCCCTCCGGGCC,.,0....                                                               \n",
       "4      ZEB1,14.5102,.,CTCACCTGCCG,0.832158029939,-0.7...                                                               \n",
       "...                                                  ...                                                               \n",
       "18410  MZF1,0.08042270172024159,GTAGGGGAAA,.,1.908352...                                                               \n",
       "18409  ZNF384,11.5106,.,TACAAAAAAATA,7.03125452052,-1...                                                               \n",
       "18411  ZNF384,11.7021,.,AATTAAAAAAAT,14.9732578533,0....                                                               \n",
       "18412  PAX5,10.5091,.,CTGCGCGACCGG,0.00464238213385,-...                                                               \n",
       "18413  GMEB2,11.5152,.,TTACGTAC,2.48928711784,-1.6381...                                                               \n",
       "\n",
       "      paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))  \\\n",
       "0      /omics/groups/OE0436/internal/nabad/results/16...                                                                                        \n",
       "1      /omics/groups/OE0436/internal/nabad/results/16...                                                                                        \n",
       "2      /omics/groups/OE0436/internal/nabad/results/16...                                                                                        \n",
       "3      /omics/groups/OE0436/internal/nabad/results/16...                                                                                        \n",
       "4                                                      .                                                                                        \n",
       "...                                                  ...                                                                                        \n",
       "18410                                                  .                                                                                        \n",
       "18409                                                  .                                                                                        \n",
       "18411                                                  .                                                                                        \n",
       "18412                                                  .                                                                                        \n",
       "18413                                                  .                                                                                        \n",
       "\n",
       "      num_original_mutations  num_final_mutations num_promoter_mutations  \\\n",
       "0                      11050                   76                    284   \n",
       "1                      40712                  223                    879   \n",
       "2                      22203                   91                    436   \n",
       "3                       1099                    5                     20   \n",
       "4                      23779                   92                    459   \n",
       "...                      ...                  ...                    ...   \n",
       "18410                 110086                  472                   2157   \n",
       "18409                 413781                 1320                   6298   \n",
       "18411                  30011                  145                    611   \n",
       "18412                   4949                   16                    102   \n",
       "18413                   2737                   13                     46   \n",
       "\n",
       "       num_recurrent_mutations       cn_score  \n",
       "0                          5.0              3  \n",
       "1                          6.0              6  \n",
       "2                          5.0              4  \n",
       "3                         95.0              3  \n",
       "4                          0.0              9  \n",
       "...                        ...            ...  \n",
       "18410                      0.0              4  \n",
       "18409                      0.0  not_available  \n",
       "18411                      0.0              3  \n",
       "18412                      0.0  not_available  \n",
       "18413                      0.0  not_available  \n",
       "\n",
       "[18468 rows x 29 columns]"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_pcawg_and_eopc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Why did the TERT C228T mutations drop out?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2s/38kg81v526j7qcgd8gdq8jd00000gn/T/ipykernel_1243/1956300189.py:13: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '-0.230602071237921' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.\n",
      "  tert_c228t_dataframe.loc[idx+1, \"zscore\"] = zscore\n"
     ]
    }
   ],
   "source": [
    "tert_c228t = combined_pcawg_and_eopc.iloc[3]\n",
    "\n",
    "tert_c228t_dataframe = pd.DataFrame()\n",
    "tert_c228t_dataframe.loc[0, \"pid\"] = tert_c228t[\"pid\"]\n",
    "tert_c228t_dataframe.loc[0, \"cohort\"] = tert_c228t[\"cohort\"]\n",
    "tert_c228t_dataframe.loc[0, \"zscore\"] = tert_c228t[\"FPKM_Z_score\"]\n",
    "\n",
    "\n",
    "for idx, entry in enumerate(tert_c228t[\"paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))\"].split(\";\")):\n",
    "    path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af = entry.split(\",\")[:14]\n",
    "    tert_c228t_dataframe.loc[idx+1, \"pid\"] = pid\n",
    "    tert_c228t_dataframe.loc[idx+1, \"cohort\"] = cohort\n",
    "    tert_c228t_dataframe.loc[idx+1, \"zscore\"] = zscore\n",
    "    \n",
    "pids_of_tert_c228ts_passing_pipeline = list(data[(data[\"POS\"] == 1295228) & (data[\"GENE\"] == \"TERT\")][\"pid\"].unique())\n",
    "\n",
    "for idx, row in tert_c228t_dataframe.iterrows():\n",
    "    tert_c228t_dataframe.loc[idx, \"passes_pipeline\"] = row[\"pid\"] in pids_of_tert_c228ts_passing_pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(96, 4)"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Result: 96 possible TERT C228T mutations\n",
    "tert_c228t_dataframe.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(65, 4)"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Result: 65 TERT C228Ts have RNA-Seq data available.\n",
    "tert_c228t_dataframe[tert_c228t_dataframe[\"zscore\"] != \"not_available\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(28, 29)"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Result: 28 passed all filter steps.\n",
    "combined_pcawg_and_eopc[(combined_pcawg_and_eopc[\"GENE\"] == \"TERT\") & (combined_pcawg_and_eopc[\"POS\"] == 1295228)].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2s/38kg81v526j7qcgd8gdq8jd00000gn/T/ipykernel_1243/971158515.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  combined_pcawg_and_eopc.iloc[:100][(combined_pcawg_and_eopc[\"GENE\"] == \"TERT\") & (combined_pcawg_and_eopc[\"POS\"] == 1295228)].shape\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(24, 29)"
      ]
     },
     "execution_count": 156,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Result: 24 were within the top 100 candidates\n",
    "combined_pcawg_and_eopc.iloc[:100][(combined_pcawg_and_eopc[\"GENE\"] == \"TERT\") & (combined_pcawg_and_eopc[\"POS\"] == 1295228)].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2s/38kg81v526j7qcgd8gdq8jd00000gn/T/ipykernel_1243/2686636836.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  tert_c228t_dataframe_ge_data_available[\"zscore\"] = tert_c228t_dataframe_ge_data_available[\"zscore\"].astype(float)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(37, 4)"
      ]
     },
     "execution_count": 157,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Result: 37 of the TERT C228Ts with GE data available did not pass due to having a zscore < 0.\n",
    "tert_c228t_dataframe_ge_data_available = tert_c228t_dataframe[(tert_c228t_dataframe[\"zscore\"] != \"not_available\")]\n",
    "tert_c228t_dataframe_ge_data_available[\"zscore\"] = tert_c228t_dataframe_ge_data_available[\"zscore\"].astype(float)\n",
    "tert_c228t_dataframe_ge_data_available[tert_c228t_dataframe_ge_data_available[\"zscore\"] < 0].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Why did C250T mutations drop?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2s/38kg81v526j7qcgd8gdq8jd00000gn/T/ipykernel_1243/1779089474.py:15: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '0.2882712716542805' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.\n",
      "  tert_c250t_dataframe.loc[idx+1, \"zscore\"] = zscore\n"
     ]
    }
   ],
   "source": [
    "tert_c250t = combined_pcawg_and_eopc.iloc[29]\n",
    "\n",
    "tert_c250t_dataframe = pd.DataFrame()\n",
    "tert_c250t_dataframe.loc[0, \"pid\"] = tert_c250t[\"pid\"]\n",
    "tert_c250t_dataframe.loc[0, \"cohort\"] = tert_c250t[\"cohort\"]\n",
    "tert_c250t_dataframe.loc[0, \"zscore\"] = tert_c250t[\"FPKM_Z_score\"]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "for idx, entry in enumerate(tert_c250t[\"paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))\"].split(\";\")):\n",
    "    path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af = entry.split(\",\")[:14]\n",
    "    tert_c250t_dataframe.loc[idx+1, \"pid\"] = pid\n",
    "    tert_c250t_dataframe.loc[idx+1, \"cohort\"] = cohort\n",
    "    tert_c250t_dataframe.loc[idx+1, \"zscore\"] = zscore\n",
    "    \n",
    "pids_of_tert_c250ts_passing_pipeline = list(data[(data[\"POS\"] == 1295250) & (data[\"GENE\"] == \"TERT\")][\"pid\"].unique())\n",
    "\n",
    "for idx, row in tert_c250t_dataframe.iterrows():\n",
    "    tert_c250t_dataframe.loc[idx, \"passes_pipeline\"] = row[\"pid\"] in pids_of_tert_c250ts_passing_pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(35, 4)"
      ]
     },
     "execution_count": 167,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Result: 35 total TERT C250Ts\n",
    "tert_c250t_dataframe.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(26, 4)"
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Result: 26 have RNA-Seq data available\n",
    "tert_c250t_dataframe[tert_c250t_dataframe[\"zscore\"] != \"not_available\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 171,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "passes_pipeline\n",
       "False    20\n",
       "True      6\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 171,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Result: 6 passed the pipeline\n",
    "tert_c250t_dataframe[tert_c250t_dataframe[\"zscore\"] != \"not_available\"][\"passes_pipeline\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 172,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pid</th>\n",
       "      <th>cohort</th>\n",
       "      <th>zscore</th>\n",
       "      <th>passes_pipeline</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>fa4d35e3-fc38-4e26-a9e0-521a648acec1</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>0.754906</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>54a77010-2f16-4da9-8ca0-746a51801a00</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>0.2882712716542805</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>a8f86b64-914c-4d89-897b-33bcdd1759f7</td>\n",
       "      <td>GBM-US</td>\n",
       "      <td>0.0130328791644564</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>c065761d-f775-457f-bda0-4c7c257a701e</td>\n",
       "      <td>GBM-US</td>\n",
       "      <td>-0.5642192569946982</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>448fe471-3f4e-4dc8-a4e0-6f147dc93abe</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>-0.4010794865307736</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3f2ecbcd-5c9e-4d57-8aa5-962e2e86a83c</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.5131991180172888</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>6c884037-49f9-41c3-b0e5-9cbcd545aeb7</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.5107224658524304</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>507620ef-f0fc-4ac3-a7c0-ea2e36c40f77</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>0.0422108897825063</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>88f2c496-3eb4-4c03-b3a5-8d5a55803aaf</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.6129595099095041</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>c21a5d55-9790-429f-a4e6-420293c07b53</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>0.2559681573260988</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>5fd77ba9-5015-4d8b-86a0-582e5c76bdd6</td>\n",
       "      <td>GBM-US</td>\n",
       "      <td>0.6562346152437472</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0dd0718d-5ddf-4c59-8c47-0f51303daeb5</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.4569483162353123</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>8294a969-59a1-4d3b-895b-f07e96bf9f7a</td>\n",
       "      <td>HNSC-US</td>\n",
       "      <td>-0.5371036860447784</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>8c46a1df-1c10-4188-992a-a55b782494c9</td>\n",
       "      <td>LGG-US</td>\n",
       "      <td>0.0668132218607219</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>e2774050-1257-4c29-b201-d79ce70fe17e</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.4379410194660893</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>5310f698-d31a-4628-84b4-dc9ac29722a5</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.0068347114243561</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>d432e99a-67fb-4609-b90f-99438eee7cae</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.5151538970017837</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>978ae91e-6ebe-4efa-97ff-cfad511ae7b3</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.5839295199428811</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>d1132127-1250-43af-9c16-425798a3d1a7</td>\n",
       "      <td>GBM-US</td>\n",
       "      <td>-0.0624706219524289</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>2b142863-b963-4cc9-8f8f-c72503c93390</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>-0.3533774556035514</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>d0e033bc-8e97-4322-b3b1-8626f1e2449b</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.0726002501514517</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>41d50a3c-49fc-4d8a-9c25-bbbcd3a5da6f</td>\n",
       "      <td>GBM-US</td>\n",
       "      <td>-0.2907228368770588</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>0c7aca3f-e006-4de3-afc2-20b4f727d4fd</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>-0.2457822675716241</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>affa4ef4-02ec-40f1-b6a2-739f8bf2afa9</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.1417866820777039</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>10bb1a92-901e-4a14-80f4-5e88f997754b</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.4366975067097288</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>fc249113-83d4-4abe-8c80-a4f7305dcd91</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.4311616315076992</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     pid   cohort               zscore  \\\n",
       "0   fa4d35e3-fc38-4e26-a9e0-521a648acec1  SKCM-US             0.754906   \n",
       "1   54a77010-2f16-4da9-8ca0-746a51801a00  SKCM-US   0.2882712716542805   \n",
       "2   a8f86b64-914c-4d89-897b-33bcdd1759f7   GBM-US   0.0130328791644564   \n",
       "3   c065761d-f775-457f-bda0-4c7c257a701e   GBM-US  -0.5642192569946982   \n",
       "4   448fe471-3f4e-4dc8-a4e0-6f147dc93abe  BLCA-US  -0.4010794865307736   \n",
       "6   3f2ecbcd-5c9e-4d57-8aa5-962e2e86a83c  SKCM-US  -0.5131991180172888   \n",
       "7   6c884037-49f9-41c3-b0e5-9cbcd545aeb7  SKCM-US  -0.5107224658524304   \n",
       "8   507620ef-f0fc-4ac3-a7c0-ea2e36c40f77  SKCM-US   0.0422108897825063   \n",
       "9   88f2c496-3eb4-4c03-b3a5-8d5a55803aaf  SKCM-US  -0.6129595099095041   \n",
       "11  c21a5d55-9790-429f-a4e6-420293c07b53  SKCM-US   0.2559681573260988   \n",
       "12  5fd77ba9-5015-4d8b-86a0-582e5c76bdd6   GBM-US   0.6562346152437472   \n",
       "14  0dd0718d-5ddf-4c59-8c47-0f51303daeb5  SKCM-US  -0.4569483162353123   \n",
       "15  8294a969-59a1-4d3b-895b-f07e96bf9f7a  HNSC-US  -0.5371036860447784   \n",
       "16  8c46a1df-1c10-4188-992a-a55b782494c9   LGG-US   0.0668132218607219   \n",
       "17  e2774050-1257-4c29-b201-d79ce70fe17e  SKCM-US  -0.4379410194660893   \n",
       "19  5310f698-d31a-4628-84b4-dc9ac29722a5  SKCM-US  -0.0068347114243561   \n",
       "20  d432e99a-67fb-4609-b90f-99438eee7cae  SKCM-US  -0.5151538970017837   \n",
       "22  978ae91e-6ebe-4efa-97ff-cfad511ae7b3  SKCM-US  -0.5839295199428811   \n",
       "24  d1132127-1250-43af-9c16-425798a3d1a7   GBM-US  -0.0624706219524289   \n",
       "26  2b142863-b963-4cc9-8f8f-c72503c93390  BLCA-US  -0.3533774556035514   \n",
       "27  d0e033bc-8e97-4322-b3b1-8626f1e2449b  SKCM-US  -0.0726002501514517   \n",
       "28  41d50a3c-49fc-4d8a-9c25-bbbcd3a5da6f   GBM-US  -0.2907228368770588   \n",
       "30  0c7aca3f-e006-4de3-afc2-20b4f727d4fd  BLCA-US  -0.2457822675716241   \n",
       "31  affa4ef4-02ec-40f1-b6a2-739f8bf2afa9  SKCM-US  -0.1417866820777039   \n",
       "33  10bb1a92-901e-4a14-80f4-5e88f997754b  SKCM-US  -0.4366975067097288   \n",
       "34  fc249113-83d4-4abe-8c80-a4f7305dcd91  SKCM-US  -0.4311616315076992   \n",
       "\n",
       "   passes_pipeline  \n",
       "0             True  \n",
       "1             True  \n",
       "2            False  \n",
       "3            False  \n",
       "4            False  \n",
       "6            False  \n",
       "7            False  \n",
       "8             True  \n",
       "9            False  \n",
       "11            True  \n",
       "12            True  \n",
       "14           False  \n",
       "15           False  \n",
       "16            True  \n",
       "17           False  \n",
       "19           False  \n",
       "20           False  \n",
       "22           False  \n",
       "24           False  \n",
       "26           False  \n",
       "27           False  \n",
       "28           False  \n",
       "30           False  \n",
       "31           False  \n",
       "33           False  \n",
       "34           False  "
      ]
     },
     "execution_count": 172,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tert_c250t_dataframe[tert_c250t_dataframe[\"zscore\"] != \"not_available\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Why did the RALY mutations drop?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 174,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>pid</th>\n",
       "      <th>cohort</th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SEQUENCE_CONTEXT</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>strand</th>\n",
       "      <th>...</th>\n",
       "      <th>destroyed_tfs_passing_tf_expression_threshold</th>\n",
       "      <th>num_remaining_tfs</th>\n",
       "      <th>remaining_tfs</th>\n",
       "      <th>JASPAR2020_CORE_vertebrates_non_redundant(tf_name,binding_affinity,seq1,seq2,raw,zscore,log,tf_sequence_logo)</th>\n",
       "      <th>paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))</th>\n",
       "      <th>num_original_mutations</th>\n",
       "      <th>num_final_mutations</th>\n",
       "      <th>num_promoter_mutations</th>\n",
       "      <th>num_recurrent_mutations</th>\n",
       "      <th>cn_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>70.38</td>\n",
       "      <td>d053f88f-09e6-472e-a691-bb293c461399</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>20</td>\n",
       "      <td>32580927</td>\n",
       "      <td>RALY</td>\n",
       "      <td>TGCCTCGTCT,TTCCGGATCC</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>ELF5,ZBTB7A</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ELK4,ELF2</td>\n",
       "      <td>ELK4,0.09259259259259259,TCTCTTCCGGA,.,5.67634...</td>\n",
       "      <td>/omics/groups/OE0436/internal/nabad/results/16...</td>\n",
       "      <td>19832</td>\n",
       "      <td>87</td>\n",
       "      <td>420</td>\n",
       "      <td>15.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>69.54</td>\n",
       "      <td>b07bad52-d44c-4b27-900a-960985bfadec</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>20</td>\n",
       "      <td>32580927</td>\n",
       "      <td>RALY</td>\n",
       "      <td>TGCCTCGTCT,TTCCGGATCC</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>ZBTB7A</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ELK4,ELF5,ELF2</td>\n",
       "      <td>ELK4,0.09259259259259259,TCTCTTCCGGA,.,3.81399...</td>\n",
       "      <td>/omics/groups/OE0436/internal/nabad/results/16...</td>\n",
       "      <td>158060</td>\n",
       "      <td>710</td>\n",
       "      <td>3344</td>\n",
       "      <td>15.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>60.91</td>\n",
       "      <td>3869ff3f-21b9-4817-8ff4-83c6fc75ab11</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>20</td>\n",
       "      <td>32580927</td>\n",
       "      <td>RALY</td>\n",
       "      <td>TGCCTCGTCT,TTCCGGATCC</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>ELF5,ZBTB7A</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ELK4,ELF2</td>\n",
       "      <td>ELK4,0.09259259259259259,TCTCTTCCGGA,.,4.02060...</td>\n",
       "      <td>/omics/groups/OE0436/internal/nabad/results/16...</td>\n",
       "      <td>171882</td>\n",
       "      <td>809</td>\n",
       "      <td>3472</td>\n",
       "      <td>15.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>58.51</td>\n",
       "      <td>e2774050-1257-4c29-b201-d79ce70fe17e</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>20</td>\n",
       "      <td>32580927</td>\n",
       "      <td>RALY</td>\n",
       "      <td>TGCCTCGTCT,TTCCGGATCC</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>ZBTB7A</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ELK4,ELF5,ELF2</td>\n",
       "      <td>ELK4,0.09259259259259259,TCTCTTCCGGA,.,6.14689...</td>\n",
       "      <td>/omics/groups/OE0436/internal/nabad/results/16...</td>\n",
       "      <td>164045</td>\n",
       "      <td>713</td>\n",
       "      <td>3175</td>\n",
       "      <td>15.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>882</th>\n",
       "      <td>52.91</td>\n",
       "      <td>3869ff3f-21b9-4817-8ff4-83c6fc75ab11</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>20</td>\n",
       "      <td>32580916</td>\n",
       "      <td>RALY</td>\n",
       "      <td>GAGTTTCTTC,TGCCTCGTCT</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>E2F6,ETV1,ETV4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>GABPA,ELF1</td>\n",
       "      <td>GABPA,0.09491538293611246,GTTTCTTCCTGCCT,.,4.9...</td>\n",
       "      <td>/omics/groups/OE0436/internal/nabad/results/16...</td>\n",
       "      <td>171882</td>\n",
       "      <td>809</td>\n",
       "      <td>3472</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161</th>\n",
       "      <td>51.08</td>\n",
       "      <td>760881cc-c623-11e3-bf01-24c6515278c0</td>\n",
       "      <td>LIRI-JP</td>\n",
       "      <td>20</td>\n",
       "      <td>32581904</td>\n",
       "      <td>RALY</td>\n",
       "      <td>CCTCCTTCCC,GCCGCGCGGC</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>RBPJ</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NRF1,ZNF263</td>\n",
       "      <td>E2F6,12.7642,.,GCGGCCGGGAAGG,4.39819574116,0.0...</td>\n",
       "      <td>.</td>\n",
       "      <td>184214</td>\n",
       "      <td>647</td>\n",
       "      <td>3397</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1508</th>\n",
       "      <td>49.47</td>\n",
       "      <td>0be08326-c623-11e3-bf01-24c6515278c0</td>\n",
       "      <td>LIRI-JP</td>\n",
       "      <td>20</td>\n",
       "      <td>32580986</td>\n",
       "      <td>RALY</td>\n",
       "      <td>CTCCAGGCCC,GCCCACCTCA</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>KLF4,EGR1,KLF5,KLF9,KLF10,KLF11,KLF15,KLF2,KLF...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>SP1,SP2,SP4,KLF16,SP3,SP8,ZNF740,MAZ,SP9</td>\n",
       "      <td>KLF4,0.06940058712896711,GGCCCCGCCCAC,.,4.6798...</td>\n",
       "      <td>.</td>\n",
       "      <td>6940</td>\n",
       "      <td>25</td>\n",
       "      <td>105</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows Ã— 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      score                                   pid   cohort #CHROM       POS  \\\n",
       "70    70.38  d053f88f-09e6-472e-a691-bb293c461399  SKCM-US     20  32580927   \n",
       "82    69.54  b07bad52-d44c-4b27-900a-960985bfadec  SKCM-US     20  32580927   \n",
       "287   60.91  3869ff3f-21b9-4817-8ff4-83c6fc75ab11  SKCM-US     20  32580927   \n",
       "409   58.51  e2774050-1257-4c29-b201-d79ce70fe17e  SKCM-US     20  32580927   \n",
       "882   52.91  3869ff3f-21b9-4817-8ff4-83c6fc75ab11  SKCM-US     20  32580916   \n",
       "1161  51.08  760881cc-c623-11e3-bf01-24c6515278c0  LIRI-JP     20  32581904   \n",
       "1508  49.47  0be08326-c623-11e3-bf01-24c6515278c0  LIRI-JP     20  32580986   \n",
       "\n",
       "      GENE       SEQUENCE_CONTEXT REF ALT strand  ...  \\\n",
       "70    RALY  TGCCTCGTCT,TTCCGGATCC   C   T      +  ...   \n",
       "82    RALY  TGCCTCGTCT,TTCCGGATCC   C   T      +  ...   \n",
       "287   RALY  TGCCTCGTCT,TTCCGGATCC   C   T      +  ...   \n",
       "409   RALY  TGCCTCGTCT,TTCCGGATCC   C   T      +  ...   \n",
       "882   RALY  GAGTTTCTTC,TGCCTCGTCT   C   T      +  ...   \n",
       "1161  RALY  CCTCCTTCCC,GCCGCGCGGC   A   G      +  ...   \n",
       "1508  RALY  CTCCAGGCCC,GCCCACCTCA   C   T      +  ...   \n",
       "\n",
       "          destroyed_tfs_passing_tf_expression_threshold  num_remaining_tfs  \\\n",
       "70                                          ELF5,ZBTB7A                2.0   \n",
       "82                                               ZBTB7A                3.0   \n",
       "287                                         ELF5,ZBTB7A                2.0   \n",
       "409                                              ZBTB7A                3.0   \n",
       "882                                      E2F6,ETV1,ETV4                2.0   \n",
       "1161                                               RBPJ                2.0   \n",
       "1508  KLF4,EGR1,KLF5,KLF9,KLF10,KLF11,KLF15,KLF2,KLF...                9.0   \n",
       "\n",
       "                                 remaining_tfs  \\\n",
       "70                                   ELK4,ELF2   \n",
       "82                              ELK4,ELF5,ELF2   \n",
       "287                                  ELK4,ELF2   \n",
       "409                             ELK4,ELF5,ELF2   \n",
       "882                                 GABPA,ELF1   \n",
       "1161                               NRF1,ZNF263   \n",
       "1508  SP1,SP2,SP4,KLF16,SP3,SP8,ZNF740,MAZ,SP9   \n",
       "\n",
       "      JASPAR2020_CORE_vertebrates_non_redundant(tf_name,binding_affinity,seq1,seq2,raw,zscore,log,tf_sequence_logo)  \\\n",
       "70    ELK4,0.09259259259259259,TCTCTTCCGGA,.,5.67634...                                                               \n",
       "82    ELK4,0.09259259259259259,TCTCTTCCGGA,.,3.81399...                                                               \n",
       "287   ELK4,0.09259259259259259,TCTCTTCCGGA,.,4.02060...                                                               \n",
       "409   ELK4,0.09259259259259259,TCTCTTCCGGA,.,6.14689...                                                               \n",
       "882   GABPA,0.09491538293611246,GTTTCTTCCTGCCT,.,4.9...                                                               \n",
       "1161  E2F6,12.7642,.,GCGGCCGGGAAGG,4.39819574116,0.0...                                                               \n",
       "1508  KLF4,0.06940058712896711,GGCCCCGCCCAC,.,4.6798...                                                               \n",
       "\n",
       "     paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))  \\\n",
       "70    /omics/groups/OE0436/internal/nabad/results/16...                                                                                        \n",
       "82    /omics/groups/OE0436/internal/nabad/results/16...                                                                                        \n",
       "287   /omics/groups/OE0436/internal/nabad/results/16...                                                                                        \n",
       "409   /omics/groups/OE0436/internal/nabad/results/16...                                                                                        \n",
       "882   /omics/groups/OE0436/internal/nabad/results/16...                                                                                        \n",
       "1161                                                  .                                                                                        \n",
       "1508                                                  .                                                                                        \n",
       "\n",
       "     num_original_mutations  num_final_mutations num_promoter_mutations  \\\n",
       "70                    19832                   87                    420   \n",
       "82                   158060                  710                   3344   \n",
       "287                  171882                  809                   3472   \n",
       "409                  164045                  713                   3175   \n",
       "882                  171882                  809                   3472   \n",
       "1161                 184214                  647                   3397   \n",
       "1508                   6940                   25                    105   \n",
       "\n",
       "      num_recurrent_mutations       cn_score  \n",
       "70                       15.0              3  \n",
       "82                       15.0  not_available  \n",
       "287                      15.0              4  \n",
       "409                      15.0              4  \n",
       "882                       1.0              4  \n",
       "1161                      0.0  not_available  \n",
       "1508                      0.0  not_available  \n",
       "\n",
       "[7 rows x 29 columns]"
      ]
     },
     "execution_count": 174,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_pcawg_and_eopc[combined_pcawg_and_eopc[\"GENE\"] == \"RALY\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2s/38kg81v526j7qcgd8gdq8jd00000gn/T/ipykernel_1243/908769902.py:14: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'not_available' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.\n",
      "  raly_dataframe.loc[idx+1, \"zscore\"] = zscore\n"
     ]
    }
   ],
   "source": [
    "raly = combined_pcawg_and_eopc.iloc[70]\n",
    "\n",
    "raly_dataframe = pd.DataFrame()\n",
    "raly_dataframe.loc[0, \"pid\"] = raly[\"pid\"]\n",
    "raly_dataframe.loc[0, \"cohort\"] = raly[\"cohort\"]\n",
    "raly_dataframe.loc[0, \"zscore\"] = raly[\"FPKM_Z_score\"]\n",
    "\n",
    "\n",
    "\n",
    "for idx, entry in enumerate(raly[\"paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))\"].split(\";\")):\n",
    "    path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af = entry.split(\",\")[:14]\n",
    "    raly_dataframe.loc[idx+1, \"pid\"] = pid\n",
    "    raly_dataframe.loc[idx+1, \"cohort\"] = cohort\n",
    "    raly_dataframe.loc[idx+1, \"zscore\"] = zscore\n",
    "    \n",
    "pids_of_raly_passing_pipeline = list(data[(data[\"POS\"] == 32580927) & (data[\"GENE\"] == \"RALY\")][\"pid\"].unique())\n",
    "\n",
    "for idx, row in raly_dataframe.iterrows():\n",
    "    raly_dataframe.loc[idx, \"passes_pipeline\"] = row[\"pid\"] in pids_of_tert_c250ts_passing_pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Top 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 180,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2s/38kg81v526j7qcgd8gdq8jd00000gn/T/ipykernel_1243/2506053702.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  combined_pcawg_and_eopc.iloc[:100][(combined_pcawg_and_eopc[\"GENE\"] == \"TERT\") & (combined_pcawg_and_eopc[\"POS\"] == 1295228)].shape\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(24, 29)"
      ]
     },
     "execution_count": 180,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_pcawg_and_eopc.iloc[:100][(combined_pcawg_and_eopc[\"GENE\"] == \"TERT\") & (combined_pcawg_and_eopc[\"POS\"] == 1295228)].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 181,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2s/38kg81v526j7qcgd8gdq8jd00000gn/T/ipykernel_1243/106539731.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  combined_pcawg_and_eopc.iloc[:100][(combined_pcawg_and_eopc[\"GENE\"] == \"TERT\") & (combined_pcawg_and_eopc[\"POS\"] == 1295250)].shape\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(6, 29)"
      ]
     },
     "execution_count": 181,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_pcawg_and_eopc.iloc[:100][(combined_pcawg_and_eopc[\"GENE\"] == \"TERT\") & (combined_pcawg_and_eopc[\"POS\"] == 1295250)].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 182,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2s/38kg81v526j7qcgd8gdq8jd00000gn/T/ipykernel_1243/925935169.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  combined_pcawg_and_eopc.iloc[:100][(combined_pcawg_and_eopc[\"GENE\"] == \"RALY\") & (combined_pcawg_and_eopc[\"POS\"] == 32580927)].shape\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(2, 29)"
      ]
     },
     "execution_count": 182,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_pcawg_and_eopc.iloc[:100][(combined_pcawg_and_eopc[\"GENE\"] == \"RALY\") & (combined_pcawg_and_eopc[\"POS\"] == 32580927)].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 189,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_100 = combined_pcawg_and_eopc.iloc[:100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 195,
   "metadata": {},
   "outputs": [],
   "source": [
    "count_top_100 = top_100.groupby([\"GENE\", \"POS\"]).agg({\"pid\": \"count\"}).sort_values(\"pid\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 197,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>pid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE</th>\n",
       "      <th>POS</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">TERT</th>\n",
       "      <th>1295228</th>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1295250</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ANKRD53</th>\n",
       "      <th>71204529</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SECISBP2</th>\n",
       "      <th>91933357</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RALY</th>\n",
       "      <th>32580927</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RSPO3</th>\n",
       "      <th>127439062</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RPL36AL</th>\n",
       "      <th>50087348</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RFX5</th>\n",
       "      <th>151320371</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PTPN9</th>\n",
       "      <th>75871490</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRKAR1A</th>\n",
       "      <th>66507534</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    pid\n",
       "GENE     POS           \n",
       "TERT     1295228     24\n",
       "         1295250      6\n",
       "ANKRD53  71204529     2\n",
       "SECISBP2 91933357     2\n",
       "RALY     32580927     2\n",
       "RSPO3    127439062    1\n",
       "RPL36AL  50087348     1\n",
       "RFX5     151320371    1\n",
       "PTPN9    75871490     1\n",
       "PRKAR1A  66507534     1"
      ]
     },
     "execution_count": 197,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_top_100.iloc[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "psnv-hunter",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
