{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Why did the TERT C228T mutations drop out?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>score</th>\n",
       "      <th>pid</th>\n",
       "      <th>cohort</th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>GENE</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>strand</th>\n",
       "      <th>...</th>\n",
       "      <th>num_promoter_mutations</th>\n",
       "      <th>num_recurrent_mutations</th>\n",
       "      <th>expression_traces</th>\n",
       "      <th>tf_summary</th>\n",
       "      <th>ncbi_gene_summary</th>\n",
       "      <th>ncbi_url</th>\n",
       "      <th>in_old_results</th>\n",
       "      <th>cn_score</th>\n",
       "      <th>cn_file_of_score</th>\n",
       "      <th>ICGC_Estimated_Purity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>93.45</td>\n",
       "      <td>45a7949d-e63f-4956-866c-df51257032de</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>2</td>\n",
       "      <td>71204529</td>\n",
       "      <td>ANKRD53</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>284</td>\n",
       "      <td>5.0</td>\n",
       "      <td>{'ANKRD53': {'BLCA-US': {'raw': [0.13674067793...</td>\n",
       "      <td>{'REL': {'summary': \"This gene encodes a prote...</td>\n",
       "      <td>Involved in mitotic metaphase plate congressio...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/79998</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>93.41</td>\n",
       "      <td>6c884037-49f9-41c3-b0e5-9cbcd545aeb7</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>9</td>\n",
       "      <td>91933357</td>\n",
       "      <td>SECISBP2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>879</td>\n",
       "      <td>6.0</td>\n",
       "      <td>{'SECISBP2': {'SKCM-US': {'raw': [5.0687620882...</td>\n",
       "      <td>{'ELK4': {'summary': 'This gene is a member of...</td>\n",
       "      <td>The protein encoded by this gene is one of the...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/79048</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>88.10</td>\n",
       "      <td>49dec0c2-8e75-4f44-a253-82b2ea605890</td>\n",
       "      <td>LUAD-US</td>\n",
       "      <td>2</td>\n",
       "      <td>71204529</td>\n",
       "      <td>ANKRD53</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>436</td>\n",
       "      <td>5.0</td>\n",
       "      <td>{'ANKRD53': {'LUAD-US': {'raw': [0.73980555761...</td>\n",
       "      <td>{'REL': {'summary': \"This gene encodes a prote...</td>\n",
       "      <td>Involved in mitotic metaphase plate congressio...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/79998</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>86.45</td>\n",
       "      <td>d692ecd0-2433-426b-9af2-a30c44a80f7c</td>\n",
       "      <td>THCA-US</td>\n",
       "      <td>5</td>\n",
       "      <td>1295228</td>\n",
       "      <td>TERT</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>95.0</td>\n",
       "      <td>{'TERT': {'THCA-US': {'raw': [0.0, 0.006731104...</td>\n",
       "      <td>{'ELK4': {'summary': 'This gene is a member of...</td>\n",
       "      <td>Telomerase is a ribonucleoprotein polymerase t...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/7015</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>86.30</td>\n",
       "      <td>9988eb07-01f6-4f83-8699-bb63e0525f08</td>\n",
       "      <td>HNSC-US</td>\n",
       "      <td>7</td>\n",
       "      <td>145813786</td>\n",
       "      <td>CNTNAP2</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>459</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'CNTNAP2': {'HNSC-US': {'raw': [0.38939713831...</td>\n",
       "      <td>{'ZEB1': {'summary': 'This gene encodes a zinc...</td>\n",
       "      <td>This gene encodes a member of the neurexin fam...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/26047</td>\n",
       "      <td>False</td>\n",
       "      <td>9</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.489</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 165 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0  score                                   pid   cohort  #CHROM  \\\n",
       "0           0  93.45  45a7949d-e63f-4956-866c-df51257032de  BLCA-US       2   \n",
       "1           1  93.41  6c884037-49f9-41c3-b0e5-9cbcd545aeb7  SKCM-US       9   \n",
       "2           2  88.10  49dec0c2-8e75-4f44-a253-82b2ea605890  LUAD-US       2   \n",
       "3           3  86.45  d692ecd0-2433-426b-9af2-a30c44a80f7c  THCA-US       5   \n",
       "4           4  86.30  9988eb07-01f6-4f83-8699-bb63e0525f08  HNSC-US       7   \n",
       "\n",
       "         POS      GENE REF ALT strand  ...  num_promoter_mutations  \\\n",
       "0   71204529   ANKRD53   G   A      +  ...                     284   \n",
       "1   91933357  SECISBP2   G   A      +  ...                     879   \n",
       "2   71204529   ANKRD53   G   A      +  ...                     436   \n",
       "3    1295228      TERT   G   A      -  ...                      20   \n",
       "4  145813786   CNTNAP2   C   A      +  ...                     459   \n",
       "\n",
       "   num_recurrent_mutations                                  expression_traces  \\\n",
       "0                      5.0  {'ANKRD53': {'BLCA-US': {'raw': [0.13674067793...   \n",
       "1                      6.0  {'SECISBP2': {'SKCM-US': {'raw': [5.0687620882...   \n",
       "2                      5.0  {'ANKRD53': {'LUAD-US': {'raw': [0.73980555761...   \n",
       "3                     95.0  {'TERT': {'THCA-US': {'raw': [0.0, 0.006731104...   \n",
       "4                      0.0  {'CNTNAP2': {'HNSC-US': {'raw': [0.38939713831...   \n",
       "\n",
       "                                          tf_summary  \\\n",
       "0  {'REL': {'summary': \"This gene encodes a prote...   \n",
       "1  {'ELK4': {'summary': 'This gene is a member of...   \n",
       "2  {'REL': {'summary': \"This gene encodes a prote...   \n",
       "3  {'ELK4': {'summary': 'This gene is a member of...   \n",
       "4  {'ZEB1': {'summary': 'This gene encodes a zinc...   \n",
       "\n",
       "                                   ncbi_gene_summary  \\\n",
       "0  Involved in mitotic metaphase plate congressio...   \n",
       "1  The protein encoded by this gene is one of the...   \n",
       "2  Involved in mitotic metaphase plate congressio...   \n",
       "3  Telomerase is a ribonucleoprotein polymerase t...   \n",
       "4  This gene encodes a member of the neurexin fam...   \n",
       "\n",
       "                                  ncbi_url in_old_results cn_score  \\\n",
       "0  https://www.ncbi.nlm.nih.gov/gene/79998           True        3   \n",
       "1  https://www.ncbi.nlm.nih.gov/gene/79048           True        6   \n",
       "2  https://www.ncbi.nlm.nih.gov/gene/79998           True        4   \n",
       "3   https://www.ncbi.nlm.nih.gov/gene/7015           True        3   \n",
       "4  https://www.ncbi.nlm.nih.gov/gene/26047          False        9   \n",
       "\n",
       "   cn_file_of_score ICGC_Estimated_Purity  \n",
       "0     not_available                 0.938  \n",
       "1     not_available                  0.78  \n",
       "2     not_available                   0.5  \n",
       "3     not_available                 0.589  \n",
       "4     not_available                 0.489  \n",
       "\n",
       "[5 rows x 165 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv(\"/Users/nicholasabad/Desktop/workspace/data/_DATA_FOR_PAPER_/FinalResults_long.csv\", delimiter=\"\\t\", nrows=100)\n",
    "\n",
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Unnamed: 0',\n",
       " 'score',\n",
       " 'pid',\n",
       " 'cohort',\n",
       " '#CHROM',\n",
       " 'POS',\n",
       " 'GENE',\n",
       " 'REF',\n",
       " 'ALT',\n",
       " 'strand',\n",
       " 'FPKM',\n",
       " 'FPKM_Z_score',\n",
       " 'num_paths_of_recurrence',\n",
       " 'within_cgc_list',\n",
       " 'within_chromhmm_promoter',\n",
       " 'allele_frequency',\n",
       " 'SuRE_score',\n",
       " 'SuRE_ID',\n",
       " 'num_created_tfs_passing_tf_expression_threshold',\n",
       " 'created_tfs_passing_tf_expression_threshold',\n",
       " 'num_destroyed_tfs_passing_tf_expression_threshold',\n",
       " 'destroyed_tfs_passing_tf_expression_threshold',\n",
       " 'num_remaining_tfs',\n",
       " 'remaining_tfs',\n",
       " 'ID',\n",
       " 'QUAL',\n",
       " 'FILTER',\n",
       " 'INFO',\n",
       " '#Chromosome',\n",
       " 'Start_position',\n",
       " 'Reference_Allele',\n",
       " 'Tumor_Seq_Allele2',\n",
       " 'Tumor_Sample_Barcode',\n",
       " 'Project_Code',\n",
       " 'Donor_ID',\n",
       " 'Strand',\n",
       " 'Variant_Classification',\n",
       " 'Variant_Type',\n",
       " 'Tumor_Seq_Allele1',\n",
       " 'dbSNP_RS',\n",
       " 'dbSNP_Val_Status',\n",
       " 'Matched_Norm_Sample_Barcode',\n",
       " 'Genome_Change',\n",
       " 'ref_context',\n",
       " 'gc_content',\n",
       " 'i_1000genomes_AF',\n",
       " 'i_1000genomes_ID',\n",
       " 'i_Callers',\n",
       " 'i_GERM1000G',\n",
       " 'i_GERMOVLP',\n",
       " 'i_LOWSUPPORT',\n",
       " 'i_NORMALPANEL',\n",
       " 'i_NumCallers',\n",
       " 'i_OXOGFAIL',\n",
       " 'i_REMAPFAIL',\n",
       " 'i_SEXF',\n",
       " 'i_VAF',\n",
       " 'i_bPcr',\n",
       " 'i_bSeq',\n",
       " 'i_qual',\n",
       " 'i_repeat_masker',\n",
       " 'i_signature_N3',\n",
       " 'i_signature_R1',\n",
       " 'i_signature_R2',\n",
       " 'i_snv_near_indel',\n",
       " 't_alt_count',\n",
       " 't_ref_count',\n",
       " 'i_model_score',\n",
       " 'i_n_vaf',\n",
       " 'enhancers',\n",
       " 'gc19_pc.3utr',\n",
       " 'gc19_pc.5utr',\n",
       " 'gc19_pc.cds',\n",
       " 'gc19_pc.promCore',\n",
       " 'gc19_pc.promDomain',\n",
       " 'gc19_pc.ss',\n",
       " 'lncrna.ncrna',\n",
       " 'lncrna.promCore',\n",
       " 'lncrna.promDomain',\n",
       " 'lncrna.ss',\n",
       " 'mirna.mat',\n",
       " 'mirna.pre',\n",
       " 'mirna.prom',\n",
       " 'smallrna.ncrna',\n",
       " 'roadmap_enhancer',\n",
       " 'roadmap_promoter',\n",
       " 'ucsc_cpgi',\n",
       " 'wgEncodeRegDnaseClusteredV3',\n",
       " 'MAPABILITY',\n",
       " 'HISEQDEPTH',\n",
       " 'SIMPLE_TANDEMREPEATS',\n",
       " 'REPEAT_MASKER',\n",
       " 'DUKE_EXCLUDED',\n",
       " 'DAC_BLACKLIST',\n",
       " 'SELFCHAIN',\n",
       " 'CpGislands',\n",
       " 'CgiMountains',\n",
       " 'Enhancers',\n",
       " 'miRNAs_snoRNAs',\n",
       " 'miRBase18',\n",
       " 'miRNAtargets',\n",
       " 'phastConsElem20bp',\n",
       " 'TFBScons',\n",
       " 'COSMIC',\n",
       " 'ENCODE_DNASE',\n",
       " 'ENCODE_TFBS',\n",
       " 'PANCAN_ENCODE_TFBS',\n",
       " 'PANCAN_NC_TRANSCRIPT',\n",
       " 'PANCAN_CODING_GENES_FLAT',\n",
       " 'PANCAN_FLAT_PROMOTERS',\n",
       " 'PANCAN_LINC_RNA_FLAT',\n",
       " 'PANCAN_CALLABLE_GENOME',\n",
       " 'PANCAN_MAPPABILITY',\n",
       " 'PANCAN_MIRBASE_20_MIRNA',\n",
       " 'PANCAN_REPLICATION_TIMING',\n",
       " 'PANCAN_ROADMAP_STR_DYADICS',\n",
       " 'PANCAN_ROADMAP_STR_ENHANCERS',\n",
       " 'PANCAN_ROADMAP_STR_PROMOTERS',\n",
       " 'PANCAN_ENCODE_TFB_SITES',\n",
       " 'DBSNP',\n",
       " '1K_GENOMES',\n",
       " 'ANNOVAR_FUNCTION',\n",
       " 'EXONIC_CLASSIFICATION',\n",
       " 'ANNOVAR_TRANSCRIPTS',\n",
       " 'SEGDUP',\n",
       " 'CYTOBAND',\n",
       " 'SEQUENCE_CONTEXT',\n",
       " 'JASPAR2020_CORE_vertebrates_non_redundant',\n",
       " 'within_pcawg_whitelist',\n",
       " 'purity_median_after_esd',\n",
       " 'purity_sd_after_esd',\n",
       " 'purity_median_before_esd',\n",
       " 'path_to_purity_file',\n",
       " 'JASPAR2020_CORE_vertebrates_non_redundant(tf_name,binding_affinity,seq1,seq2,raw,zscore,log)',\n",
       " 'promoter_start',\n",
       " 'promoter_end',\n",
       " 'distance_from_gene_start',\n",
       " 'distance_from_start_of_promoter',\n",
       " 'element',\n",
       " 'genome_region',\n",
       " 'within_basic_promoter',\n",
       " 'promoter_start.1',\n",
       " 'promoter_end.1',\n",
       " 'distance_from_gene_start.1',\n",
       " 'distance_from_start_of_promoter.1',\n",
       " 'element.1',\n",
       " 'genome_region.1',\n",
       " 'paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))',\n",
       " 'number_of_pcawg_recurrent_mutations',\n",
       " 'path_to_pcawg_recurrence_dict',\n",
       " 'JASPAR2020_CORE_vertebrates_non_redundant(tf_name,binding_affinity,seq1,seq2,raw,zscore,log,tf_sequence_logo)',\n",
       " 'transcription_factor_pfms',\n",
       " 'jaspar_matrix_ids',\n",
       " 'num_original_mutations',\n",
       " 'num_final_mutations',\n",
       " 'num_promoter_mutations',\n",
       " 'num_recurrent_mutations',\n",
       " 'expression_traces',\n",
       " 'tf_summary',\n",
       " 'ncbi_gene_summary',\n",
       " 'ncbi_url',\n",
       " 'in_old_results',\n",
       " 'cn_score',\n",
       " 'cn_file_of_score',\n",
       " 'ICGC_Estimated_Purity']"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(data.columns)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "tert_c228t = data.iloc[3]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2s/38kg81v526j7qcgd8gdq8jd00000gn/T/ipykernel_64597/437269650.py:12: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '-0.230602071237921' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.\n",
      "  tert_c228t_dataframe.loc[idx+1, \"zscore\"] = zscore\n"
     ]
    }
   ],
   "source": [
    "tert_c228t_dataframe = pd.DataFrame()\n",
    "tert_c228t_dataframe.loc[0, \"pid\"] = tert_c228t[\"pid\"]\n",
    "tert_c228t_dataframe.loc[0, \"cohort\"] = tert_c228t[\"cohort\"]\n",
    "tert_c228t_dataframe.loc[0, \"zscore\"] = tert_c228t[\"FPKM_Z_score\"]\n",
    "\n",
    "\n",
    "\n",
    "for idx, entry in enumerate(tert_c228t[\"paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))\"].split(\";\")):\n",
    "    path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af = entry.split(\",\")[:14]\n",
    "    tert_c228t_dataframe.loc[idx+1, \"pid\"] = pid\n",
    "    tert_c228t_dataframe.loc[idx+1, \"cohort\"] = cohort\n",
    "    tert_c228t_dataframe.loc[idx+1, \"zscore\"] = zscore\n",
    "    \n",
    "pids_of_tert_c228ts_passing_pipeline = list(data[(data[\"POS\"] == 1295228) & (data[\"GENE\"] == \"TERT\")][\"pid\"].unique())\n",
    "\n",
    "for idx, row in tert_c228t_dataframe.iterrows():\n",
    "    tert_c228t_dataframe.loc[idx, \"passes_pipeline\"] = row[\"pid\"] in pids_of_tert_c228ts_passing_pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "passes_pipeline\n",
       "False    73\n",
       "True     23\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tert_c228t_dataframe[\"passes_pipeline\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Are they the same?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2s/38kg81v526j7qcgd8gdq8jd00000gn/T/ipykernel_64597/2502772735.py:1: DtypeWarning: Columns (88) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  pcawg_results = pd.read_csv(\"/Users/nicholasabad/Desktop/workspace/data/_DATA_FOR_PAPER_/FinalResults_long.csv\", delimiter=\"\\t\")\n"
     ]
    }
   ],
   "source": [
    "pcawg_results = pd.read_csv(\"/Users/nicholasabad/Desktop/workspace/data/_DATA_FOR_PAPER_/FinalResults_long.csv\", delimiter=\"\\t\")\n",
    "pcawg_results.drop(\"Unnamed: 0\", axis=1, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>pid</th>\n",
       "      <th>cohort</th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>GENE</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>strand</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>...</th>\n",
       "      <th>num_promoter_mutations</th>\n",
       "      <th>num_recurrent_mutations</th>\n",
       "      <th>expression_traces</th>\n",
       "      <th>tf_summary</th>\n",
       "      <th>ncbi_gene_summary</th>\n",
       "      <th>ncbi_url</th>\n",
       "      <th>in_old_results</th>\n",
       "      <th>cn_score</th>\n",
       "      <th>cn_file_of_score</th>\n",
       "      <th>ICGC_Estimated_Purity</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>93.45</td>\n",
       "      <td>45a7949d-e63f-4956-866c-df51257032de</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>2</td>\n",
       "      <td>71204529</td>\n",
       "      <td>ANKRD53</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>1.428501</td>\n",
       "      <td>...</td>\n",
       "      <td>284</td>\n",
       "      <td>5.0</td>\n",
       "      <td>{'ANKRD53': {'BLCA-US': {'raw': [0.13674067793...</td>\n",
       "      <td>{'REL': {'summary': \"This gene encodes a prote...</td>\n",
       "      <td>Involved in mitotic metaphase plate congressio...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/79998</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.938</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>93.41</td>\n",
       "      <td>6c884037-49f9-41c3-b0e5-9cbcd545aeb7</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>9</td>\n",
       "      <td>91933357</td>\n",
       "      <td>SECISBP2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>15.079190</td>\n",
       "      <td>...</td>\n",
       "      <td>879</td>\n",
       "      <td>6.0</td>\n",
       "      <td>{'SECISBP2': {'SKCM-US': {'raw': [5.0687620882...</td>\n",
       "      <td>{'ELK4': {'summary': 'This gene is a member of...</td>\n",
       "      <td>The protein encoded by this gene is one of the...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/79048</td>\n",
       "      <td>True</td>\n",
       "      <td>6</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.78</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>88.10</td>\n",
       "      <td>49dec0c2-8e75-4f44-a253-82b2ea605890</td>\n",
       "      <td>LUAD-US</td>\n",
       "      <td>2</td>\n",
       "      <td>71204529</td>\n",
       "      <td>ANKRD53</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>1.764278</td>\n",
       "      <td>...</td>\n",
       "      <td>436</td>\n",
       "      <td>5.0</td>\n",
       "      <td>{'ANKRD53': {'LUAD-US': {'raw': [0.73980555761...</td>\n",
       "      <td>{'REL': {'summary': \"This gene encodes a prote...</td>\n",
       "      <td>Involved in mitotic metaphase plate congressio...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/79998</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>86.45</td>\n",
       "      <td>d692ecd0-2433-426b-9af2-a30c44a80f7c</td>\n",
       "      <td>THCA-US</td>\n",
       "      <td>5</td>\n",
       "      <td>1295228</td>\n",
       "      <td>TERT</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>0.380236</td>\n",
       "      <td>...</td>\n",
       "      <td>20</td>\n",
       "      <td>95.0</td>\n",
       "      <td>{'TERT': {'THCA-US': {'raw': [0.0, 0.006731104...</td>\n",
       "      <td>{'ELK4': {'summary': 'This gene is a member of...</td>\n",
       "      <td>Telomerase is a ribonucleoprotein polymerase t...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/7015</td>\n",
       "      <td>True</td>\n",
       "      <td>3</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.589</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>86.30</td>\n",
       "      <td>9988eb07-01f6-4f83-8699-bb63e0525f08</td>\n",
       "      <td>HNSC-US</td>\n",
       "      <td>7</td>\n",
       "      <td>145813786</td>\n",
       "      <td>CNTNAP2</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>21.329213</td>\n",
       "      <td>...</td>\n",
       "      <td>459</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'CNTNAP2': {'HNSC-US': {'raw': [0.38939713831...</td>\n",
       "      <td>{'ZEB1': {'summary': 'This gene encodes a zinc...</td>\n",
       "      <td>This gene encodes a member of the neurexin fam...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/26047</td>\n",
       "      <td>False</td>\n",
       "      <td>9</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18409</th>\n",
       "      <td>2.02</td>\n",
       "      <td>6ca5c1bb-275b-4d05-948a-3c6c7d03fab9</td>\n",
       "      <td>COAD-US</td>\n",
       "      <td>19</td>\n",
       "      <td>39615618</td>\n",
       "      <td>PAK4</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>14.560673</td>\n",
       "      <td>...</td>\n",
       "      <td>6298</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'PAK4': {'COAD-US': {'raw': [23.9501634252, 4...</td>\n",
       "      <td>{'ZNF384': {'summary': \"This gene encodes a C2...</td>\n",
       "      <td>Not Available</td>\n",
       "      <td>Not Available</td>\n",
       "      <td>False</td>\n",
       "      <td>not_available</td>\n",
       "      <td>not_available</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18410</th>\n",
       "      <td>2.02</td>\n",
       "      <td>978ae91e-6ebe-4efa-97ff-cfad511ae7b3</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>17</td>\n",
       "      <td>1645475</td>\n",
       "      <td>SERPINF2</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>1.000828</td>\n",
       "      <td>...</td>\n",
       "      <td>2157</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'SERPINF2': {'SKCM-US': {'raw': [0.1213962340...</td>\n",
       "      <td>{'MZF1': {'summary': 'Enables DNA-binding tran...</td>\n",
       "      <td>Not Available</td>\n",
       "      <td>Not Available</td>\n",
       "      <td>False</td>\n",
       "      <td>4</td>\n",
       "      <td>not_available</td>\n",
       "      <td>0.239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18411</th>\n",
       "      <td>2.01</td>\n",
       "      <td>a96eddfc-3afb-4bf8-a440-c91778113fbd</td>\n",
       "      <td>LUSC-US</td>\n",
       "      <td>2</td>\n",
       "      <td>155824463</td>\n",
       "      <td>AC092625.1</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>0.110200</td>\n",
       "      <td>...</td>\n",
       "      <td>611</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'AC092625.1': {'LUSC-US': {'raw': [0.11019988...</td>\n",
       "      <td>{'ZNF384': {'summary': \"This gene encodes a C2...</td>\n",
       "      <td>Not Available</td>\n",
       "      <td>Not Available</td>\n",
       "      <td>False</td>\n",
       "      <td>3</td>\n",
       "      <td>not_available</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18412</th>\n",
       "      <td>2.00</td>\n",
       "      <td>bb659e7c-0b97-41b5-ac10-a66bb1f6f283</td>\n",
       "      <td>RECA-EU</td>\n",
       "      <td>14</td>\n",
       "      <td>96829898</td>\n",
       "      <td>ATG2B</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>3.022303</td>\n",
       "      <td>...</td>\n",
       "      <td>102</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'ATG2B': {'RECA-EU': {'raw': [4.19346821296, ...</td>\n",
       "      <td>{'ZBTB14': {'summary': 'Enables DNA-binding tr...</td>\n",
       "      <td>Not Available</td>\n",
       "      <td>Not Available</td>\n",
       "      <td>False</td>\n",
       "      <td>not_available</td>\n",
       "      <td>not_available</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18413</th>\n",
       "      <td>2.00</td>\n",
       "      <td>14dd86c1-dfb3-4124-a453-4d64223bc048</td>\n",
       "      <td>KIRP-US</td>\n",
       "      <td>1</td>\n",
       "      <td>202317452</td>\n",
       "      <td>PPP1R12B</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>+</td>\n",
       "      <td>1.224841</td>\n",
       "      <td>...</td>\n",
       "      <td>46</td>\n",
       "      <td>0.0</td>\n",
       "      <td>{'PPP1R12B': {'KIRP-US': {'raw': [0.9522805347...</td>\n",
       "      <td>{'GMEB2': {'summary': 'This gene is a member o...</td>\n",
       "      <td>Myosin phosphatase is a protein complex compri...</td>\n",
       "      <td>https://www.ncbi.nlm.nih.gov/gene/4660</td>\n",
       "      <td>False</td>\n",
       "      <td>not_available</td>\n",
       "      <td>not_available</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18414 rows × 164 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       score                                   pid   cohort #CHROM        POS  \\\n",
       "0      93.45  45a7949d-e63f-4956-866c-df51257032de  BLCA-US      2   71204529   \n",
       "1      93.41  6c884037-49f9-41c3-b0e5-9cbcd545aeb7  SKCM-US      9   91933357   \n",
       "2      88.10  49dec0c2-8e75-4f44-a253-82b2ea605890  LUAD-US      2   71204529   \n",
       "3      86.45  d692ecd0-2433-426b-9af2-a30c44a80f7c  THCA-US      5    1295228   \n",
       "4      86.30  9988eb07-01f6-4f83-8699-bb63e0525f08  HNSC-US      7  145813786   \n",
       "...      ...                                   ...      ...    ...        ...   \n",
       "18409   2.02  6ca5c1bb-275b-4d05-948a-3c6c7d03fab9  COAD-US     19   39615618   \n",
       "18410   2.02  978ae91e-6ebe-4efa-97ff-cfad511ae7b3  SKCM-US     17    1645475   \n",
       "18411   2.01  a96eddfc-3afb-4bf8-a440-c91778113fbd  LUSC-US      2  155824463   \n",
       "18412   2.00  bb659e7c-0b97-41b5-ac10-a66bb1f6f283  RECA-EU     14   96829898   \n",
       "18413   2.00  14dd86c1-dfb3-4124-a453-4d64223bc048  KIRP-US      1  202317452   \n",
       "\n",
       "             GENE REF ALT strand       FPKM  ...  num_promoter_mutations  \\\n",
       "0         ANKRD53   G   A      +   1.428501  ...                     284   \n",
       "1        SECISBP2   G   A      +  15.079190  ...                     879   \n",
       "2         ANKRD53   G   A      +   1.764278  ...                     436   \n",
       "3            TERT   G   A      -   0.380236  ...                      20   \n",
       "4         CNTNAP2   C   A      +  21.329213  ...                     459   \n",
       "...           ...  ..  ..    ...        ...  ...                     ...   \n",
       "18409        PAK4   T   A      +  14.560673  ...                    6298   \n",
       "18410    SERPINF2   G   A      +   1.000828  ...                    2157   \n",
       "18411  AC092625.1   C   A      -   0.110200  ...                     611   \n",
       "18412       ATG2B   G   T      -   3.022303  ...                     102   \n",
       "18413    PPP1R12B   T   G      +   1.224841  ...                      46   \n",
       "\n",
       "       num_recurrent_mutations  \\\n",
       "0                          5.0   \n",
       "1                          6.0   \n",
       "2                          5.0   \n",
       "3                         95.0   \n",
       "4                          0.0   \n",
       "...                        ...   \n",
       "18409                      0.0   \n",
       "18410                      0.0   \n",
       "18411                      0.0   \n",
       "18412                      0.0   \n",
       "18413                      0.0   \n",
       "\n",
       "                                       expression_traces  \\\n",
       "0      {'ANKRD53': {'BLCA-US': {'raw': [0.13674067793...   \n",
       "1      {'SECISBP2': {'SKCM-US': {'raw': [5.0687620882...   \n",
       "2      {'ANKRD53': {'LUAD-US': {'raw': [0.73980555761...   \n",
       "3      {'TERT': {'THCA-US': {'raw': [0.0, 0.006731104...   \n",
       "4      {'CNTNAP2': {'HNSC-US': {'raw': [0.38939713831...   \n",
       "...                                                  ...   \n",
       "18409  {'PAK4': {'COAD-US': {'raw': [23.9501634252, 4...   \n",
       "18410  {'SERPINF2': {'SKCM-US': {'raw': [0.1213962340...   \n",
       "18411  {'AC092625.1': {'LUSC-US': {'raw': [0.11019988...   \n",
       "18412  {'ATG2B': {'RECA-EU': {'raw': [4.19346821296, ...   \n",
       "18413  {'PPP1R12B': {'KIRP-US': {'raw': [0.9522805347...   \n",
       "\n",
       "                                              tf_summary  \\\n",
       "0      {'REL': {'summary': \"This gene encodes a prote...   \n",
       "1      {'ELK4': {'summary': 'This gene is a member of...   \n",
       "2      {'REL': {'summary': \"This gene encodes a prote...   \n",
       "3      {'ELK4': {'summary': 'This gene is a member of...   \n",
       "4      {'ZEB1': {'summary': 'This gene encodes a zinc...   \n",
       "...                                                  ...   \n",
       "18409  {'ZNF384': {'summary': \"This gene encodes a C2...   \n",
       "18410  {'MZF1': {'summary': 'Enables DNA-binding tran...   \n",
       "18411  {'ZNF384': {'summary': \"This gene encodes a C2...   \n",
       "18412  {'ZBTB14': {'summary': 'Enables DNA-binding tr...   \n",
       "18413  {'GMEB2': {'summary': 'This gene is a member o...   \n",
       "\n",
       "                                       ncbi_gene_summary  \\\n",
       "0      Involved in mitotic metaphase plate congressio...   \n",
       "1      The protein encoded by this gene is one of the...   \n",
       "2      Involved in mitotic metaphase plate congressio...   \n",
       "3      Telomerase is a ribonucleoprotein polymerase t...   \n",
       "4      This gene encodes a member of the neurexin fam...   \n",
       "...                                                  ...   \n",
       "18409                                      Not Available   \n",
       "18410                                      Not Available   \n",
       "18411                                      Not Available   \n",
       "18412                                      Not Available   \n",
       "18413  Myosin phosphatase is a protein complex compri...   \n",
       "\n",
       "                                      ncbi_url in_old_results       cn_score  \\\n",
       "0      https://www.ncbi.nlm.nih.gov/gene/79998           True              3   \n",
       "1      https://www.ncbi.nlm.nih.gov/gene/79048           True              6   \n",
       "2      https://www.ncbi.nlm.nih.gov/gene/79998           True              4   \n",
       "3       https://www.ncbi.nlm.nih.gov/gene/7015           True              3   \n",
       "4      https://www.ncbi.nlm.nih.gov/gene/26047          False              9   \n",
       "...                                        ...            ...            ...   \n",
       "18409                            Not Available          False  not_available   \n",
       "18410                            Not Available          False              4   \n",
       "18411                            Not Available          False              3   \n",
       "18412                            Not Available          False  not_available   \n",
       "18413   https://www.ncbi.nlm.nih.gov/gene/4660          False  not_available   \n",
       "\n",
       "      cn_file_of_score  ICGC_Estimated_Purity  \n",
       "0        not_available                  0.938  \n",
       "1        not_available                   0.78  \n",
       "2        not_available                    0.5  \n",
       "3        not_available                  0.589  \n",
       "4        not_available                  0.489  \n",
       "...                ...                    ...  \n",
       "18409    not_available          not_available  \n",
       "18410    not_available                  0.239  \n",
       "18411    not_available          not_available  \n",
       "18412    not_available          not_available  \n",
       "18413    not_available          not_available  \n",
       "\n",
       "[18414 rows x 164 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pcawg_results = pcawg_results[(pcawg_results[\"REF\"].str.len() == 1) & (pcawg_results[\"ALT\"].str.len() == 1)].reset_index(drop=True)\n",
    "pcawg_results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "pcawg_results.rename(\n",
    "    {\n",
    "        \"within_chromhmm_promoter\": \"open_chromatin\"\n",
    "    },\n",
    "    axis=1,\n",
    "    inplace=True\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Combine subset_long with EOPC-DE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "eopc_de = pd.read_csv(\"/Users/nicholasabad/Desktop/workspace/data/_DATA_FOR_PAPER_/SupplementaryTables/2_eopc_de_remaining_mutations_28Feb2024.csv\")\n",
    "eopc_de = eopc_de[(eopc_de[\"REF\"].str.len() == 1) & (eopc_de[\"ALT\"].str.len() == 1)].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# NOTE: All columns were taken from EOPC-DE dataframe except for cn_score.\n",
    "\n",
    "necessary_columns = [\n",
    "    'score',\n",
    "    'pid',\n",
    "    'cohort',\n",
    "    '#CHROM',\n",
    "    'POS',\n",
    "    'GENE',\n",
    "    'SEQUENCE_CONTEXT',\n",
    "    'REF',\n",
    "    'ALT',\n",
    "    'strand',\n",
    "    'FPKM',\n",
    "    'FPKM_Z_score',\n",
    "    'within_cgc_list',\n",
    "    'open_chromatin',\n",
    "    'allele_frequency',\n",
    "    'ICGC_Estimated_Purity',\n",
    "    'num_created_tfs_passing_tf_expression_threshold',\n",
    "    'created_tfs_passing_tf_expression_threshold',\n",
    "    'num_destroyed_tfs_passing_tf_expression_threshold',\n",
    "    'destroyed_tfs_passing_tf_expression_threshold',\n",
    "    'num_remaining_tfs',\n",
    "    'remaining_tfs',\n",
    "    'JASPAR2020_CORE_vertebrates_non_redundant(tf_name,binding_affinity,seq1,seq2,raw,zscore,log,tf_sequence_logo)',\n",
    "    'paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))',\n",
    "    'num_original_mutations',\n",
    "    'num_final_mutations',\n",
    "    'num_promoter_mutations',\n",
    "    'num_recurrent_mutations',\n",
    "    'cn_score'\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_pcawg_and_eopc = pd.concat([pcawg_results, eopc_de], ignore_index=True)\n",
    "combined_pcawg_and_eopc.sort_values(\"score\", ascending=False, inplace=True)\n",
    "\n",
    "combined_pcawg_and_eopc = combined_pcawg_and_eopc[necessary_columns]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "combined_pcawg_and_eopc.to_csv(\n",
    "    \"/Users/nicholasabad/Desktop/workspace/data/_DATA_FOR_PAPER_/pcawg_with_eopc_de_all_mutations_18March2025.csv\",\n",
    "    sep=\",\", index=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>pid</th>\n",
       "      <th>cohort</th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SEQUENCE_CONTEXT</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>strand</th>\n",
       "      <th>...</th>\n",
       "      <th>destroyed_tfs_passing_tf_expression_threshold</th>\n",
       "      <th>num_remaining_tfs</th>\n",
       "      <th>remaining_tfs</th>\n",
       "      <th>JASPAR2020_CORE_vertebrates_non_redundant(tf_name,binding_affinity,seq1,seq2,raw,zscore,log,tf_sequence_logo)</th>\n",
       "      <th>paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))</th>\n",
       "      <th>num_original_mutations</th>\n",
       "      <th>num_final_mutations</th>\n",
       "      <th>num_promoter_mutations</th>\n",
       "      <th>num_recurrent_mutations</th>\n",
       "      <th>cn_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>93.45</td>\n",
       "      <td>45a7949d-e63f-4956-866c-df51257032de</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>2</td>\n",
       "      <td>71204529</td>\n",
       "      <td>ANKRD53</td>\n",
       "      <td>GGGAAAGGGA,ACTCCCTTTC</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NFKB1,NFKB2</td>\n",
       "      <td>REL,13.2727,.,GGGAGTTTCC,0.298502304678,-1.254...</td>\n",
       "      <td>/omics/groups/OE0436/internal/nabad/results/16...</td>\n",
       "      <td>11050</td>\n",
       "      <td>76</td>\n",
       "      <td>284</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>93.41</td>\n",
       "      <td>6c884037-49f9-41c3-b0e5-9cbcd545aeb7</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>9</td>\n",
       "      <td>91933357</td>\n",
       "      <td>SECISBP2</td>\n",
       "      <td>CCGGCCGGAA,AAAGGGCGGG</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>ELK4,ZBTB7A</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ELK1,ELK3</td>\n",
       "      <td>ELK1,0.10067117472187072,GCCGGAAGAA,.,14.99847...</td>\n",
       "      <td>/omics/groups/OE0436/internal/nabad/results/16...</td>\n",
       "      <td>40712</td>\n",
       "      <td>223</td>\n",
       "      <td>879</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>88.10</td>\n",
       "      <td>49dec0c2-8e75-4f44-a253-82b2ea605890</td>\n",
       "      <td>LUAD-US</td>\n",
       "      <td>2</td>\n",
       "      <td>71204529</td>\n",
       "      <td>ANKRD53</td>\n",
       "      <td>GGGAAAGGGA,ACTCCCTTTC</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NFKB1,NFKB2</td>\n",
       "      <td>REL,13.2727,.,GGGAGTTTCC,1.13424410656,-0.0631...</td>\n",
       "      <td>/omics/groups/OE0436/internal/nabad/results/16...</td>\n",
       "      <td>22203</td>\n",
       "      <td>91</td>\n",
       "      <td>436</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>86.45</td>\n",
       "      <td>d692ecd0-2433-426b-9af2-a30c44a80f7c</td>\n",
       "      <td>THCA-US</td>\n",
       "      <td>5</td>\n",
       "      <td>1295228</td>\n",
       "      <td>TERT</td>\n",
       "      <td>AGGGCCCGGA,GGGGCTGGGC</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>SP8,TFAP2A,TFAP2C,CTCFL,MAZ,ZNF148</td>\n",
       "      <td>9.0</td>\n",
       "      <td>TFAP2A,ELF1,ELF4,ETV6,KLF16,ZBTB7A,TFAP2B,TFAP...</td>\n",
       "      <td>TFAP2A,0.09557122923715045,GCCCCCTCCGGGCC,.,0....</td>\n",
       "      <td>/omics/groups/OE0436/internal/nabad/results/16...</td>\n",
       "      <td>1099</td>\n",
       "      <td>5</td>\n",
       "      <td>20</td>\n",
       "      <td>95.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>86.30</td>\n",
       "      <td>9988eb07-01f6-4f83-8699-bb63e0525f08</td>\n",
       "      <td>HNSC-US</td>\n",
       "      <td>7</td>\n",
       "      <td>145813786</td>\n",
       "      <td>CNTNAP2</td>\n",
       "      <td>GCTGAGACTC,CCTGCCGCTC</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>ZNF148</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NHLH2,SNAI3</td>\n",
       "      <td>ZEB1,14.5102,.,CTCACCTGCCG,0.832158029939,-0.7...</td>\n",
       "      <td>.</td>\n",
       "      <td>23779</td>\n",
       "      <td>92</td>\n",
       "      <td>459</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18410</th>\n",
       "      <td>2.02</td>\n",
       "      <td>978ae91e-6ebe-4efa-97ff-cfad511ae7b3</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>17</td>\n",
       "      <td>1645475</td>\n",
       "      <td>SERPINF2</td>\n",
       "      <td>TCTTGGGTAG,GGAAAGAGAC</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>MZF1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>MAZ</td>\n",
       "      <td>MZF1,0.08042270172024159,GTAGGGGAAA,.,1.908352...</td>\n",
       "      <td>.</td>\n",
       "      <td>110086</td>\n",
       "      <td>472</td>\n",
       "      <td>2157</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18409</th>\n",
       "      <td>2.02</td>\n",
       "      <td>6ca5c1bb-275b-4d05-948a-3c6c7d03fab9</td>\n",
       "      <td>COAD-US</td>\n",
       "      <td>19</td>\n",
       "      <td>39615618</td>\n",
       "      <td>PAK4</td>\n",
       "      <td>ATACAAAAAA,TAGCTGGGTG</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZNF384,11.5106,.,TACAAAAAAATA,7.03125452052,-1...</td>\n",
       "      <td>.</td>\n",
       "      <td>413781</td>\n",
       "      <td>1320</td>\n",
       "      <td>6298</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18411</th>\n",
       "      <td>2.01</td>\n",
       "      <td>a96eddfc-3afb-4bf8-a440-c91778113fbd</td>\n",
       "      <td>LUSC-US</td>\n",
       "      <td>2</td>\n",
       "      <td>155824463</td>\n",
       "      <td>AC092625.1</td>\n",
       "      <td>GTTGAAATTA,AAAAATGTTG</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZNF384,11.7021,.,AATTAAAAAAAT,14.9732578533,0....</td>\n",
       "      <td>.</td>\n",
       "      <td>30011</td>\n",
       "      <td>145</td>\n",
       "      <td>611</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18412</th>\n",
       "      <td>2.00</td>\n",
       "      <td>bb659e7c-0b97-41b5-ac10-a66bb1f6f283</td>\n",
       "      <td>RECA-EU</td>\n",
       "      <td>14</td>\n",
       "      <td>96829898</td>\n",
       "      <td>ATG2B</td>\n",
       "      <td>AGGACGCCGG,CGCGCAGGTG</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>ZBTB14</td>\n",
       "      <td>2.0</td>\n",
       "      <td>PAX5,KLF15</td>\n",
       "      <td>PAX5,10.5091,.,CTGCGCGACCGG,0.00464238213385,-...</td>\n",
       "      <td>.</td>\n",
       "      <td>4949</td>\n",
       "      <td>16</td>\n",
       "      <td>102</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18413</th>\n",
       "      <td>2.00</td>\n",
       "      <td>14dd86c1-dfb3-4124-a453-4d64223bc048</td>\n",
       "      <td>KIRP-US</td>\n",
       "      <td>1</td>\n",
       "      <td>202317452</td>\n",
       "      <td>PPP1R12B</td>\n",
       "      <td>GTTTAGGTAC,TAATGTTTAG</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GMEB2,11.5152,.,TTACGTAC,2.48928711784,-1.6381...</td>\n",
       "      <td>.</td>\n",
       "      <td>2737</td>\n",
       "      <td>13</td>\n",
       "      <td>46</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18468 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       score                                   pid   cohort #CHROM        POS  \\\n",
       "0      93.45  45a7949d-e63f-4956-866c-df51257032de  BLCA-US      2   71204529   \n",
       "1      93.41  6c884037-49f9-41c3-b0e5-9cbcd545aeb7  SKCM-US      9   91933357   \n",
       "2      88.10  49dec0c2-8e75-4f44-a253-82b2ea605890  LUAD-US      2   71204529   \n",
       "3      86.45  d692ecd0-2433-426b-9af2-a30c44a80f7c  THCA-US      5    1295228   \n",
       "4      86.30  9988eb07-01f6-4f83-8699-bb63e0525f08  HNSC-US      7  145813786   \n",
       "...      ...                                   ...      ...    ...        ...   \n",
       "18410   2.02  978ae91e-6ebe-4efa-97ff-cfad511ae7b3  SKCM-US     17    1645475   \n",
       "18409   2.02  6ca5c1bb-275b-4d05-948a-3c6c7d03fab9  COAD-US     19   39615618   \n",
       "18411   2.01  a96eddfc-3afb-4bf8-a440-c91778113fbd  LUSC-US      2  155824463   \n",
       "18412   2.00  bb659e7c-0b97-41b5-ac10-a66bb1f6f283  RECA-EU     14   96829898   \n",
       "18413   2.00  14dd86c1-dfb3-4124-a453-4d64223bc048  KIRP-US      1  202317452   \n",
       "\n",
       "             GENE       SEQUENCE_CONTEXT REF ALT strand  ...  \\\n",
       "0         ANKRD53  GGGAAAGGGA,ACTCCCTTTC   G   A      +  ...   \n",
       "1        SECISBP2  CCGGCCGGAA,AAAGGGCGGG   G   A      +  ...   \n",
       "2         ANKRD53  GGGAAAGGGA,ACTCCCTTTC   G   A      +  ...   \n",
       "3            TERT  AGGGCCCGGA,GGGGCTGGGC   G   A      -  ...   \n",
       "4         CNTNAP2  GCTGAGACTC,CCTGCCGCTC   C   A      +  ...   \n",
       "...           ...                    ...  ..  ..    ...  ...   \n",
       "18410    SERPINF2  TCTTGGGTAG,GGAAAGAGAC   G   A      +  ...   \n",
       "18409        PAK4  ATACAAAAAA,TAGCTGGGTG   T   A      +  ...   \n",
       "18411  AC092625.1  GTTGAAATTA,AAAAATGTTG   C   A      -  ...   \n",
       "18412       ATG2B  AGGACGCCGG,CGCGCAGGTG   G   T      -  ...   \n",
       "18413    PPP1R12B  GTTTAGGTAC,TAATGTTTAG   T   G      +  ...   \n",
       "\n",
       "       destroyed_tfs_passing_tf_expression_threshold  num_remaining_tfs  \\\n",
       "0                                                NaN                2.0   \n",
       "1                                        ELK4,ZBTB7A                2.0   \n",
       "2                                                NaN                2.0   \n",
       "3                 SP8,TFAP2A,TFAP2C,CTCFL,MAZ,ZNF148                9.0   \n",
       "4                                             ZNF148                2.0   \n",
       "...                                              ...                ...   \n",
       "18410                                           MZF1                1.0   \n",
       "18409                                            NaN                0.0   \n",
       "18411                                            NaN                0.0   \n",
       "18412                                         ZBTB14                2.0   \n",
       "18413                                            NaN                0.0   \n",
       "\n",
       "                                           remaining_tfs  \\\n",
       "0                                            NFKB1,NFKB2   \n",
       "1                                              ELK1,ELK3   \n",
       "2                                            NFKB1,NFKB2   \n",
       "3      TFAP2A,ELF1,ELF4,ETV6,KLF16,ZBTB7A,TFAP2B,TFAP...   \n",
       "4                                            NHLH2,SNAI3   \n",
       "...                                                  ...   \n",
       "18410                                                MAZ   \n",
       "18409                                                NaN   \n",
       "18411                                                NaN   \n",
       "18412                                         PAX5,KLF15   \n",
       "18413                                                NaN   \n",
       "\n",
       "       JASPAR2020_CORE_vertebrates_non_redundant(tf_name,binding_affinity,seq1,seq2,raw,zscore,log,tf_sequence_logo)  \\\n",
       "0      REL,13.2727,.,GGGAGTTTCC,0.298502304678,-1.254...                                                               \n",
       "1      ELK1,0.10067117472187072,GCCGGAAGAA,.,14.99847...                                                               \n",
       "2      REL,13.2727,.,GGGAGTTTCC,1.13424410656,-0.0631...                                                               \n",
       "3      TFAP2A,0.09557122923715045,GCCCCCTCCGGGCC,.,0....                                                               \n",
       "4      ZEB1,14.5102,.,CTCACCTGCCG,0.832158029939,-0.7...                                                               \n",
       "...                                                  ...                                                               \n",
       "18410  MZF1,0.08042270172024159,GTAGGGGAAA,.,1.908352...                                                               \n",
       "18409  ZNF384,11.5106,.,TACAAAAAAATA,7.03125452052,-1...                                                               \n",
       "18411  ZNF384,11.7021,.,AATTAAAAAAAT,14.9732578533,0....                                                               \n",
       "18412  PAX5,10.5091,.,CTGCGCGACCGG,0.00464238213385,-...                                                               \n",
       "18413  GMEB2,11.5152,.,TTACGTAC,2.48928711784,-1.6381...                                                               \n",
       "\n",
       "      paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))  \\\n",
       "0      /omics/groups/OE0436/internal/nabad/results/16...                                                                                        \n",
       "1      /omics/groups/OE0436/internal/nabad/results/16...                                                                                        \n",
       "2      /omics/groups/OE0436/internal/nabad/results/16...                                                                                        \n",
       "3      /omics/groups/OE0436/internal/nabad/results/16...                                                                                        \n",
       "4                                                      .                                                                                        \n",
       "...                                                  ...                                                                                        \n",
       "18410                                                  .                                                                                        \n",
       "18409                                                  .                                                                                        \n",
       "18411                                                  .                                                                                        \n",
       "18412                                                  .                                                                                        \n",
       "18413                                                  .                                                                                        \n",
       "\n",
       "      num_original_mutations  num_final_mutations num_promoter_mutations  \\\n",
       "0                      11050                   76                    284   \n",
       "1                      40712                  223                    879   \n",
       "2                      22203                   91                    436   \n",
       "3                       1099                    5                     20   \n",
       "4                      23779                   92                    459   \n",
       "...                      ...                  ...                    ...   \n",
       "18410                 110086                  472                   2157   \n",
       "18409                 413781                 1320                   6298   \n",
       "18411                  30011                  145                    611   \n",
       "18412                   4949                   16                    102   \n",
       "18413                   2737                   13                     46   \n",
       "\n",
       "       num_recurrent_mutations       cn_score  \n",
       "0                          5.0              3  \n",
       "1                          6.0              6  \n",
       "2                          5.0              4  \n",
       "3                         95.0              3  \n",
       "4                          0.0              9  \n",
       "...                        ...            ...  \n",
       "18410                      0.0              4  \n",
       "18409                      0.0  not_available  \n",
       "18411                      0.0              3  \n",
       "18412                      0.0  not_available  \n",
       "18413                      0.0  not_available  \n",
       "\n",
       "[18468 rows x 29 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_pcawg_and_eopc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Why did the TERT C228T mutations drop out?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2s/38kg81v526j7qcgd8gdq8jd00000gn/T/ipykernel_64597/1956300189.py:13: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '-0.230602071237921' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.\n",
      "  tert_c228t_dataframe.loc[idx+1, \"zscore\"] = zscore\n"
     ]
    }
   ],
   "source": [
    "tert_c228t = combined_pcawg_and_eopc.iloc[3]\n",
    "\n",
    "tert_c228t_dataframe = pd.DataFrame()\n",
    "tert_c228t_dataframe.loc[0, \"pid\"] = tert_c228t[\"pid\"]\n",
    "tert_c228t_dataframe.loc[0, \"cohort\"] = tert_c228t[\"cohort\"]\n",
    "tert_c228t_dataframe.loc[0, \"zscore\"] = tert_c228t[\"FPKM_Z_score\"]\n",
    "\n",
    "\n",
    "for idx, entry in enumerate(tert_c228t[\"paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))\"].split(\";\")):\n",
    "    path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af = entry.split(\",\")[:14]\n",
    "    tert_c228t_dataframe.loc[idx+1, \"pid\"] = pid\n",
    "    tert_c228t_dataframe.loc[idx+1, \"cohort\"] = cohort\n",
    "    tert_c228t_dataframe.loc[idx+1, \"zscore\"] = zscore\n",
    "    \n",
    "pids_of_tert_c228ts_passing_pipeline = list(data[(data[\"POS\"] == 1295228) & (data[\"GENE\"] == \"TERT\")][\"pid\"].unique())\n",
    "\n",
    "for idx, row in tert_c228t_dataframe.iterrows():\n",
    "    tert_c228t_dataframe.loc[idx, \"passes_pipeline\"] = row[\"pid\"] in pids_of_tert_c228ts_passing_pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(96, 4)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Result: 96 possible TERT C228T mutations\n",
    "tert_c228t_dataframe.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(65, 4)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Result: 65 TERT C228Ts have RNA-Seq data available.\n",
    "tert_c228t_dataframe[tert_c228t_dataframe[\"zscore\"] != \"not_available\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(28, 29)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Result: 28 passed all filter steps.\n",
    "combined_pcawg_and_eopc[(combined_pcawg_and_eopc[\"GENE\"] == \"TERT\") & (combined_pcawg_and_eopc[\"POS\"] == 1295228)].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2s/38kg81v526j7qcgd8gdq8jd00000gn/T/ipykernel_64597/971158515.py:2: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  combined_pcawg_and_eopc.iloc[:100][(combined_pcawg_and_eopc[\"GENE\"] == \"TERT\") & (combined_pcawg_and_eopc[\"POS\"] == 1295228)].shape\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(24, 29)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Result: 24 were within the top 100 candidates\n",
    "combined_pcawg_and_eopc.iloc[:100][(combined_pcawg_and_eopc[\"GENE\"] == \"TERT\") & (combined_pcawg_and_eopc[\"POS\"] == 1295228)].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2s/38kg81v526j7qcgd8gdq8jd00000gn/T/ipykernel_64597/2686636836.py:3: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  tert_c228t_dataframe_ge_data_available[\"zscore\"] = tert_c228t_dataframe_ge_data_available[\"zscore\"].astype(float)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(37, 4)"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Result: 37 of the TERT C228Ts with GE data available did not pass due to having a zscore < 0.\n",
    "tert_c228t_dataframe_ge_data_available = tert_c228t_dataframe[(tert_c228t_dataframe[\"zscore\"] != \"not_available\")]\n",
    "tert_c228t_dataframe_ge_data_available[\"zscore\"] = tert_c228t_dataframe_ge_data_available[\"zscore\"].astype(float)\n",
    "tert_c228t_dataframe_ge_data_available[tert_c228t_dataframe_ge_data_available[\"zscore\"] < 0].shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Why did C250T mutations drop?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2s/38kg81v526j7qcgd8gdq8jd00000gn/T/ipykernel_64597/1779089474.py:15: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value '0.2882712716542805' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.\n",
      "  tert_c250t_dataframe.loc[idx+1, \"zscore\"] = zscore\n"
     ]
    }
   ],
   "source": [
    "tert_c250t = combined_pcawg_and_eopc.iloc[29]\n",
    "\n",
    "tert_c250t_dataframe = pd.DataFrame()\n",
    "tert_c250t_dataframe.loc[0, \"pid\"] = tert_c250t[\"pid\"]\n",
    "tert_c250t_dataframe.loc[0, \"cohort\"] = tert_c250t[\"cohort\"]\n",
    "tert_c250t_dataframe.loc[0, \"zscore\"] = tert_c250t[\"FPKM_Z_score\"]\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "for idx, entry in enumerate(tert_c250t[\"paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))\"].split(\";\")):\n",
    "    path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af = entry.split(\",\")[:14]\n",
    "    tert_c250t_dataframe.loc[idx+1, \"pid\"] = pid\n",
    "    tert_c250t_dataframe.loc[idx+1, \"cohort\"] = cohort\n",
    "    tert_c250t_dataframe.loc[idx+1, \"zscore\"] = zscore\n",
    "    \n",
    "pids_of_tert_c250ts_passing_pipeline = list(data[(data[\"POS\"] == 1295250) & (data[\"GENE\"] == \"TERT\")][\"pid\"].unique())\n",
    "\n",
    "for idx, row in tert_c250t_dataframe.iterrows():\n",
    "    tert_c250t_dataframe.loc[idx, \"passes_pipeline\"] = row[\"pid\"] in pids_of_tert_c250ts_passing_pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(35, 4)"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Result: 35 total TERT C250Ts\n",
    "tert_c250t_dataframe.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(26, 4)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Result: 26 have RNA-Seq data available\n",
    "tert_c250t_dataframe[tert_c250t_dataframe[\"zscore\"] != \"not_available\"].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "passes_pipeline\n",
       "False    20\n",
       "True      6\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Result: 6 passed the pipeline\n",
    "tert_c250t_dataframe[tert_c250t_dataframe[\"zscore\"] != \"not_available\"][\"passes_pipeline\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pid</th>\n",
       "      <th>cohort</th>\n",
       "      <th>zscore</th>\n",
       "      <th>passes_pipeline</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>fa4d35e3-fc38-4e26-a9e0-521a648acec1</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>0.754906</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>54a77010-2f16-4da9-8ca0-746a51801a00</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>0.2882712716542805</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>a8f86b64-914c-4d89-897b-33bcdd1759f7</td>\n",
       "      <td>GBM-US</td>\n",
       "      <td>0.0130328791644564</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>c065761d-f775-457f-bda0-4c7c257a701e</td>\n",
       "      <td>GBM-US</td>\n",
       "      <td>-0.5642192569946982</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>448fe471-3f4e-4dc8-a4e0-6f147dc93abe</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>-0.4010794865307736</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>3f2ecbcd-5c9e-4d57-8aa5-962e2e86a83c</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.5131991180172888</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>6c884037-49f9-41c3-b0e5-9cbcd545aeb7</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.5107224658524304</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>507620ef-f0fc-4ac3-a7c0-ea2e36c40f77</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>0.0422108897825063</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>88f2c496-3eb4-4c03-b3a5-8d5a55803aaf</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.6129595099095041</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>c21a5d55-9790-429f-a4e6-420293c07b53</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>0.2559681573260988</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>5fd77ba9-5015-4d8b-86a0-582e5c76bdd6</td>\n",
       "      <td>GBM-US</td>\n",
       "      <td>0.6562346152437472</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>0dd0718d-5ddf-4c59-8c47-0f51303daeb5</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.4569483162353123</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>8294a969-59a1-4d3b-895b-f07e96bf9f7a</td>\n",
       "      <td>HNSC-US</td>\n",
       "      <td>-0.5371036860447784</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>8c46a1df-1c10-4188-992a-a55b782494c9</td>\n",
       "      <td>LGG-US</td>\n",
       "      <td>0.0668132218607219</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>e2774050-1257-4c29-b201-d79ce70fe17e</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.4379410194660893</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>5310f698-d31a-4628-84b4-dc9ac29722a5</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.0068347114243561</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>d432e99a-67fb-4609-b90f-99438eee7cae</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.5151538970017837</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>978ae91e-6ebe-4efa-97ff-cfad511ae7b3</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.5839295199428811</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>d1132127-1250-43af-9c16-425798a3d1a7</td>\n",
       "      <td>GBM-US</td>\n",
       "      <td>-0.0624706219524289</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>2b142863-b963-4cc9-8f8f-c72503c93390</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>-0.3533774556035514</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>d0e033bc-8e97-4322-b3b1-8626f1e2449b</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.0726002501514517</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>41d50a3c-49fc-4d8a-9c25-bbbcd3a5da6f</td>\n",
       "      <td>GBM-US</td>\n",
       "      <td>-0.2907228368770588</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>0c7aca3f-e006-4de3-afc2-20b4f727d4fd</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>-0.2457822675716241</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>affa4ef4-02ec-40f1-b6a2-739f8bf2afa9</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.1417866820777039</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>10bb1a92-901e-4a14-80f4-5e88f997754b</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.4366975067097288</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td>fc249113-83d4-4abe-8c80-a4f7305dcd91</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>-0.4311616315076992</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                     pid   cohort               zscore  \\\n",
       "0   fa4d35e3-fc38-4e26-a9e0-521a648acec1  SKCM-US             0.754906   \n",
       "1   54a77010-2f16-4da9-8ca0-746a51801a00  SKCM-US   0.2882712716542805   \n",
       "2   a8f86b64-914c-4d89-897b-33bcdd1759f7   GBM-US   0.0130328791644564   \n",
       "3   c065761d-f775-457f-bda0-4c7c257a701e   GBM-US  -0.5642192569946982   \n",
       "4   448fe471-3f4e-4dc8-a4e0-6f147dc93abe  BLCA-US  -0.4010794865307736   \n",
       "6   3f2ecbcd-5c9e-4d57-8aa5-962e2e86a83c  SKCM-US  -0.5131991180172888   \n",
       "7   6c884037-49f9-41c3-b0e5-9cbcd545aeb7  SKCM-US  -0.5107224658524304   \n",
       "8   507620ef-f0fc-4ac3-a7c0-ea2e36c40f77  SKCM-US   0.0422108897825063   \n",
       "9   88f2c496-3eb4-4c03-b3a5-8d5a55803aaf  SKCM-US  -0.6129595099095041   \n",
       "11  c21a5d55-9790-429f-a4e6-420293c07b53  SKCM-US   0.2559681573260988   \n",
       "12  5fd77ba9-5015-4d8b-86a0-582e5c76bdd6   GBM-US   0.6562346152437472   \n",
       "14  0dd0718d-5ddf-4c59-8c47-0f51303daeb5  SKCM-US  -0.4569483162353123   \n",
       "15  8294a969-59a1-4d3b-895b-f07e96bf9f7a  HNSC-US  -0.5371036860447784   \n",
       "16  8c46a1df-1c10-4188-992a-a55b782494c9   LGG-US   0.0668132218607219   \n",
       "17  e2774050-1257-4c29-b201-d79ce70fe17e  SKCM-US  -0.4379410194660893   \n",
       "19  5310f698-d31a-4628-84b4-dc9ac29722a5  SKCM-US  -0.0068347114243561   \n",
       "20  d432e99a-67fb-4609-b90f-99438eee7cae  SKCM-US  -0.5151538970017837   \n",
       "22  978ae91e-6ebe-4efa-97ff-cfad511ae7b3  SKCM-US  -0.5839295199428811   \n",
       "24  d1132127-1250-43af-9c16-425798a3d1a7   GBM-US  -0.0624706219524289   \n",
       "26  2b142863-b963-4cc9-8f8f-c72503c93390  BLCA-US  -0.3533774556035514   \n",
       "27  d0e033bc-8e97-4322-b3b1-8626f1e2449b  SKCM-US  -0.0726002501514517   \n",
       "28  41d50a3c-49fc-4d8a-9c25-bbbcd3a5da6f   GBM-US  -0.2907228368770588   \n",
       "30  0c7aca3f-e006-4de3-afc2-20b4f727d4fd  BLCA-US  -0.2457822675716241   \n",
       "31  affa4ef4-02ec-40f1-b6a2-739f8bf2afa9  SKCM-US  -0.1417866820777039   \n",
       "33  10bb1a92-901e-4a14-80f4-5e88f997754b  SKCM-US  -0.4366975067097288   \n",
       "34  fc249113-83d4-4abe-8c80-a4f7305dcd91  SKCM-US  -0.4311616315076992   \n",
       "\n",
       "   passes_pipeline  \n",
       "0             True  \n",
       "1             True  \n",
       "2            False  \n",
       "3            False  \n",
       "4            False  \n",
       "6            False  \n",
       "7            False  \n",
       "8             True  \n",
       "9            False  \n",
       "11            True  \n",
       "12            True  \n",
       "14           False  \n",
       "15           False  \n",
       "16            True  \n",
       "17           False  \n",
       "19           False  \n",
       "20           False  \n",
       "22           False  \n",
       "24           False  \n",
       "26           False  \n",
       "27           False  \n",
       "28           False  \n",
       "30           False  \n",
       "31           False  \n",
       "33           False  \n",
       "34           False  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tert_c250t_dataframe[tert_c250t_dataframe[\"zscore\"] != \"not_available\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Why did the RALY mutations drop?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>pid</th>\n",
       "      <th>cohort</th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SEQUENCE_CONTEXT</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>strand</th>\n",
       "      <th>...</th>\n",
       "      <th>destroyed_tfs_passing_tf_expression_threshold</th>\n",
       "      <th>num_remaining_tfs</th>\n",
       "      <th>remaining_tfs</th>\n",
       "      <th>JASPAR2020_CORE_vertebrates_non_redundant(tf_name,binding_affinity,seq1,seq2,raw,zscore,log,tf_sequence_logo)</th>\n",
       "      <th>paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))</th>\n",
       "      <th>num_original_mutations</th>\n",
       "      <th>num_final_mutations</th>\n",
       "      <th>num_promoter_mutations</th>\n",
       "      <th>num_recurrent_mutations</th>\n",
       "      <th>cn_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>70.38</td>\n",
       "      <td>d053f88f-09e6-472e-a691-bb293c461399</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>20</td>\n",
       "      <td>32580927</td>\n",
       "      <td>RALY</td>\n",
       "      <td>TGCCTCGTCT,TTCCGGATCC</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>ELF5,ZBTB7A</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ELK4,ELF2</td>\n",
       "      <td>ELK4,0.09259259259259259,TCTCTTCCGGA,.,5.67634...</td>\n",
       "      <td>/omics/groups/OE0436/internal/nabad/results/16...</td>\n",
       "      <td>19832</td>\n",
       "      <td>87</td>\n",
       "      <td>420</td>\n",
       "      <td>15.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82</th>\n",
       "      <td>69.54</td>\n",
       "      <td>b07bad52-d44c-4b27-900a-960985bfadec</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>20</td>\n",
       "      <td>32580927</td>\n",
       "      <td>RALY</td>\n",
       "      <td>TGCCTCGTCT,TTCCGGATCC</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>ZBTB7A</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ELK4,ELF5,ELF2</td>\n",
       "      <td>ELK4,0.09259259259259259,TCTCTTCCGGA,.,3.81399...</td>\n",
       "      <td>/omics/groups/OE0436/internal/nabad/results/16...</td>\n",
       "      <td>158060</td>\n",
       "      <td>710</td>\n",
       "      <td>3344</td>\n",
       "      <td>15.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>287</th>\n",
       "      <td>60.91</td>\n",
       "      <td>3869ff3f-21b9-4817-8ff4-83c6fc75ab11</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>20</td>\n",
       "      <td>32580927</td>\n",
       "      <td>RALY</td>\n",
       "      <td>TGCCTCGTCT,TTCCGGATCC</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>ELF5,ZBTB7A</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ELK4,ELF2</td>\n",
       "      <td>ELK4,0.09259259259259259,TCTCTTCCGGA,.,4.02060...</td>\n",
       "      <td>/omics/groups/OE0436/internal/nabad/results/16...</td>\n",
       "      <td>171882</td>\n",
       "      <td>809</td>\n",
       "      <td>3472</td>\n",
       "      <td>15.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>409</th>\n",
       "      <td>58.51</td>\n",
       "      <td>e2774050-1257-4c29-b201-d79ce70fe17e</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>20</td>\n",
       "      <td>32580927</td>\n",
       "      <td>RALY</td>\n",
       "      <td>TGCCTCGTCT,TTCCGGATCC</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>ZBTB7A</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ELK4,ELF5,ELF2</td>\n",
       "      <td>ELK4,0.09259259259259259,TCTCTTCCGGA,.,6.14689...</td>\n",
       "      <td>/omics/groups/OE0436/internal/nabad/results/16...</td>\n",
       "      <td>164045</td>\n",
       "      <td>713</td>\n",
       "      <td>3175</td>\n",
       "      <td>15.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>882</th>\n",
       "      <td>52.91</td>\n",
       "      <td>3869ff3f-21b9-4817-8ff4-83c6fc75ab11</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>20</td>\n",
       "      <td>32580916</td>\n",
       "      <td>RALY</td>\n",
       "      <td>GAGTTTCTTC,TGCCTCGTCT</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>E2F6,ETV1,ETV4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>GABPA,ELF1</td>\n",
       "      <td>GABPA,0.09491538293611246,GTTTCTTCCTGCCT,.,4.9...</td>\n",
       "      <td>/omics/groups/OE0436/internal/nabad/results/16...</td>\n",
       "      <td>171882</td>\n",
       "      <td>809</td>\n",
       "      <td>3472</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1161</th>\n",
       "      <td>51.08</td>\n",
       "      <td>760881cc-c623-11e3-bf01-24c6515278c0</td>\n",
       "      <td>LIRI-JP</td>\n",
       "      <td>20</td>\n",
       "      <td>32581904</td>\n",
       "      <td>RALY</td>\n",
       "      <td>CCTCCTTCCC,GCCGCGCGGC</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>RBPJ</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NRF1,ZNF263</td>\n",
       "      <td>E2F6,12.7642,.,GCGGCCGGGAAGG,4.39819574116,0.0...</td>\n",
       "      <td>.</td>\n",
       "      <td>184214</td>\n",
       "      <td>647</td>\n",
       "      <td>3397</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1508</th>\n",
       "      <td>49.47</td>\n",
       "      <td>0be08326-c623-11e3-bf01-24c6515278c0</td>\n",
       "      <td>LIRI-JP</td>\n",
       "      <td>20</td>\n",
       "      <td>32580986</td>\n",
       "      <td>RALY</td>\n",
       "      <td>CTCCAGGCCC,GCCCACCTCA</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>KLF4,EGR1,KLF5,KLF9,KLF10,KLF11,KLF15,KLF2,KLF...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>SP1,SP2,SP4,KLF16,SP3,SP8,ZNF740,MAZ,SP9</td>\n",
       "      <td>KLF4,0.06940058712896711,GGCCCCGCCCAC,.,4.6798...</td>\n",
       "      <td>.</td>\n",
       "      <td>6940</td>\n",
       "      <td>25</td>\n",
       "      <td>105</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      score                                   pid   cohort #CHROM       POS  \\\n",
       "70    70.38  d053f88f-09e6-472e-a691-bb293c461399  SKCM-US     20  32580927   \n",
       "82    69.54  b07bad52-d44c-4b27-900a-960985bfadec  SKCM-US     20  32580927   \n",
       "287   60.91  3869ff3f-21b9-4817-8ff4-83c6fc75ab11  SKCM-US     20  32580927   \n",
       "409   58.51  e2774050-1257-4c29-b201-d79ce70fe17e  SKCM-US     20  32580927   \n",
       "882   52.91  3869ff3f-21b9-4817-8ff4-83c6fc75ab11  SKCM-US     20  32580916   \n",
       "1161  51.08  760881cc-c623-11e3-bf01-24c6515278c0  LIRI-JP     20  32581904   \n",
       "1508  49.47  0be08326-c623-11e3-bf01-24c6515278c0  LIRI-JP     20  32580986   \n",
       "\n",
       "      GENE       SEQUENCE_CONTEXT REF ALT strand  ...  \\\n",
       "70    RALY  TGCCTCGTCT,TTCCGGATCC   C   T      +  ...   \n",
       "82    RALY  TGCCTCGTCT,TTCCGGATCC   C   T      +  ...   \n",
       "287   RALY  TGCCTCGTCT,TTCCGGATCC   C   T      +  ...   \n",
       "409   RALY  TGCCTCGTCT,TTCCGGATCC   C   T      +  ...   \n",
       "882   RALY  GAGTTTCTTC,TGCCTCGTCT   C   T      +  ...   \n",
       "1161  RALY  CCTCCTTCCC,GCCGCGCGGC   A   G      +  ...   \n",
       "1508  RALY  CTCCAGGCCC,GCCCACCTCA   C   T      +  ...   \n",
       "\n",
       "          destroyed_tfs_passing_tf_expression_threshold  num_remaining_tfs  \\\n",
       "70                                          ELF5,ZBTB7A                2.0   \n",
       "82                                               ZBTB7A                3.0   \n",
       "287                                         ELF5,ZBTB7A                2.0   \n",
       "409                                              ZBTB7A                3.0   \n",
       "882                                      E2F6,ETV1,ETV4                2.0   \n",
       "1161                                               RBPJ                2.0   \n",
       "1508  KLF4,EGR1,KLF5,KLF9,KLF10,KLF11,KLF15,KLF2,KLF...                9.0   \n",
       "\n",
       "                                 remaining_tfs  \\\n",
       "70                                   ELK4,ELF2   \n",
       "82                              ELK4,ELF5,ELF2   \n",
       "287                                  ELK4,ELF2   \n",
       "409                             ELK4,ELF5,ELF2   \n",
       "882                                 GABPA,ELF1   \n",
       "1161                               NRF1,ZNF263   \n",
       "1508  SP1,SP2,SP4,KLF16,SP3,SP8,ZNF740,MAZ,SP9   \n",
       "\n",
       "      JASPAR2020_CORE_vertebrates_non_redundant(tf_name,binding_affinity,seq1,seq2,raw,zscore,log,tf_sequence_logo)  \\\n",
       "70    ELK4,0.09259259259259259,TCTCTTCCGGA,.,5.67634...                                                               \n",
       "82    ELK4,0.09259259259259259,TCTCTTCCGGA,.,3.81399...                                                               \n",
       "287   ELK4,0.09259259259259259,TCTCTTCCGGA,.,4.02060...                                                               \n",
       "409   ELK4,0.09259259259259259,TCTCTTCCGGA,.,6.14689...                                                               \n",
       "882   GABPA,0.09491538293611246,GTTTCTTCCTGCCT,.,4.9...                                                               \n",
       "1161  E2F6,12.7642,.,GCGGCCGGGAAGG,4.39819574116,0.0...                                                               \n",
       "1508  KLF4,0.06940058712896711,GGCCCCGCCCAC,.,4.6798...                                                               \n",
       "\n",
       "     paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))  \\\n",
       "70    /omics/groups/OE0436/internal/nabad/results/16...                                                                                        \n",
       "82    /omics/groups/OE0436/internal/nabad/results/16...                                                                                        \n",
       "287   /omics/groups/OE0436/internal/nabad/results/16...                                                                                        \n",
       "409   /omics/groups/OE0436/internal/nabad/results/16...                                                                                        \n",
       "882   /omics/groups/OE0436/internal/nabad/results/16...                                                                                        \n",
       "1161                                                  .                                                                                        \n",
       "1508                                                  .                                                                                        \n",
       "\n",
       "     num_original_mutations  num_final_mutations num_promoter_mutations  \\\n",
       "70                    19832                   87                    420   \n",
       "82                   158060                  710                   3344   \n",
       "287                  171882                  809                   3472   \n",
       "409                  164045                  713                   3175   \n",
       "882                  171882                  809                   3472   \n",
       "1161                 184214                  647                   3397   \n",
       "1508                   6940                   25                    105   \n",
       "\n",
       "      num_recurrent_mutations       cn_score  \n",
       "70                       15.0              3  \n",
       "82                       15.0  not_available  \n",
       "287                      15.0              4  \n",
       "409                      15.0              4  \n",
       "882                       1.0              4  \n",
       "1161                      0.0  not_available  \n",
       "1508                      0.0  not_available  \n",
       "\n",
       "[7 rows x 29 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_pcawg_and_eopc[combined_pcawg_and_eopc[\"GENE\"] == \"RALY\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2s/38kg81v526j7qcgd8gdq8jd00000gn/T/ipykernel_64597/908769902.py:14: FutureWarning: Setting an item of incompatible dtype is deprecated and will raise an error in a future version of pandas. Value 'not_available' has dtype incompatible with float64, please explicitly cast to a compatible dtype first.\n",
      "  raly_dataframe.loc[idx+1, \"zscore\"] = zscore\n"
     ]
    }
   ],
   "source": [
    "raly = combined_pcawg_and_eopc.iloc[70]\n",
    "\n",
    "raly_dataframe = pd.DataFrame()\n",
    "raly_dataframe.loc[0, \"pid\"] = raly[\"pid\"]\n",
    "raly_dataframe.loc[0, \"cohort\"] = raly[\"cohort\"]\n",
    "raly_dataframe.loc[0, \"zscore\"] = raly[\"FPKM_Z_score\"]\n",
    "\n",
    "\n",
    "\n",
    "for idx, entry in enumerate(raly[\"paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))\"].split(\";\")):\n",
    "    path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af = entry.split(\",\")[:14]\n",
    "    raly_dataframe.loc[idx+1, \"pid\"] = pid\n",
    "    raly_dataframe.loc[idx+1, \"cohort\"] = cohort\n",
    "    raly_dataframe.loc[idx+1, \"zscore\"] = zscore\n",
    "    \n",
    "pids_of_raly_passing_pipeline = list(data[(data[\"POS\"] == 32580927) & (data[\"GENE\"] == \"RALY\")][\"pid\"].unique())\n",
    "\n",
    "for idx, row in raly_dataframe.iterrows():\n",
    "    raly_dataframe.loc[idx, \"passes_pipeline\"] = row[\"pid\"] in pids_of_tert_c250ts_passing_pipeline"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Top 100"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2s/38kg81v526j7qcgd8gdq8jd00000gn/T/ipykernel_64597/2506053702.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  combined_pcawg_and_eopc.iloc[:100][(combined_pcawg_and_eopc[\"GENE\"] == \"TERT\") & (combined_pcawg_and_eopc[\"POS\"] == 1295228)].shape\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(24, 29)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_pcawg_and_eopc.iloc[:100][(combined_pcawg_and_eopc[\"GENE\"] == \"TERT\") & (combined_pcawg_and_eopc[\"POS\"] == 1295228)].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2s/38kg81v526j7qcgd8gdq8jd00000gn/T/ipykernel_8957/106539731.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  combined_pcawg_and_eopc.iloc[:100][(combined_pcawg_and_eopc[\"GENE\"] == \"TERT\") & (combined_pcawg_and_eopc[\"POS\"] == 1295250)].shape\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(6, 29)"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_pcawg_and_eopc.iloc[:100][(combined_pcawg_and_eopc[\"GENE\"] == \"TERT\") & (combined_pcawg_and_eopc[\"POS\"] == 1295250)].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/2s/38kg81v526j7qcgd8gdq8jd00000gn/T/ipykernel_8957/925935169.py:1: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  combined_pcawg_and_eopc.iloc[:100][(combined_pcawg_and_eopc[\"GENE\"] == \"RALY\") & (combined_pcawg_and_eopc[\"POS\"] == 32580927)].shape\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(2, 29)"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "combined_pcawg_and_eopc.iloc[:100][(combined_pcawg_and_eopc[\"GENE\"] == \"RALY\") & (combined_pcawg_and_eopc[\"POS\"] == 32580927)].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "top_100 = combined_pcawg_and_eopc.iloc[:100]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "count_top_100 = top_100.groupby([\"GENE\", \"POS\"]).agg({\"pid\": \"count\"}).sort_values(\"pid\", ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>pid</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GENE</th>\n",
       "      <th>POS</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">TERT</th>\n",
       "      <th>1295228</th>\n",
       "      <td>24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1295250</th>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ANKRD53</th>\n",
       "      <th>71204529</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SECISBP2</th>\n",
       "      <th>91933357</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RALY</th>\n",
       "      <th>32580927</th>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RSPO3</th>\n",
       "      <th>127439062</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RPL36AL</th>\n",
       "      <th>50087348</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RFX5</th>\n",
       "      <th>151320371</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PTPN9</th>\n",
       "      <th>75871490</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRKAR1A</th>\n",
       "      <th>66507534</th>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    pid\n",
       "GENE     POS           \n",
       "TERT     1295228     24\n",
       "         1295250      6\n",
       "ANKRD53  71204529     2\n",
       "SECISBP2 91933357     2\n",
       "RALY     32580927     2\n",
       "RSPO3    127439062    1\n",
       "RPL36AL  50087348     1\n",
       "RFX5     151320371    1\n",
       "PTPN9    75871490     1\n",
       "PRKAR1A  66507534     1"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "count_top_100.iloc[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1     6.0\n",
       "61    6.0\n",
       "Name: num_recurrent_mutations, dtype: float64"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_100[top_100[\"GENE\"] == \"SECISBP2\"][\"num_recurrent_mutations\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>score</th>\n",
       "      <th>pid</th>\n",
       "      <th>cohort</th>\n",
       "      <th>#CHROM</th>\n",
       "      <th>POS</th>\n",
       "      <th>GENE</th>\n",
       "      <th>SEQUENCE_CONTEXT</th>\n",
       "      <th>REF</th>\n",
       "      <th>ALT</th>\n",
       "      <th>strand</th>\n",
       "      <th>...</th>\n",
       "      <th>destroyed_tfs_passing_tf_expression_threshold</th>\n",
       "      <th>num_remaining_tfs</th>\n",
       "      <th>remaining_tfs</th>\n",
       "      <th>JASPAR2020_CORE_vertebrates_non_redundant(tf_name,binding_affinity,seq1,seq2,raw,zscore,log,tf_sequence_logo)</th>\n",
       "      <th>paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))</th>\n",
       "      <th>num_original_mutations</th>\n",
       "      <th>num_final_mutations</th>\n",
       "      <th>num_promoter_mutations</th>\n",
       "      <th>num_recurrent_mutations</th>\n",
       "      <th>cn_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>86.30</td>\n",
       "      <td>9988eb07-01f6-4f83-8699-bb63e0525f08</td>\n",
       "      <td>HNSC-US</td>\n",
       "      <td>7</td>\n",
       "      <td>145813786</td>\n",
       "      <td>CNTNAP2</td>\n",
       "      <td>GCTGAGACTC,CCTGCCGCTC</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>ZNF148</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NHLH2,SNAI3</td>\n",
       "      <td>ZEB1,14.5102,.,CTCACCTGCCG,0.832158029939,-0.7...</td>\n",
       "      <td>.</td>\n",
       "      <td>23779</td>\n",
       "      <td>92</td>\n",
       "      <td>459</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>82.68</td>\n",
       "      <td>04c09376-8756-41d9-bbca-f464170a8cfb</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>12</td>\n",
       "      <td>54332280</td>\n",
       "      <td>HOXC13</td>\n",
       "      <td>CCCTCCCTCT,CCTCCCTGGG</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>E2F6</td>\n",
       "      <td>4.0</td>\n",
       "      <td>GABPA,KLF5,ZNF460,ZNF148</td>\n",
       "      <td>GABPA,10.4714,.,CCCTCTTCCTCCCT,5.56252436512,-...</td>\n",
       "      <td>.</td>\n",
       "      <td>32589</td>\n",
       "      <td>176</td>\n",
       "      <td>629</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>82.00</td>\n",
       "      <td>5ead73fe-1c34-48ed-b20d-89fc3c82dbd6</td>\n",
       "      <td>LIHC-US</td>\n",
       "      <td>5</td>\n",
       "      <td>170736726</td>\n",
       "      <td>TLX3</td>\n",
       "      <td>CGGTCTCCAG,CTTGGCGGTC</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NFIC</td>\n",
       "      <td>NFIC,9.54472,.,GACTTGGCGGT,27.9809277799,0.238...</td>\n",
       "      <td>.</td>\n",
       "      <td>2157</td>\n",
       "      <td>9</td>\n",
       "      <td>42</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>80.72</td>\n",
       "      <td>199bbb0f-996c-40c1-b06d-2066f04be778</td>\n",
       "      <td>LUAD-US</td>\n",
       "      <td>15</td>\n",
       "      <td>57210473</td>\n",
       "      <td>TCF12</td>\n",
       "      <td>GGGCAGCGGA,CCAGATGCTT</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ZBTB18,TWIST1,HAND2</td>\n",
       "      <td>HINFP,13.0606,.,CTGGGTCCGCTG,2.8152864897,0.62...</td>\n",
       "      <td>.</td>\n",
       "      <td>17648</td>\n",
       "      <td>68</td>\n",
       "      <td>412</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>80.13</td>\n",
       "      <td>3869ff3f-21b9-4817-8ff4-83c6fc75ab11</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>3</td>\n",
       "      <td>169381814</td>\n",
       "      <td>MECOM</td>\n",
       "      <td>AGCCAGGGGG,AAAAAAACTT</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>INSM1</td>\n",
       "      <td>INSM1,12.0408,.,AGCCAGGGGGAA,0.0,-0.3815070014...</td>\n",
       "      <td>.</td>\n",
       "      <td>171882</td>\n",
       "      <td>809</td>\n",
       "      <td>3472</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>77.52</td>\n",
       "      <td>dcf45005-7e6d-4c26-9994-f973da7e814f</td>\n",
       "      <td>LIHC-US</td>\n",
       "      <td>22</td>\n",
       "      <td>29999735</td>\n",
       "      <td>NF2</td>\n",
       "      <td>GCCGGGTCGC,CCTGCACCGA</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>PAX5</td>\n",
       "      <td>8.0</td>\n",
       "      <td>TCF3,TBX4,TBX5,TCF4,ASCL1,NHLH2,ASCL1,TCF12</td>\n",
       "      <td>PAX5,0.07818058150716525,AGGCGCGACCCG,.,0.0093...</td>\n",
       "      <td>.</td>\n",
       "      <td>4582</td>\n",
       "      <td>19</td>\n",
       "      <td>106</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>77.00</td>\n",
       "      <td>04c09376-8756-41d9-bbca-f464170a8cfb</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>1</td>\n",
       "      <td>190446811</td>\n",
       "      <td>BRINP3</td>\n",
       "      <td>GGGGGGTGGG,AATGGTGGGG</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>MZF1,KLF16,RELB,KLF10,SP9,ZNF148</td>\n",
       "      <td>9.0</td>\n",
       "      <td>KLF4,RREB1,SP1,KLF5,SP3,SP8,ZNF740,KLF11,KLF2</td>\n",
       "      <td>KLF4,0.7129713823848856,TTCCCCACCCCC,TTTCCCACC...</td>\n",
       "      <td>.</td>\n",
       "      <td>32589</td>\n",
       "      <td>176</td>\n",
       "      <td>629</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>76.92</td>\n",
       "      <td>66d312fc-809d-428b-98e3-a29d211db35c</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>1</td>\n",
       "      <td>14026387</td>\n",
       "      <td>PRDM2</td>\n",
       "      <td>GATATTTAAT,ACCTCCAACG</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>RORA,HOXC9,HOXC11,POU6F2,HOXC10,HOXD11,NR1H4,H...</td>\n",
       "      <td>21.0</td>\n",
       "      <td>PDX1,ALX3,NKX6-1,NKX6-2,MEOX2,VAX1,VAX2,HOXD12...</td>\n",
       "      <td>RORA,0.08166665305555783,TTGGAGGTCA,.,0.463684...</td>\n",
       "      <td>.</td>\n",
       "      <td>40281</td>\n",
       "      <td>168</td>\n",
       "      <td>742</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>74.79</td>\n",
       "      <td>0c7aca3f-e006-4de3-afc2-20b4f727d4fd</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>11</td>\n",
       "      <td>113929845</td>\n",
       "      <td>ZBTB16</td>\n",
       "      <td>GGACTGCAGC,CCGAGCCATG</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>TFAP2A</td>\n",
       "      <td>2.0</td>\n",
       "      <td>TFAP2C,TFAP2B</td>\n",
       "      <td>TFAP2C,0.09271535458987362,TGGCTCGGGGCT,.,16.6...</td>\n",
       "      <td>.</td>\n",
       "      <td>7978</td>\n",
       "      <td>37</td>\n",
       "      <td>170</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>74.72</td>\n",
       "      <td>2399ab13-abfa-480e-9fda-7947edc420be</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>1</td>\n",
       "      <td>35658792</td>\n",
       "      <td>SFPQ</td>\n",
       "      <td>GCTGGCACGG,CCCGCCTACT</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>SP1,EGR1,KLF5,KLF16,KLF2</td>\n",
       "      <td>8.0</td>\n",
       "      <td>PAX5,NFIX,EGR3,SP3,SP8,KLF10,KLF11,SP9</td>\n",
       "      <td>PAX5,10.6909,.,AGGCGGGACCGT,0.0041284764193,-0...</td>\n",
       "      <td>.</td>\n",
       "      <td>26144</td>\n",
       "      <td>143</td>\n",
       "      <td>532</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>74.02</td>\n",
       "      <td>dfa20d2d-1ba3-476a-b4b9-21d21470bee2</td>\n",
       "      <td>HNSC-US</td>\n",
       "      <td>20</td>\n",
       "      <td>43976570</td>\n",
       "      <td>SDC4</td>\n",
       "      <td>AGGGCCATCT,GGATCCTAAA</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>PLAG1,NEUROD1</td>\n",
       "      <td>PLAG1,6.08163,.,AGGGCCATCTGGGA,1.6972677727,0....</td>\n",
       "      <td>.</td>\n",
       "      <td>28551</td>\n",
       "      <td>160</td>\n",
       "      <td>726</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>74.00</td>\n",
       "      <td>3b7810f7-f8ff-4d62-b766-3ba06170194c</td>\n",
       "      <td>RECA-EU</td>\n",
       "      <td>3</td>\n",
       "      <td>187463635</td>\n",
       "      <td>BCL6</td>\n",
       "      <td>TAAGAGGGGG,GTGCGCTTTG</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>MSANTD3</td>\n",
       "      <td>4.0</td>\n",
       "      <td>EGR1,KLF16,ZNF740,KLF10</td>\n",
       "      <td>EGR1,10.2114,.,GCGCACCCCCCCTC,4.70543294151,-0...</td>\n",
       "      <td>.</td>\n",
       "      <td>21836</td>\n",
       "      <td>111</td>\n",
       "      <td>490</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>73.46</td>\n",
       "      <td>b73523d7-f5a5-4140-8537-4df4d1ecf465</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>19</td>\n",
       "      <td>13048717</td>\n",
       "      <td>CALR</td>\n",
       "      <td>CCCCCTCTGT,TGCTCAGGAA</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>FOXD2</td>\n",
       "      <td>FOXD2,10.7857,.,CCTGAGCAAACAG,0.417828374569,-...</td>\n",
       "      <td>.</td>\n",
       "      <td>30204</td>\n",
       "      <td>174</td>\n",
       "      <td>1034</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td>73.42</td>\n",
       "      <td>04aa6b77-8074-480c-872e-a1a47afa5314</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>3</td>\n",
       "      <td>71633880</td>\n",
       "      <td>FOXP1</td>\n",
       "      <td>TGCAAGGAAA,AGAAACCCCG</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>IRF9,PRDM4</td>\n",
       "      <td>4.0</td>\n",
       "      <td>STAT1,STAT2,IRF8,IRF4</td>\n",
       "      <td>STAT1,0.9441198356465754,GGGGTTTCTGTTTCC,GGGGT...</td>\n",
       "      <td>.</td>\n",
       "      <td>342233</td>\n",
       "      <td>1348</td>\n",
       "      <td>6140</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>73.34</td>\n",
       "      <td>448fe471-3f4e-4dc8-a4e0-6f147dc93abe</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>6</td>\n",
       "      <td>37138454</td>\n",
       "      <td>PIM1</td>\n",
       "      <td>CCCCCGCCTC,GGCCCGGGGA</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>ZBTB7A,KLF11,TFAP2E</td>\n",
       "      <td>8.0</td>\n",
       "      <td>TFAP2A,KLF16,TFAP2A,TFAP2B,TFAP2C,MAZ,SP9,ZNF148</td>\n",
       "      <td>TFAP2A,0.9195551902087045,CCCGCCTCCGGCCC,CCCGC...</td>\n",
       "      <td>.</td>\n",
       "      <td>19009</td>\n",
       "      <td>95</td>\n",
       "      <td>488</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>73.03</td>\n",
       "      <td>9321341c-c622-11e3-bf01-24c6515278c0</td>\n",
       "      <td>LIRI-JP</td>\n",
       "      <td>3</td>\n",
       "      <td>30647905</td>\n",
       "      <td>TGFBR2</td>\n",
       "      <td>GGGCGGTGAG,GGCAGCTGAA</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>ZNF148</td>\n",
       "      <td>2.0</td>\n",
       "      <td>PKNOX1,PBX3</td>\n",
       "      <td>SMAD2,11.3443,.,CTGCCTCTCACCG,7.06965725325,1....</td>\n",
       "      <td>.</td>\n",
       "      <td>3410</td>\n",
       "      <td>13</td>\n",
       "      <td>49</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>73.00</td>\n",
       "      <td>3869ff3f-21b9-4817-8ff4-83c6fc75ab11</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>8</td>\n",
       "      <td>68658276</td>\n",
       "      <td>CPA6</td>\n",
       "      <td>ATAAAGGTGG,TGTGCCCCGG</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>ZNF341</td>\n",
       "      <td>7.0</td>\n",
       "      <td>SP1,ESR2,KLF13,NFKB2,KLF9,KLF3,SP9</td>\n",
       "      <td>SP1,3.38298,.,CGGGGCACACCCACC,13.0141602848,-1...</td>\n",
       "      <td>.</td>\n",
       "      <td>171882</td>\n",
       "      <td>809</td>\n",
       "      <td>3472</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>72.69</td>\n",
       "      <td>75ba6722-1148-4a52-a9ed-68d890238205</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>12</td>\n",
       "      <td>95945540</td>\n",
       "      <td>USP44</td>\n",
       "      <td>CTTCCCTGAT,CCCAGATGGA</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>MZF1</td>\n",
       "      <td>3.0</td>\n",
       "      <td>STAT3,RHOXF1,HAND2</td>\n",
       "      <td>MZF1,0.08081461128171973,CTGATCCCCAGAT,.,2.029...</td>\n",
       "      <td>.</td>\n",
       "      <td>127544</td>\n",
       "      <td>641</td>\n",
       "      <td>2368</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40</th>\n",
       "      <td>72.54</td>\n",
       "      <td>14c5b81d-da49-4db1-9834-77711c2b1d38</td>\n",
       "      <td>READ-US</td>\n",
       "      <td>4</td>\n",
       "      <td>106067000</td>\n",
       "      <td>TET2</td>\n",
       "      <td>TCTGTGAGAC,CCCCCTTCTA</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>CTCFL,KLF11,SP9</td>\n",
       "      <td>6.0</td>\n",
       "      <td>KLF13,SP4,KLF16,SP3,KLF10,KLF6</td>\n",
       "      <td>KLF13,0.4498060359495164,GTGAGACGCCCCCTTCTA,GT...</td>\n",
       "      <td>.</td>\n",
       "      <td>548085</td>\n",
       "      <td>1685</td>\n",
       "      <td>7490</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43</th>\n",
       "      <td>72.26</td>\n",
       "      <td>0bbc7ede-5022-4084-925c-d65baaf7abc2</td>\n",
       "      <td>LUSC-US</td>\n",
       "      <td>22</td>\n",
       "      <td>41486807</td>\n",
       "      <td>EP300</td>\n",
       "      <td>TTTTTCCCAT,ATTACATTAG</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>STAT1,PROP1</td>\n",
       "      <td>4.0</td>\n",
       "      <td>MEOX2,PHOX2A,FOXB1,DUXA</td>\n",
       "      <td>STAT1,0.08986906077844581,ATTATGGGAAA,.,81.441...</td>\n",
       "      <td>.</td>\n",
       "      <td>17905</td>\n",
       "      <td>92</td>\n",
       "      <td>304</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44</th>\n",
       "      <td>72.18</td>\n",
       "      <td>49dec0c2-8e75-4f44-a253-82b2ea605890</td>\n",
       "      <td>LUAD-US</td>\n",
       "      <td>6</td>\n",
       "      <td>127439062</td>\n",
       "      <td>RSPO3</td>\n",
       "      <td>AAGGAAGAAA,AAAGAAAAGA</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EWSR1-FLI1,12.602921305366829,GGAAGAAAGAAAGAAA...</td>\n",
       "      <td>.</td>\n",
       "      <td>22203</td>\n",
       "      <td>91</td>\n",
       "      <td>436</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>71.45</td>\n",
       "      <td>04c09376-8756-41d9-bbca-f464170a8cfb</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>10</td>\n",
       "      <td>88126034</td>\n",
       "      <td>GRID1</td>\n",
       "      <td>CGTTGGGGCC,CCGCCCAGCC</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>KLF4,SP3,KLF15,KLF2,KLF6,PLAGL2,SP9,ZNF148</td>\n",
       "      <td>5.0</td>\n",
       "      <td>KLF5,SP8,KLF11,MAZ,ZBTB14</td>\n",
       "      <td>KLF4,0.07352941176470588,GCCCCCGCCCAG,.,6.2973...</td>\n",
       "      <td>.</td>\n",
       "      <td>32589</td>\n",
       "      <td>176</td>\n",
       "      <td>629</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>71.00</td>\n",
       "      <td>2290b078-6a5b-4c83-9dfb-b525bbf14e4e</td>\n",
       "      <td>BRCA-US</td>\n",
       "      <td>20</td>\n",
       "      <td>42876352</td>\n",
       "      <td>GDAP1L1</td>\n",
       "      <td>GGGCCCCAGG,GCACAGTTCT</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>EBF1,MYOD1,SNAI2,NHLH2</td>\n",
       "      <td>ZEB1,12.7143,.,TGCACCTGGGG,0.433314718291,-1.3...</td>\n",
       "      <td>.</td>\n",
       "      <td>3408</td>\n",
       "      <td>24</td>\n",
       "      <td>102</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>71.00</td>\n",
       "      <td>760881cc-c623-11e3-bf01-24c6515278c0</td>\n",
       "      <td>LIRI-JP</td>\n",
       "      <td>1</td>\n",
       "      <td>87793862</td>\n",
       "      <td>LMO4</td>\n",
       "      <td>CAGGTGAGGC,CCGCCCTGCG</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>KLF4,SP3,KLF15,KLF2,KLF3,KLF6,ZNF148</td>\n",
       "      <td>6.0</td>\n",
       "      <td>SP1,NFKB1,ZNF263,KLF5,MAZ,SP9</td>\n",
       "      <td>KLF4,0.07198957590940831,GGCCCCGCCCTG,.,50.971...</td>\n",
       "      <td>.</td>\n",
       "      <td>184214</td>\n",
       "      <td>647</td>\n",
       "      <td>3397</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>71.00</td>\n",
       "      <td>14630c19-9ee0-4ab3-9d8b-8c3309d1e7d0</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>15</td>\n",
       "      <td>75871490</td>\n",
       "      <td>PTPN9</td>\n",
       "      <td>GAAATTTCCC,AAGGCCGCCG</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>HSF1,HSF2,HSF4</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>HSF1,0.08079502302658156,TTCGGGAAATTTC,.,44.11...</td>\n",
       "      <td>.</td>\n",
       "      <td>24354</td>\n",
       "      <td>106</td>\n",
       "      <td>472</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>71.00</td>\n",
       "      <td>4c2ad4a0-5d57-4e27-9f35-058b2f205f50</td>\n",
       "      <td>LUSC-US</td>\n",
       "      <td>5</td>\n",
       "      <td>157097791</td>\n",
       "      <td>C5orf52</td>\n",
       "      <td>TTACACAAGA,GGAAGTGAGA</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>ELK1,SPIB,ELF1,ERG,EHF,ELF3,ETV3,ELF2</td>\n",
       "      <td>ELK1,10.6889,.,GACGGAAGTG,18.4019310658,0.0856...</td>\n",
       "      <td>.</td>\n",
       "      <td>28620</td>\n",
       "      <td>209</td>\n",
       "      <td>668</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63</th>\n",
       "      <td>70.87</td>\n",
       "      <td>448fe471-3f4e-4dc8-a4e0-6f147dc93abe</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>4</td>\n",
       "      <td>40058235</td>\n",
       "      <td>N4BP2</td>\n",
       "      <td>GGAGAGGTGG,AGCGCCCGCG</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>TBX18</td>\n",
       "      <td>3.0</td>\n",
       "      <td>E2F6,TBX15,KLF9</td>\n",
       "      <td>E2F6,0.0970798384591488,AGAGGTGGGAGCG,.,3.2812...</td>\n",
       "      <td>.</td>\n",
       "      <td>19009</td>\n",
       "      <td>95</td>\n",
       "      <td>488</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>65</th>\n",
       "      <td>70.68</td>\n",
       "      <td>04c09376-8756-41d9-bbca-f464170a8cfb</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>12</td>\n",
       "      <td>54332288</td>\n",
       "      <td>HOXC13</td>\n",
       "      <td>CTCCCTCCCT,GGGCCTGCCG</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>TFAP2A,TFAP2C,TFAP2E</td>\n",
       "      <td>3.0</td>\n",
       "      <td>EBF1,KLF5,EBF3</td>\n",
       "      <td>TFAP2A,0.0863155350700019,CAGGCCCCAGGGAG,.,10....</td>\n",
       "      <td>.</td>\n",
       "      <td>32589</td>\n",
       "      <td>176</td>\n",
       "      <td>629</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>70.63</td>\n",
       "      <td>2df02f2b-9f1c-4249-b3b4-b03079cd97d9</td>\n",
       "      <td>COAD-US</td>\n",
       "      <td>10</td>\n",
       "      <td>23216440</td>\n",
       "      <td>ARMC3</td>\n",
       "      <td>AAGTGGCTGA,GTGATGACTG</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.0</td>\n",
       "      <td>FOSB,JUN,JUN,JUNB,JUNB,FOSL2,JUND,BACH2</td>\n",
       "      <td>SREBF1,11.2073,.,ATCACGTCAG,33.4166268472,0.63...</td>\n",
       "      <td>.</td>\n",
       "      <td>875975</td>\n",
       "      <td>2705</td>\n",
       "      <td>11145</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>70.49</td>\n",
       "      <td>987528ac-437a-4eb8-a335-4f2076d5c006</td>\n",
       "      <td>BRCA-US</td>\n",
       "      <td>9</td>\n",
       "      <td>123837465</td>\n",
       "      <td>CNTRL</td>\n",
       "      <td>ATGGATTCGG,GCGGGGGCGC</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>KLF4,E2F6,KLF5,CTCFL,KLF2,KLF6,MAZ</td>\n",
       "      <td>9.0</td>\n",
       "      <td>EGR2,KLF16,SP3,SP8,ZNF740,KLF11,KLF15,SP9,ZNF148</td>\n",
       "      <td>KLF4,0.07115158845921235,GCCCCCGCCCCG,.,1.9048...</td>\n",
       "      <td>.</td>\n",
       "      <td>8630</td>\n",
       "      <td>44</td>\n",
       "      <td>212</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>70.44</td>\n",
       "      <td>04aa6b77-8074-480c-872e-a1a47afa5314</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>12</td>\n",
       "      <td>81332010</td>\n",
       "      <td>LIN7A</td>\n",
       "      <td>AAATGTTTGG,ATGTAATTAC</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>HOXA5,HOXB5,HOXB8</td>\n",
       "      <td>3.0</td>\n",
       "      <td>HOXA1,HOXA7,HOXC8</td>\n",
       "      <td>TEAD1,13.439,.,TTACATTCCAAAC,13.2052531214,-0....</td>\n",
       "      <td>.</td>\n",
       "      <td>342233</td>\n",
       "      <td>1348</td>\n",
       "      <td>6140</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18414</th>\n",
       "      <td>69.75</td>\n",
       "      <td>3c92c0e9-68e0-4354-8ed1-b6506878145d</td>\n",
       "      <td>EOPC-DE</td>\n",
       "      <td>6</td>\n",
       "      <td>135501964</td>\n",
       "      <td>MYB</td>\n",
       "      <td>AGTTGTAAAC,TTGACGAAAA</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>FOXG1,FOXK1</td>\n",
       "      <td>FOXD1,13.52,.,GTAAACAT,5.62,4.407493119305658,...</td>\n",
       "      <td>.</td>\n",
       "      <td>1560</td>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>78</th>\n",
       "      <td>69.72</td>\n",
       "      <td>f5c8f98a-c34a-49e8-a655-d8f29487b55c</td>\n",
       "      <td>CLLE-ES</td>\n",
       "      <td>12</td>\n",
       "      <td>92539203</td>\n",
       "      <td>BTG1</td>\n",
       "      <td>AAGGTCTGCA,CTGTCGCTCG</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>ASCL1</td>\n",
       "      <td>8.0</td>\n",
       "      <td>MYOD1,MYOG,SNAI2,BHLHA15,NHLH2,ASCL1,BHLHE22,MYF5</td>\n",
       "      <td>MYOD1,0.809704733676805,CTGCAGCTGTCGC,CTGCATCT...</td>\n",
       "      <td>.</td>\n",
       "      <td>1015</td>\n",
       "      <td>8</td>\n",
       "      <td>26</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>69.70</td>\n",
       "      <td>2df02f2b-9f1c-4249-b3b4-b03079cd97d9</td>\n",
       "      <td>COAD-US</td>\n",
       "      <td>4</td>\n",
       "      <td>55523332</td>\n",
       "      <td>KIT</td>\n",
       "      <td>CAGAGGCAGA,AAATCATTAC</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>BARX2</td>\n",
       "      <td>PBX2,12.3465,.,CAGATAAATCATT,10.724192169,0.04...</td>\n",
       "      <td>.</td>\n",
       "      <td>875975</td>\n",
       "      <td>2705</td>\n",
       "      <td>11145</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>69.55</td>\n",
       "      <td>760881cc-c623-11e3-bf01-24c6515278c0</td>\n",
       "      <td>LIRI-JP</td>\n",
       "      <td>19</td>\n",
       "      <td>55582410</td>\n",
       "      <td>EPS8L1</td>\n",
       "      <td>CAAGCCCCAG,GGAGAATTCT</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>TFAP2A,TFAP2B,TFAP2E</td>\n",
       "      <td>6.0</td>\n",
       "      <td>TFAP2A,EBF1,TFAP2C,TFAP2B,TFAP2C,EBF3</td>\n",
       "      <td>TFAP2A,0.09213447948625815,CAAGCCCCAGGGGA,.,2....</td>\n",
       "      <td>.</td>\n",
       "      <td>184214</td>\n",
       "      <td>647</td>\n",
       "      <td>3397</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83</th>\n",
       "      <td>69.19</td>\n",
       "      <td>94108975-b7a0-40ba-ad39-e44cc62e8cc0</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>14</td>\n",
       "      <td>61125535</td>\n",
       "      <td>SIX1</td>\n",
       "      <td>TATTAAAAAA,AGAACTTCAA</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>MEF2A</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ZNF384</td>\n",
       "      <td>MEF2A,0.08685607080506892,ATTAAAAAATAGAAC,.,4....</td>\n",
       "      <td>.</td>\n",
       "      <td>30920</td>\n",
       "      <td>145</td>\n",
       "      <td>699</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>84</th>\n",
       "      <td>69.12</td>\n",
       "      <td>a762dd44-a8ac-4269-8163-bbf6d0f8d4b3</td>\n",
       "      <td>LUSC-US</td>\n",
       "      <td>19</td>\n",
       "      <td>9731577</td>\n",
       "      <td>ZNF561</td>\n",
       "      <td>AGTAGCCGGG,GTGGTGGCAC</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>SP8,KLF2,KLF3,KLF6</td>\n",
       "      <td>1.0</td>\n",
       "      <td>SP3</td>\n",
       "      <td>SP3,0.10810810810810811,CACCACGCCCGGC,.,16.347...</td>\n",
       "      <td>.</td>\n",
       "      <td>22038</td>\n",
       "      <td>116</td>\n",
       "      <td>460</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>85</th>\n",
       "      <td>69.01</td>\n",
       "      <td>0980e7fd-051d-45e9-9ca6-2baf073da4e8</td>\n",
       "      <td>COAD-US</td>\n",
       "      <td>17</td>\n",
       "      <td>66507534</td>\n",
       "      <td>PRKAR1A</td>\n",
       "      <td>CTCTGATGGA,AAAAAAAATG</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>NFATC2,NFAT5,NFATC1,NFATC3,NFATC4</td>\n",
       "      <td>1.0</td>\n",
       "      <td>ZNF384</td>\n",
       "      <td>NFATC2,0.08195243480683811,TTTTCCA,.,2.8420286...</td>\n",
       "      <td>.</td>\n",
       "      <td>411973</td>\n",
       "      <td>1329</td>\n",
       "      <td>6006</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>86</th>\n",
       "      <td>69.00</td>\n",
       "      <td>760881cc-c623-11e3-bf01-24c6515278c0</td>\n",
       "      <td>LIRI-JP</td>\n",
       "      <td>6</td>\n",
       "      <td>150389763</td>\n",
       "      <td>ULBP3</td>\n",
       "      <td>CCCATGTCTT,ACCCCAGCTC</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>NR2F1,NR2C1</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NR2C2</td>\n",
       "      <td>NR2F1,0.0895792462802218,CTGGGGTCAAGAC,.,2.232...</td>\n",
       "      <td>.</td>\n",
       "      <td>184214</td>\n",
       "      <td>647</td>\n",
       "      <td>3397</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>87</th>\n",
       "      <td>69.00</td>\n",
       "      <td>04aa6b77-8074-480c-872e-a1a47afa5314</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>20</td>\n",
       "      <td>55904848</td>\n",
       "      <td>SPO11</td>\n",
       "      <td>GAAAGCGCGG,AAAGGCACGC</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>E2F6,TFDP1</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>E2F6,0.08760709967935801,AAGCGCGGGAAAG,.,4.716...</td>\n",
       "      <td>.</td>\n",
       "      <td>342233</td>\n",
       "      <td>1348</td>\n",
       "      <td>6140</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>69.00</td>\n",
       "      <td>22edd867-dcd6-42b5-92bb-025ceb84b106</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>6</td>\n",
       "      <td>54711199</td>\n",
       "      <td>FAM83B</td>\n",
       "      <td>CGCCTTCTTC,CAGGTGAGCG</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>ZEB1,RBPJ</td>\n",
       "      <td>1.0</td>\n",
       "      <td>EBF1</td>\n",
       "      <td>ZEB1,0.07877799573023264,CTCACCTGGGA,.,4.42931...</td>\n",
       "      <td>.</td>\n",
       "      <td>27192</td>\n",
       "      <td>125</td>\n",
       "      <td>502</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>89</th>\n",
       "      <td>68.96</td>\n",
       "      <td>faff4626-615b-416a-b7a6-9d177dcc94a9</td>\n",
       "      <td>READ-US</td>\n",
       "      <td>8</td>\n",
       "      <td>31497371</td>\n",
       "      <td>NRG1</td>\n",
       "      <td>CCCCCAATAA,TAAATAAAAG</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>HOXA13</td>\n",
       "      <td>10.0</td>\n",
       "      <td>FOXC1,HNF1A,FOXP1,ONECUT1,ONECUT3,FOXB1,FOXC2,...</td>\n",
       "      <td>FOXC1,1.1451876019575855,TAAATAAATAA,TAAGTAAAT...</td>\n",
       "      <td>.</td>\n",
       "      <td>11590</td>\n",
       "      <td>53</td>\n",
       "      <td>199</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>90</th>\n",
       "      <td>68.48</td>\n",
       "      <td>e2774050-1257-4c29-b201-d79ce70fe17e</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>19</td>\n",
       "      <td>2720917</td>\n",
       "      <td>DIRAS1</td>\n",
       "      <td>CCACCCCGGG,AGGGGGCGCG</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>KLF4,ZNF263,KLF5,ZNF740,MAZ</td>\n",
       "      <td>7.0</td>\n",
       "      <td>NRF1,ZIC3,CTCFL,TFDP1,ZNF460,EBF3,ZNF148</td>\n",
       "      <td>KLF4,0.07790346202985261,GCCCCCTCCCCG,.,1.4874...</td>\n",
       "      <td>.</td>\n",
       "      <td>164045</td>\n",
       "      <td>713</td>\n",
       "      <td>3175</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>91</th>\n",
       "      <td>68.25</td>\n",
       "      <td>affa4ef4-02ec-40f1-b6a2-739f8bf2afa9</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>4</td>\n",
       "      <td>126238008</td>\n",
       "      <td>FAT4</td>\n",
       "      <td>CAAGGAAGAC,GTAGCAGCGG</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>OSR1,OSR2</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>OSR1,0.07198957590940831,TGCTACTGTC,.,0.038173...</td>\n",
       "      <td>.</td>\n",
       "      <td>3563</td>\n",
       "      <td>20</td>\n",
       "      <td>71</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>68.25</td>\n",
       "      <td>45a7949d-e63f-4956-866c-df51257032de</td>\n",
       "      <td>BLCA-US</td>\n",
       "      <td>1</td>\n",
       "      <td>149890174</td>\n",
       "      <td>SV2A</td>\n",
       "      <td>TTCCCGGGCA,GAAGAAGTAA</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>E2F6,ELF3,ETV1,ETV4</td>\n",
       "      <td>2.0</td>\n",
       "      <td>GABPA,ELF1</td>\n",
       "      <td>GABPA,0.09345794392523366,ACTTCTTCCTGCCC,.,11....</td>\n",
       "      <td>.</td>\n",
       "      <td>11050</td>\n",
       "      <td>76</td>\n",
       "      <td>284</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>68.18</td>\n",
       "      <td>fa4d35e3-fc38-4e26-a9e0-521a648acec1</td>\n",
       "      <td>SKCM-US</td>\n",
       "      <td>17</td>\n",
       "      <td>37025887</td>\n",
       "      <td>LASP1</td>\n",
       "      <td>TGGGCCGGAA,AGAGCTAAGA</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>ZBTB7A,ETV5</td>\n",
       "      <td>5.0</td>\n",
       "      <td>ELK1,ELK4,ELF1,ELK3,TFDP1</td>\n",
       "      <td>ELK1,0.09453761651761236,GCCGGAAGAG,.,26.41848...</td>\n",
       "      <td>.</td>\n",
       "      <td>42973</td>\n",
       "      <td>163</td>\n",
       "      <td>857</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>98</th>\n",
       "      <td>68.11</td>\n",
       "      <td>760881cc-c623-11e3-bf01-24c6515278c0</td>\n",
       "      <td>LIRI-JP</td>\n",
       "      <td>2</td>\n",
       "      <td>47629535</td>\n",
       "      <td>MSH2</td>\n",
       "      <td>CGTGAGCTAC,GCGCCCTGCC</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>+</td>\n",
       "      <td>...</td>\n",
       "      <td>KLF2,KLF3,OSR1</td>\n",
       "      <td>2.0</td>\n",
       "      <td>CTCFL,KLF15</td>\n",
       "      <td>CTCFL,0.7975471818008515,GGCAGGGCGCGG,GGCAGGGC...</td>\n",
       "      <td>.</td>\n",
       "      <td>184214</td>\n",
       "      <td>647</td>\n",
       "      <td>3397</td>\n",
       "      <td>0.0</td>\n",
       "      <td>not_available</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>47 rows × 29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       score                                   pid   cohort #CHROM        POS  \\\n",
       "4      86.30  9988eb07-01f6-4f83-8699-bb63e0525f08  HNSC-US      7  145813786   \n",
       "7      82.68  04c09376-8756-41d9-bbca-f464170a8cfb  SKCM-US     12   54332280   \n",
       "9      82.00  5ead73fe-1c34-48ed-b20d-89fc3c82dbd6  LIHC-US      5  170736726   \n",
       "13     80.72  199bbb0f-996c-40c1-b06d-2066f04be778  LUAD-US     15   57210473   \n",
       "14     80.13  3869ff3f-21b9-4817-8ff4-83c6fc75ab11  SKCM-US      3  169381814   \n",
       "17     77.52  dcf45005-7e6d-4c26-9994-f973da7e814f  LIHC-US     22   29999735   \n",
       "19     77.00  04c09376-8756-41d9-bbca-f464170a8cfb  SKCM-US      1  190446811   \n",
       "20     76.92  66d312fc-809d-428b-98e3-a29d211db35c  SKCM-US      1   14026387   \n",
       "25     74.79  0c7aca3f-e006-4de3-afc2-20b4f727d4fd  BLCA-US     11  113929845   \n",
       "26     74.72  2399ab13-abfa-480e-9fda-7947edc420be  SKCM-US      1   35658792   \n",
       "27     74.02  dfa20d2d-1ba3-476a-b4b9-21d21470bee2  HNSC-US     20   43976570   \n",
       "28     74.00  3b7810f7-f8ff-4d62-b766-3ba06170194c  RECA-EU      3  187463635   \n",
       "31     73.46  b73523d7-f5a5-4140-8537-4df4d1ecf465  BLCA-US     19   13048717   \n",
       "32     73.42  04aa6b77-8074-480c-872e-a1a47afa5314  SKCM-US      3   71633880   \n",
       "33     73.34  448fe471-3f4e-4dc8-a4e0-6f147dc93abe  BLCA-US      6   37138454   \n",
       "36     73.03  9321341c-c622-11e3-bf01-24c6515278c0  LIRI-JP      3   30647905   \n",
       "37     73.00  3869ff3f-21b9-4817-8ff4-83c6fc75ab11  SKCM-US      8   68658276   \n",
       "39     72.69  75ba6722-1148-4a52-a9ed-68d890238205  SKCM-US     12   95945540   \n",
       "40     72.54  14c5b81d-da49-4db1-9834-77711c2b1d38  READ-US      4  106067000   \n",
       "43     72.26  0bbc7ede-5022-4084-925c-d65baaf7abc2  LUSC-US     22   41486807   \n",
       "44     72.18  49dec0c2-8e75-4f44-a253-82b2ea605890  LUAD-US      6  127439062   \n",
       "49     71.45  04c09376-8756-41d9-bbca-f464170a8cfb  SKCM-US     10   88126034   \n",
       "59     71.00  2290b078-6a5b-4c83-9dfb-b525bbf14e4e  BRCA-US     20   42876352   \n",
       "58     71.00  760881cc-c623-11e3-bf01-24c6515278c0  LIRI-JP      1   87793862   \n",
       "57     71.00  14630c19-9ee0-4ab3-9d8b-8c3309d1e7d0  SKCM-US     15   75871490   \n",
       "56     71.00  4c2ad4a0-5d57-4e27-9f35-058b2f205f50  LUSC-US      5  157097791   \n",
       "63     70.87  448fe471-3f4e-4dc8-a4e0-6f147dc93abe  BLCA-US      4   40058235   \n",
       "65     70.68  04c09376-8756-41d9-bbca-f464170a8cfb  SKCM-US     12   54332288   \n",
       "66     70.63  2df02f2b-9f1c-4249-b3b4-b03079cd97d9  COAD-US     10   23216440   \n",
       "67     70.49  987528ac-437a-4eb8-a335-4f2076d5c006  BRCA-US      9  123837465   \n",
       "69     70.44  04aa6b77-8074-480c-872e-a1a47afa5314  SKCM-US     12   81332010   \n",
       "18414  69.75  3c92c0e9-68e0-4354-8ed1-b6506878145d  EOPC-DE      6  135501964   \n",
       "78     69.72  f5c8f98a-c34a-49e8-a655-d8f29487b55c  CLLE-ES     12   92539203   \n",
       "79     69.70  2df02f2b-9f1c-4249-b3b4-b03079cd97d9  COAD-US      4   55523332   \n",
       "81     69.55  760881cc-c623-11e3-bf01-24c6515278c0  LIRI-JP     19   55582410   \n",
       "83     69.19  94108975-b7a0-40ba-ad39-e44cc62e8cc0  BLCA-US     14   61125535   \n",
       "84     69.12  a762dd44-a8ac-4269-8163-bbf6d0f8d4b3  LUSC-US     19    9731577   \n",
       "85     69.01  0980e7fd-051d-45e9-9ca6-2baf073da4e8  COAD-US     17   66507534   \n",
       "86     69.00  760881cc-c623-11e3-bf01-24c6515278c0  LIRI-JP      6  150389763   \n",
       "87     69.00  04aa6b77-8074-480c-872e-a1a47afa5314  SKCM-US     20   55904848   \n",
       "88     69.00  22edd867-dcd6-42b5-92bb-025ceb84b106  SKCM-US      6   54711199   \n",
       "89     68.96  faff4626-615b-416a-b7a6-9d177dcc94a9  READ-US      8   31497371   \n",
       "90     68.48  e2774050-1257-4c29-b201-d79ce70fe17e  SKCM-US     19    2720917   \n",
       "91     68.25  affa4ef4-02ec-40f1-b6a2-739f8bf2afa9  SKCM-US      4  126238008   \n",
       "92     68.25  45a7949d-e63f-4956-866c-df51257032de  BLCA-US      1  149890174   \n",
       "94     68.18  fa4d35e3-fc38-4e26-a9e0-521a648acec1  SKCM-US     17   37025887   \n",
       "98     68.11  760881cc-c623-11e3-bf01-24c6515278c0  LIRI-JP      2   47629535   \n",
       "\n",
       "          GENE       SEQUENCE_CONTEXT REF ALT strand  ...  \\\n",
       "4      CNTNAP2  GCTGAGACTC,CCTGCCGCTC   C   A      +  ...   \n",
       "7       HOXC13  CCCTCCCTCT,CCTCCCTGGG   C   T      +  ...   \n",
       "9         TLX3  CGGTCTCCAG,CTTGGCGGTC   C   A      +  ...   \n",
       "13       TCF12  GGGCAGCGGA,CCAGATGCTT   T   C      +  ...   \n",
       "14       MECOM  AGCCAGGGGG,AAAAAAACTT   G   A      -  ...   \n",
       "17         NF2  GCCGGGTCGC,CCTGCACCGA   G   A      +  ...   \n",
       "19      BRINP3  GGGGGGTGGG,AATGGTGGGG   G   A      -  ...   \n",
       "20       PRDM2  GATATTTAAT,ACCTCCAACG   G   T      +  ...   \n",
       "25      ZBTB16  GGACTGCAGC,CCGAGCCATG   C   A      +  ...   \n",
       "26        SFPQ  GCTGGCACGG,CCCGCCTACT   C   T      -  ...   \n",
       "27        SDC4  AGGGCCATCT,GGATCCTAAA   C   G      -  ...   \n",
       "28        BCL6  TAAGAGGGGG,GTGCGCTTTG   A   G      -  ...   \n",
       "31        CALR  CCCCCTCTGT,TGCTCAGGAA   C   T      +  ...   \n",
       "32       FOXP1  TGCAAGGAAA,AGAAACCCCG   C   T      -  ...   \n",
       "33        PIM1  CCCCCGCCTC,GGCCCGGGGA   C   T      +  ...   \n",
       "36      TGFBR2  GGGCGGTGAG,GGCAGCTGAA   G   A      +  ...   \n",
       "37        CPA6  ATAAAGGTGG,TGTGCCCCGG   C   G      -  ...   \n",
       "39       USP44  CTTCCCTGAT,CCCAGATGGA   C   T      -  ...   \n",
       "40        TET2  TCTGTGAGAC,CCCCCTTCTA   G   A      +  ...   \n",
       "43       EP300  TTTTTCCCAT,ATTACATTAG   A   G      +  ...   \n",
       "44       RSPO3  AAGGAAGAAA,AAAGAAAAGA   G   C      +  ...   \n",
       "49       GRID1  CGTTGGGGCC,CCGCCCAGCC   C   T      -  ...   \n",
       "59     GDAP1L1  GGGCCCCAGG,GCACAGTTCT   A   T      +  ...   \n",
       "58        LMO4  CAGGTGAGGC,CCGCCCTGCG   C   T      +  ...   \n",
       "57       PTPN9  GAAATTTCCC,AAGGCCGCCG   G   A      -  ...   \n",
       "56     C5orf52  TTACACAAGA,GGAAGTGAGA   G   C      +  ...   \n",
       "63       N4BP2  GGAGAGGTGG,AGCGCCCGCG   G   C      +  ...   \n",
       "65      HOXC13  CTCCCTCCCT,GGGCCTGCCG   G   A      +  ...   \n",
       "66       ARMC3  AAGTGGCTGA,GTGATGACTG   A   C      +  ...   \n",
       "67       CNTRL  ATGGATTCGG,GCGGGGGCGC   G   T      +  ...   \n",
       "69       LIN7A  AAATGTTTGG,ATGTAATTAC   G   A      -  ...   \n",
       "18414      MYB  AGTTGTAAAC,TTGACGAAAA   C   A      +  ...   \n",
       "78        BTG1  AAGGTCTGCA,CTGTCGCTCG   G   T      -  ...   \n",
       "79         KIT  CAGAGGCAGA,AAATCATTAC   G   T      +  ...   \n",
       "81      EPS8L1  CAAGCCCCAG,GGAGAATTCT   G   T      +  ...   \n",
       "83        SIX1  TATTAAAAAA,AGAACTTCAA   T   C      -  ...   \n",
       "84      ZNF561  AGTAGCCGGG,GTGGTGGCAC   C   A      -  ...   \n",
       "85     PRKAR1A  CTCTGATGGA,AAAAAAAATG   A   C      +  ...   \n",
       "86       ULBP3  CCCATGTCTT,ACCCCAGCTC   G   A      -  ...   \n",
       "87       SPO11  GAAAGCGCGG,AAAGGCACGC   G   A      +  ...   \n",
       "88      FAM83B  CGCCTTCTTC,CAGGTGAGCG   C   T      +  ...   \n",
       "89        NRG1  CCCCCAATAA,TAAATAAAAG   A   G      +  ...   \n",
       "90      DIRAS1  CCACCCCGGG,AGGGGGCGCG   G   A      -  ...   \n",
       "91        FAT4  CAAGGAAGAC,GTAGCAGCGG   A   T      +  ...   \n",
       "92        SV2A  TTCCCGGGCA,GAAGAAGTAA   G   A      -  ...   \n",
       "94       LASP1  TGGGCCGGAA,AGAGCTAAGA   G   A      +  ...   \n",
       "98        MSH2  CGTGAGCTAC,GCGCCCTGCC   C   A      +  ...   \n",
       "\n",
       "           destroyed_tfs_passing_tf_expression_threshold  num_remaining_tfs  \\\n",
       "4                                                 ZNF148                2.0   \n",
       "7                                                   E2F6                4.0   \n",
       "9                                                    NaN                1.0   \n",
       "13                                                   NaN                3.0   \n",
       "14                                                   NaN                1.0   \n",
       "17                                                  PAX5                8.0   \n",
       "19                      MZF1,KLF16,RELB,KLF10,SP9,ZNF148                9.0   \n",
       "20     RORA,HOXC9,HOXC11,POU6F2,HOXC10,HOXD11,NR1H4,H...               21.0   \n",
       "25                                                TFAP2A                2.0   \n",
       "26                              SP1,EGR1,KLF5,KLF16,KLF2                8.0   \n",
       "27                                                   NaN                2.0   \n",
       "28                                               MSANTD3                4.0   \n",
       "31                                                   NaN                1.0   \n",
       "32                                            IRF9,PRDM4                4.0   \n",
       "33                                   ZBTB7A,KLF11,TFAP2E                8.0   \n",
       "36                                                ZNF148                2.0   \n",
       "37                                                ZNF341                7.0   \n",
       "39                                                  MZF1                3.0   \n",
       "40                                       CTCFL,KLF11,SP9                6.0   \n",
       "43                                           STAT1,PROP1                4.0   \n",
       "44                                                   NaN                0.0   \n",
       "49            KLF4,SP3,KLF15,KLF2,KLF6,PLAGL2,SP9,ZNF148                5.0   \n",
       "59                                                   NaN                4.0   \n",
       "58                  KLF4,SP3,KLF15,KLF2,KLF3,KLF6,ZNF148                6.0   \n",
       "57                                        HSF1,HSF2,HSF4                0.0   \n",
       "56                                                   NaN                8.0   \n",
       "63                                                 TBX18                3.0   \n",
       "65                                  TFAP2A,TFAP2C,TFAP2E                3.0   \n",
       "66                                                   NaN                8.0   \n",
       "67                    KLF4,E2F6,KLF5,CTCFL,KLF2,KLF6,MAZ                9.0   \n",
       "69                                     HOXA5,HOXB5,HOXB8                3.0   \n",
       "18414                                                NaN                2.0   \n",
       "78                                                 ASCL1                8.0   \n",
       "79                                                   NaN                1.0   \n",
       "81                                  TFAP2A,TFAP2B,TFAP2E                6.0   \n",
       "83                                                 MEF2A                1.0   \n",
       "84                                    SP8,KLF2,KLF3,KLF6                1.0   \n",
       "85                     NFATC2,NFAT5,NFATC1,NFATC3,NFATC4                1.0   \n",
       "86                                           NR2F1,NR2C1                1.0   \n",
       "87                                            E2F6,TFDP1                0.0   \n",
       "88                                             ZEB1,RBPJ                1.0   \n",
       "89                                                HOXA13               10.0   \n",
       "90                           KLF4,ZNF263,KLF5,ZNF740,MAZ                7.0   \n",
       "91                                             OSR1,OSR2                0.0   \n",
       "92                                   E2F6,ELF3,ETV1,ETV4                2.0   \n",
       "94                                           ZBTB7A,ETV5                5.0   \n",
       "98                                        KLF2,KLF3,OSR1                2.0   \n",
       "\n",
       "                                           remaining_tfs  \\\n",
       "4                                            NHLH2,SNAI3   \n",
       "7                               GABPA,KLF5,ZNF460,ZNF148   \n",
       "9                                                   NFIC   \n",
       "13                                   ZBTB18,TWIST1,HAND2   \n",
       "14                                                 INSM1   \n",
       "17           TCF3,TBX4,TBX5,TCF4,ASCL1,NHLH2,ASCL1,TCF12   \n",
       "19         KLF4,RREB1,SP1,KLF5,SP3,SP8,ZNF740,KLF11,KLF2   \n",
       "20     PDX1,ALX3,NKX6-1,NKX6-2,MEOX2,VAX1,VAX2,HOXD12...   \n",
       "25                                         TFAP2C,TFAP2B   \n",
       "26                PAX5,NFIX,EGR3,SP3,SP8,KLF10,KLF11,SP9   \n",
       "27                                         PLAG1,NEUROD1   \n",
       "28                               EGR1,KLF16,ZNF740,KLF10   \n",
       "31                                                 FOXD2   \n",
       "32                                 STAT1,STAT2,IRF8,IRF4   \n",
       "33      TFAP2A,KLF16,TFAP2A,TFAP2B,TFAP2C,MAZ,SP9,ZNF148   \n",
       "36                                           PKNOX1,PBX3   \n",
       "37                    SP1,ESR2,KLF13,NFKB2,KLF9,KLF3,SP9   \n",
       "39                                    STAT3,RHOXF1,HAND2   \n",
       "40                        KLF13,SP4,KLF16,SP3,KLF10,KLF6   \n",
       "43                               MEOX2,PHOX2A,FOXB1,DUXA   \n",
       "44                                                   NaN   \n",
       "49                             KLF5,SP8,KLF11,MAZ,ZBTB14   \n",
       "59                                EBF1,MYOD1,SNAI2,NHLH2   \n",
       "58                         SP1,NFKB1,ZNF263,KLF5,MAZ,SP9   \n",
       "57                                                   NaN   \n",
       "56                 ELK1,SPIB,ELF1,ERG,EHF,ELF3,ETV3,ELF2   \n",
       "63                                       E2F6,TBX15,KLF9   \n",
       "65                                        EBF1,KLF5,EBF3   \n",
       "66               FOSB,JUN,JUN,JUNB,JUNB,FOSL2,JUND,BACH2   \n",
       "67      EGR2,KLF16,SP3,SP8,ZNF740,KLF11,KLF15,SP9,ZNF148   \n",
       "69                                     HOXA1,HOXA7,HOXC8   \n",
       "18414                                        FOXG1,FOXK1   \n",
       "78     MYOD1,MYOG,SNAI2,BHLHA15,NHLH2,ASCL1,BHLHE22,MYF5   \n",
       "79                                                 BARX2   \n",
       "81                 TFAP2A,EBF1,TFAP2C,TFAP2B,TFAP2C,EBF3   \n",
       "83                                                ZNF384   \n",
       "84                                                   SP3   \n",
       "85                                                ZNF384   \n",
       "86                                                 NR2C2   \n",
       "87                                                   NaN   \n",
       "88                                                  EBF1   \n",
       "89     FOXC1,HNF1A,FOXP1,ONECUT1,ONECUT3,FOXB1,FOXC2,...   \n",
       "90              NRF1,ZIC3,CTCFL,TFDP1,ZNF460,EBF3,ZNF148   \n",
       "91                                                   NaN   \n",
       "92                                            GABPA,ELF1   \n",
       "94                             ELK1,ELK4,ELF1,ELK3,TFDP1   \n",
       "98                                           CTCFL,KLF15   \n",
       "\n",
       "       JASPAR2020_CORE_vertebrates_non_redundant(tf_name,binding_affinity,seq1,seq2,raw,zscore,log,tf_sequence_logo)  \\\n",
       "4      ZEB1,14.5102,.,CTCACCTGCCG,0.832158029939,-0.7...                                                               \n",
       "7      GABPA,10.4714,.,CCCTCTTCCTCCCT,5.56252436512,-...                                                               \n",
       "9      NFIC,9.54472,.,GACTTGGCGGT,27.9809277799,0.238...                                                               \n",
       "13     HINFP,13.0606,.,CTGGGTCCGCTG,2.8152864897,0.62...                                                               \n",
       "14     INSM1,12.0408,.,AGCCAGGGGGAA,0.0,-0.3815070014...                                                               \n",
       "17     PAX5,0.07818058150716525,AGGCGCGACCCG,.,0.0093...                                                               \n",
       "19     KLF4,0.7129713823848856,TTCCCCACCCCC,TTTCCCACC...                                                               \n",
       "20     RORA,0.08166665305555783,TTGGAGGTCA,.,0.463684...                                                               \n",
       "25     TFAP2C,0.09271535458987362,TGGCTCGGGGCT,.,16.6...                                                               \n",
       "26     PAX5,10.6909,.,AGGCGGGACCGT,0.0041284764193,-0...                                                               \n",
       "27     PLAG1,6.08163,.,AGGGCCATCTGGGA,1.6972677727,0....                                                               \n",
       "28     EGR1,10.2114,.,GCGCACCCCCCCTC,4.70543294151,-0...                                                               \n",
       "31     FOXD2,10.7857,.,CCTGAGCAAACAG,0.417828374569,-...                                                               \n",
       "32     STAT1,0.9441198356465754,GGGGTTTCTGTTTCC,GGGGT...                                                               \n",
       "33     TFAP2A,0.9195551902087045,CCCGCCTCCGGCCC,CCCGC...                                                               \n",
       "36     SMAD2,11.3443,.,CTGCCTCTCACCG,7.06965725325,1....                                                               \n",
       "37     SP1,3.38298,.,CGGGGCACACCCACC,13.0141602848,-1...                                                               \n",
       "39     MZF1,0.08081461128171973,CTGATCCCCAGAT,.,2.029...                                                               \n",
       "40     KLF13,0.4498060359495164,GTGAGACGCCCCCTTCTA,GT...                                                               \n",
       "43     STAT1,0.08986906077844581,ATTATGGGAAA,.,81.441...                                                               \n",
       "44     EWSR1-FLI1,12.602921305366829,GGAAGAAAGAAAGAAA...                                                               \n",
       "49     KLF4,0.07352941176470588,GCCCCCGCCCAG,.,6.2973...                                                               \n",
       "59     ZEB1,12.7143,.,TGCACCTGGGG,0.433314718291,-1.3...                                                               \n",
       "58     KLF4,0.07198957590940831,GGCCCCGCCCTG,.,50.971...                                                               \n",
       "57     HSF1,0.08079502302658156,TTCGGGAAATTTC,.,44.11...                                                               \n",
       "56     ELK1,10.6889,.,GACGGAAGTG,18.4019310658,0.0856...                                                               \n",
       "63     E2F6,0.0970798384591488,AGAGGTGGGAGCG,.,3.2812...                                                               \n",
       "65     TFAP2A,0.0863155350700019,CAGGCCCCAGGGAG,.,10....                                                               \n",
       "66     SREBF1,11.2073,.,ATCACGTCAG,33.4166268472,0.63...                                                               \n",
       "67     KLF4,0.07115158845921235,GCCCCCGCCCCG,.,1.9048...                                                               \n",
       "69     TEAD1,13.439,.,TTACATTCCAAAC,13.2052531214,-0....                                                               \n",
       "18414  FOXD1,13.52,.,GTAAACAT,5.62,4.407493119305658,...                                                               \n",
       "78     MYOD1,0.809704733676805,CTGCAGCTGTCGC,CTGCATCT...                                                               \n",
       "79     PBX2,12.3465,.,CAGATAAATCATT,10.724192169,0.04...                                                               \n",
       "81     TFAP2A,0.09213447948625815,CAAGCCCCAGGGGA,.,2....                                                               \n",
       "83     MEF2A,0.08685607080506892,ATTAAAAAATAGAAC,.,4....                                                               \n",
       "84     SP3,0.10810810810810811,CACCACGCCCGGC,.,16.347...                                                               \n",
       "85     NFATC2,0.08195243480683811,TTTTCCA,.,2.8420286...                                                               \n",
       "86     NR2F1,0.0895792462802218,CTGGGGTCAAGAC,.,2.232...                                                               \n",
       "87     E2F6,0.08760709967935801,AAGCGCGGGAAAG,.,4.716...                                                               \n",
       "88     ZEB1,0.07877799573023264,CTCACCTGGGA,.,4.42931...                                                               \n",
       "89     FOXC1,1.1451876019575855,TAAATAAATAA,TAAGTAAAT...                                                               \n",
       "90     KLF4,0.07790346202985261,GCCCCCTCCCCG,.,1.4874...                                                               \n",
       "91     OSR1,0.07198957590940831,TGCTACTGTC,.,0.038173...                                                               \n",
       "92     GABPA,0.09345794392523366,ACTTCTTCCTGCCC,.,11....                                                               \n",
       "94     ELK1,0.09453761651761236,GCCGGAAGAG,.,26.41848...                                                               \n",
       "98     CTCFL,0.7975471818008515,GGCAGGGCGCGG,GGCAGGGC...                                                               \n",
       "\n",
       "      paths_with_recurrence(format=path,pid,cohort,bp,ref,alt,gene,chr,raw_score,zscore,log_score,confidence,purity,af,(tfs_seperated_by_//))  \\\n",
       "4                                                      .                                                                                        \n",
       "7                                                      .                                                                                        \n",
       "9                                                      .                                                                                        \n",
       "13                                                     .                                                                                        \n",
       "14                                                     .                                                                                        \n",
       "17                                                     .                                                                                        \n",
       "19                                                     .                                                                                        \n",
       "20                                                     .                                                                                        \n",
       "25                                                     .                                                                                        \n",
       "26                                                     .                                                                                        \n",
       "27                                                     .                                                                                        \n",
       "28                                                     .                                                                                        \n",
       "31                                                     .                                                                                        \n",
       "32                                                     .                                                                                        \n",
       "33                                                     .                                                                                        \n",
       "36                                                     .                                                                                        \n",
       "37                                                     .                                                                                        \n",
       "39                                                     .                                                                                        \n",
       "40                                                     .                                                                                        \n",
       "43                                                     .                                                                                        \n",
       "44                                                     .                                                                                        \n",
       "49                                                     .                                                                                        \n",
       "59                                                     .                                                                                        \n",
       "58                                                     .                                                                                        \n",
       "57                                                     .                                                                                        \n",
       "56                                                     .                                                                                        \n",
       "63                                                     .                                                                                        \n",
       "65                                                     .                                                                                        \n",
       "66                                                     .                                                                                        \n",
       "67                                                     .                                                                                        \n",
       "69                                                     .                                                                                        \n",
       "18414                                                  .                                                                                        \n",
       "78                                                     .                                                                                        \n",
       "79                                                     .                                                                                        \n",
       "81                                                     .                                                                                        \n",
       "83                                                     .                                                                                        \n",
       "84                                                     .                                                                                        \n",
       "85                                                     .                                                                                        \n",
       "86                                                     .                                                                                        \n",
       "87                                                     .                                                                                        \n",
       "88                                                     .                                                                                        \n",
       "89                                                     .                                                                                        \n",
       "90                                                     .                                                                                        \n",
       "91                                                     .                                                                                        \n",
       "92                                                     .                                                                                        \n",
       "94                                                     .                                                                                        \n",
       "98                                                     .                                                                                        \n",
       "\n",
       "      num_original_mutations  num_final_mutations num_promoter_mutations  \\\n",
       "4                      23779                   92                    459   \n",
       "7                      32589                  176                    629   \n",
       "9                       2157                    9                     42   \n",
       "13                     17648                   68                    412   \n",
       "14                    171882                  809                   3472   \n",
       "17                      4582                   19                    106   \n",
       "19                     32589                  176                    629   \n",
       "20                     40281                  168                    742   \n",
       "25                      7978                   37                    170   \n",
       "26                     26144                  143                    532   \n",
       "27                     28551                  160                    726   \n",
       "28                     21836                  111                    490   \n",
       "31                     30204                  174                   1034   \n",
       "32                    342233                 1348                   6140   \n",
       "33                     19009                   95                    488   \n",
       "36                      3410                   13                     49   \n",
       "37                    171882                  809                   3472   \n",
       "39                    127544                  641                   2368   \n",
       "40                    548085                 1685                   7490   \n",
       "43                     17905                   92                    304   \n",
       "44                     22203                   91                    436   \n",
       "49                     32589                  176                    629   \n",
       "59                      3408                   24                    102   \n",
       "58                    184214                  647                   3397   \n",
       "57                     24354                  106                    472   \n",
       "56                     28620                  209                    668   \n",
       "63                     19009                   95                    488   \n",
       "65                     32589                  176                    629   \n",
       "66                    875975                 2705                  11145   \n",
       "67                      8630                   44                    212   \n",
       "69                    342233                 1348                   6140   \n",
       "18414                   1560                    1                     13   \n",
       "78                      1015                    8                     26   \n",
       "79                    875975                 2705                  11145   \n",
       "81                    184214                  647                   3397   \n",
       "83                     30920                  145                    699   \n",
       "84                     22038                  116                    460   \n",
       "85                    411973                 1329                   6006   \n",
       "86                    184214                  647                   3397   \n",
       "87                    342233                 1348                   6140   \n",
       "88                     27192                  125                    502   \n",
       "89                     11590                   53                    199   \n",
       "90                    164045                  713                   3175   \n",
       "91                      3563                   20                     71   \n",
       "92                     11050                   76                    284   \n",
       "94                     42973                  163                    857   \n",
       "98                    184214                  647                   3397   \n",
       "\n",
       "       num_recurrent_mutations       cn_score  \n",
       "4                          0.0              9  \n",
       "7                          0.0              3  \n",
       "9                          0.0              4  \n",
       "13                         0.0  not_available  \n",
       "14                         0.0              3  \n",
       "17                         0.0              6  \n",
       "19                         0.0              3  \n",
       "20                         0.0  not_available  \n",
       "25                         0.0              3  \n",
       "26                         0.0  not_available  \n",
       "27                         0.0  not_available  \n",
       "28                         0.0  not_available  \n",
       "31                         0.0              3  \n",
       "32                         0.0              4  \n",
       "33                         0.0              4  \n",
       "36                         0.0  not_available  \n",
       "37                         0.0  not_available  \n",
       "39                         0.0              3  \n",
       "40                         0.0  not_available  \n",
       "43                         0.0              8  \n",
       "44                         0.0              4  \n",
       "49                         0.0  not_available  \n",
       "59                         0.0              3  \n",
       "58                         0.0  not_available  \n",
       "57                         0.0             49  \n",
       "56                         0.0              1  \n",
       "63                         0.0              4  \n",
       "65                         0.0              3  \n",
       "66                         0.0  not_available  \n",
       "67                         0.0  not_available  \n",
       "69                         0.0              3  \n",
       "18414                      0.0            NaN  \n",
       "78                         0.0  not_available  \n",
       "79                         0.0  not_available  \n",
       "81                         0.0  not_available  \n",
       "83                         0.0              4  \n",
       "84                         0.0              5  \n",
       "85                         0.0  not_available  \n",
       "86                         0.0  not_available  \n",
       "87                         0.0              3  \n",
       "88                         0.0              3  \n",
       "89                         0.0  not_available  \n",
       "90                         0.0              5  \n",
       "91                         0.0  not_available  \n",
       "92                         0.0              4  \n",
       "94                         0.0  not_available  \n",
       "98                         0.0  not_available  \n",
       "\n",
       "[47 rows x 29 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "top_100[top_100[\"num_recurrent_mutations\"] == 0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'num'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m~/miniconda3/envs/psnv-hunter/lib/python3.11/site-packages/pandas/core/indexes/base.py:3805\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3804\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 3805\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_engine\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mcasted_key\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   3806\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[0;32mindex.pyx:167\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mindex.pyx:196\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7081\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "File \u001b[0;32mpandas/_libs/hashtable_class_helper.pxi:7089\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'num'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[45], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m top_100[\u001b[43mtop_100\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mnum\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m]\n",
      "File \u001b[0;32m~/miniconda3/envs/psnv-hunter/lib/python3.11/site-packages/pandas/core/frame.py:4102\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   4100\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[1;32m   4101\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[0;32m-> 4102\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcolumns\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget_loc\u001b[49m\u001b[43m(\u001b[49m\u001b[43mkey\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   4103\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[1;32m   4104\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[0;32m~/miniconda3/envs/psnv-hunter/lib/python3.11/site-packages/pandas/core/indexes/base.py:3812\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   3807\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[1;32m   3808\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[1;32m   3809\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[1;32m   3810\u001b[0m     ):\n\u001b[1;32m   3811\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[0;32m-> 3812\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01merr\u001b[39;00m\n\u001b[1;32m   3813\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[1;32m   3814\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[1;32m   3815\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[1;32m   3816\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[1;32m   3817\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[0;31mKeyError\u001b[0m: 'num'"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly.graph_objects as go\n",
    "from plotly.offline import init_notebook_mode"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "def af_greater_than_25_scatterplot(\n",
    "    path_to_overview_file: str, \n",
    "    only_relevant: bool = True\n",
    "):\n",
    "    data = pd.read_csv(path_to_overview_file, delimiter=\"\\t\")\n",
    "\n",
    "    gene_name = path_to_overview_file.split(\"/\")[-5]\n",
    "    position = path_to_overview_file.split(\"/\")[-4]\n",
    "\n",
    "    cohorts = {}\n",
    "    for idx, row in data.iterrows():\n",
    "        cohort_file = row[\"Cohort_File\"].replace(\".txt\", \"\")\n",
    "        af = float(row[\"SNPs_AF>25_%\"])\n",
    "\n",
    "        cohort = cohort_file.split(\"_\")[2]\n",
    "        if cohort not in cohorts:\n",
    "            cohorts[cohort] = {\"tumor\": -1, \"control\": -1}\n",
    "        if \"control\" in cohort_file:\n",
    "            cohorts[cohort][\"control\"] = float(af)\n",
    "        else:\n",
    "            cohorts[cohort][\"tumor\"] = float(af)\n",
    "\n",
    "    num_original_cohorts = len(cohorts.keys())\n",
    "\n",
    "    if only_relevant:\n",
    "        cohorts_to_remove = []\n",
    "        for cohort in cohorts:\n",
    "            control_value = cohorts[cohort][\"control\"]\n",
    "            tumor_value = cohorts[cohort][\"tumor\"]\n",
    "            if control_value == 0 and tumor_value == 0:\n",
    "                cohorts_to_remove.append(cohort)\n",
    "        for cohort in cohorts_to_remove:\n",
    "            cohorts.pop(cohort, None)\n",
    "\n",
    "    x_axis = list(cohorts.keys())\n",
    "\n",
    "    num_current_cohorts = len(x_axis)\n",
    "\n",
    "    if len(x_axis) == 0:\n",
    "        return af_greater_than_25_scatterplot(path_to_overview_file, False)\n",
    "\n",
    "    else:\n",
    "        fig = go.Figure()\n",
    "\n",
    "        fig.add_trace(\n",
    "            go.Scatter(\n",
    "                x=x_axis,\n",
    "                y=[cohorts[cohort][\"tumor\"] for cohort in x_axis],\n",
    "                mode=\"markers\",\n",
    "                marker_color=\"red\",\n",
    "                marker={\"size\": 10},\n",
    "                name=\"Tumor\",\n",
    "            )\n",
    "        )\n",
    "        fig.add_trace(\n",
    "            go.Scatter(\n",
    "                x=x_axis,\n",
    "                y=[cohorts[cohort][\"control\"] for cohort in x_axis],\n",
    "                mode=\"markers\",\n",
    "                marker_color=\"green\",\n",
    "                marker={\"symbol\": \"circle-x-open\", \"size\": 10},\n",
    "                name=\"Control\",\n",
    "            )\n",
    "        )\n",
    "\n",
    "        fig.update_layout(\n",
    "            title=f\"<b>Patients with a minor allele frequency > 25%</b><br><sup>Gene Name: {gene_name} / Position: {position} / Displaying {num_current_cohorts} of {num_original_cohorts} cohorts\",\n",
    "            xaxis_title=\"Cohorts\",\n",
    "            yaxis_title=\"Percent of Patients\",\n",
    "            # yaxis_range=[-2, 100],\n",
    "        )\n",
    "        \n",
    "        fig.update_layout(\n",
    "            plot_bgcolor='white',  # Set plotting area background to white\n",
    "            paper_bgcolor='white',  # Set overall figure background to white\n",
    "            xaxis=dict(\n",
    "                showline=True,  # Show axis lines\n",
    "                linecolor='black',  # Color of axis lines\n",
    "                linewidth=1  # Width of axis lines\n",
    "            ),\n",
    "            yaxis=dict(\n",
    "                showline=True,  # Show axis lines\n",
    "                linecolor='black',  # Color of axis lines\n",
    "                linewidth=1,  # Width of axis lines\n",
    "                ticks='outside',  # Display ticks outside the axis\n",
    "                tickcolor='black',  # Color of ticks\n",
    "                tickwidth=2  # Width of ticks\n",
    "            )\n",
    "        )\n",
    "\n",
    "        fig.update_xaxes(tickangle=45)\n",
    "\n",
    "        return fig\n",
    "\n",
    "\n",
    "def at_least_two_variant_alleles_scatterplot(\n",
    "    path_to_overview_file: str, \n",
    "    only_relevant: bool = True,\n",
    "    ref: str = None,\n",
    "    alt: str = None,\n",
    "):\n",
    "    data = pd.read_csv(path_to_overview_file, delimiter=\"\\t\")\n",
    "\n",
    "    gene_name = path_to_overview_file.split(\"/\")[-5]\n",
    "    position = path_to_overview_file.split(\"/\")[-4]\n",
    "\n",
    "    cohorts = {}\n",
    "\n",
    "    for _, row in data.iterrows():\n",
    "        # Get the maximum.\n",
    "        maximum = np.max(\n",
    "            [row[\"min_2_A\"], row[\"min_2_C\"], row[\"min_2_G\"], row[\"min_2_T\"]]\n",
    "        )\n",
    "\n",
    "        # Get the cohort.\n",
    "        cohort = row[\"Cohort_File\"].split(\"_\")[2].replace(\".txt\", \"\")\n",
    "        tumor_or_control = row[\"Cohort_File\"].split(\"_\")[1]\n",
    "\n",
    "        # Get the counts.\n",
    "        min_2_A = float(row[\"min_2_A\"]) if float(row[\"min_2_A\"]) != maximum else 0\n",
    "        min_2_C = float(row[\"min_2_C\"]) if float(row[\"min_2_C\"]) != maximum else 0\n",
    "        min_2_G = float(row[\"min_2_G\"]) if float(row[\"min_2_G\"]) != maximum else 0\n",
    "        min_2_T = float(row[\"min_2_T\"]) if float(row[\"min_2_T\"]) != maximum else 0\n",
    "\n",
    "        if cohort not in cohorts:\n",
    "            cohorts[cohort] = {\n",
    "                \"tumor\": {\"A\": -1, \"C\": -1, \"G\": -1, \"T\": -1},\n",
    "                \"control\": {\"A\": -1, \"C\": -1, \"G\": -1, \"T\": -1},\n",
    "            }\n",
    "        cohorts[cohort][tumor_or_control][\"A\"] = min_2_A\n",
    "        cohorts[cohort][tumor_or_control][\"C\"] = min_2_C\n",
    "        cohorts[cohort][tumor_or_control][\"G\"] = min_2_G\n",
    "        cohorts[cohort][tumor_or_control][\"T\"] = min_2_T\n",
    "\n",
    "    num_original_cohorts = len(cohorts.keys())\n",
    "\n",
    "    if only_relevant:\n",
    "        cohorts_to_remove = []\n",
    "        for cohort in cohorts:\n",
    "            tumor_sum = np.sum(\n",
    "                [\n",
    "                    cohorts[cohort][\"tumor\"][nucleotide]\n",
    "                    for nucleotide in [\"A\", \"C\", \"G\", \"T\"]\n",
    "                ]\n",
    "            )\n",
    "            control_sum = np.sum(\n",
    "                [\n",
    "                    cohorts[cohort][\"control\"][nucleotide]\n",
    "                    for nucleotide in [\"A\", \"C\", \"G\", \"T\"]\n",
    "                ]\n",
    "            )\n",
    "            if (tumor_sum == 0) and (control_sum == 0):\n",
    "                cohorts_to_remove.append(cohort)\n",
    "\n",
    "        for cohort in cohorts_to_remove:\n",
    "            cohorts.pop(cohort, None)\n",
    "\n",
    "    x_axis = list(cohorts.keys())\n",
    "\n",
    "    num_current_cohorts = len(x_axis)\n",
    "\n",
    "    if len(x_axis) == 0:\n",
    "        return at_least_two_variant_alleles_scatterplot(path_to_overview_file, False, ref, alt)\n",
    "    else:\n",
    "        fig = go.Figure()\n",
    "\n",
    "        color_dict = {\"A\": \"green\", \"C\": \"blue\", \"G\": \"orange\", \"T\": \"red\"}\n",
    "\n",
    "        for nucleotide in [\"A\", \"C\", \"T\", \"G\"]:\n",
    "            if (ref and alt):\n",
    "                fig.add_trace(\n",
    "                    go.Scatter(\n",
    "                        x=[cohort for cohort in cohorts],\n",
    "                        y=[cohorts[cohort][\"tumor\"][nucleotide] for cohort in cohorts],\n",
    "                        name=f\"{nucleotide} (tumor)\",\n",
    "                        mode=\"markers\",\n",
    "                        marker_color=color_dict[nucleotide],\n",
    "                        marker={\"size\": 10},\n",
    "                        visible = True if (nucleotide in [ref, alt]) else \"legendonly\"\n",
    "                    )\n",
    "                )\n",
    "                fig.add_trace(\n",
    "                    go.Scatter(\n",
    "                        x=[cohort for cohort in cohorts],\n",
    "                        y=[cohorts[cohort][\"control\"][nucleotide] for cohort in cohorts],\n",
    "                        name=f\"{nucleotide} (control)\",\n",
    "                        mode=\"markers\",\n",
    "                        marker_color=color_dict[nucleotide],\n",
    "                        marker={\"symbol\": \"circle-x-open\", \"size\": 10},\n",
    "                        visible = True if (nucleotide in [ref, alt]) else \"legendonly\"\n",
    "                    )\n",
    "                )\n",
    "            else:\n",
    "                fig.add_trace(\n",
    "                    go.Scatter(\n",
    "                        x=[cohort for cohort in cohorts],\n",
    "                        y=[cohorts[cohort][\"tumor\"][nucleotide] for cohort in cohorts],\n",
    "                        name=f\"{nucleotide} (tumor)\",\n",
    "                        mode=\"markers\",\n",
    "                        marker_color=color_dict[nucleotide],\n",
    "                        marker={\"size\": 10},\n",
    "                    )\n",
    "                )\n",
    "                fig.add_trace(\n",
    "                    go.Scatter(\n",
    "                        x=[cohort for cohort in cohorts],\n",
    "                        y=[cohorts[cohort][\"control\"][nucleotide] for cohort in cohorts],\n",
    "                        name=f\"{nucleotide} (control)\",\n",
    "                        mode=\"markers\",\n",
    "                        marker_color=color_dict[nucleotide],\n",
    "                        marker={\"symbol\": \"circle-x-open\", \"size\": 10},\n",
    "                    )\n",
    "                )\n",
    "\n",
    "            fig.update_layout(\n",
    "                title=f\"<b>Patients with at least 2 Variant Alleles</b><br><sup>Gene Name: {gene_name} / Position: {position} / Displaying {num_current_cohorts} of {num_original_cohorts} cohorts\",\n",
    "                xaxis_title=\"Cohorts\",\n",
    "                yaxis_title=\"Percent of Affected Patients\",\n",
    "                # yaxis_range=[-2, 100],\n",
    "            )\n",
    "            \n",
    "            fig.update_layout(\n",
    "                plot_bgcolor='white',  # Set plotting area background to white\n",
    "                paper_bgcolor='white',  # Set overall figure background to white\n",
    "                xaxis=dict(\n",
    "                    showline=True,  # Show axis lines\n",
    "                    linecolor='black',  # Color of axis lines\n",
    "                    linewidth=1  # Width of axis lines\n",
    "                ),\n",
    "                yaxis=dict(\n",
    "                    showline=True,  # Show axis lines\n",
    "                    linecolor='black',  # Color of axis lines\n",
    "                    linewidth=1,  # Width of axis lines\n",
    "                    ticks='outside',  # Display ticks outside the axis\n",
    "                    tickcolor='black',  # Color of ticks\n",
    "                    tickwidth=2  # Width of ticks\n",
    "                )\n",
    "            )\n",
    "\n",
    "            fig.update_xaxes(tickangle=45)\n",
    "\n",
    "        return fig\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.plotly.v1+json": {
       "config": {
        "plotlyServerURL": "https://plot.ly"
       },
       "data": [
        {
         "marker": {
          "color": "red",
          "size": 10
         },
         "mode": "markers",
         "name": "Tumor",
         "type": "scatter",
         "x": [
          "BLCA-US",
          "BOCA-UK",
          "BRCA-EU",
          "BRCA-UK",
          "BRCA-US",
          "CLLE-ES",
          "COAD-US",
          "EOPC-DE",
          "ESAD-UK",
          "GACA-CN",
          "GBM-US",
          "HNSC-US",
          "KICH-US",
          "KIRC-US",
          "KIRP-US",
          "LAML-KR",
          "LAML-US",
          "LIHC-US",
          "LINC-JP",
          "LIRI-JP",
          "LUAD-US",
          "LUSC-US",
          "MELA-AU",
          "ORCA-IN",
          "OV-US",
          "PACA-AU",
          "PACA-CA",
          "PAEN-AU",
          "PBCA-DE",
          "PRAD-CA",
          "PRAD-US",
          "READ-US",
          "RECA-EU",
          "SARC-US",
          "UCEC-US"
         ],
         "y": [
          4.3478260869565215,
          2.631578947368421,
          5.063291139240507,
          4.444444444444445,
          0,
          14.141414141414142,
          9.75609756097561,
          11.11111111111111,
          3,
          16.666666666666668,
          14.634146341463415,
          0,
          8.16326530612245,
          15,
          2.9411764705882355,
          10,
          3.0303030303030303,
          7.407407407407407,
          9.67741935483871,
          34.136546184738954,
          0,
          4.166666666666667,
          5.714285714285714,
          7.6923076923076925,
          0,
          2.0202020202020203,
          4.109589041095891,
          4.761904761904762,
          9.163346613545816,
          4.032258064516129,
          0,
          0,
          1.0526315789473684,
          5.882352941176471,
          4.081632653061225
         ]
        },
        {
         "marker": {
          "color": "green",
          "size": 10,
          "symbol": "circle-x-open"
         },
         "mode": "markers",
         "name": "Control",
         "type": "scatter",
         "x": [
          "BLCA-US",
          "BOCA-UK",
          "BRCA-EU",
          "BRCA-UK",
          "BRCA-US",
          "CLLE-ES",
          "COAD-US",
          "EOPC-DE",
          "ESAD-UK",
          "GACA-CN",
          "GBM-US",
          "HNSC-US",
          "KICH-US",
          "KIRC-US",
          "KIRP-US",
          "LAML-KR",
          "LAML-US",
          "LIHC-US",
          "LINC-JP",
          "LIRI-JP",
          "LUAD-US",
          "LUSC-US",
          "MELA-AU",
          "ORCA-IN",
          "OV-US",
          "PACA-AU",
          "PACA-CA",
          "PAEN-AU",
          "PBCA-DE",
          "PRAD-CA",
          "PRAD-US",
          "READ-US",
          "RECA-EU",
          "SARC-US",
          "UCEC-US"
         ],
         "y": [
          0,
          5.2631578947368425,
          2.5316455696202533,
          13.043478260869565,
          4.3478260869565215,
          15,
          11.904761904761903,
          2.7777777777777777,
          5,
          2.380952380952381,
          7.317073170731708,
          4.545454545454546,
          14.285714285714286,
          25,
          11.764705882352942,
          0,
          12.12121212121212,
          20.37037037037037,
          12.903225806451612,
          36.54618473895582,
          11.904761904761903,
          8.333333333333334,
          12.857142857142858,
          0,
          4.444444444444445,
          7.070707070707071,
          6.164383561643835,
          5.714285714285714,
          9.6,
          4.032258064516129,
          5,
          30.76923076923077,
          3.1578947368421053,
          2.9411764705882355,
          4
         ]
        }
       ],
       "layout": {
        "paper_bgcolor": "white",
        "plot_bgcolor": "white",
        "template": {
         "data": {
          "bar": [
           {
            "error_x": {
             "color": "#2a3f5f"
            },
            "error_y": {
             "color": "#2a3f5f"
            },
            "marker": {
             "line": {
              "color": "#E5ECF6",
              "width": 0.5
             },
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "bar"
           }
          ],
          "barpolar": [
           {
            "marker": {
             "line": {
              "color": "#E5ECF6",
              "width": 0.5
             },
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "barpolar"
           }
          ],
          "carpet": [
           {
            "aaxis": {
             "endlinecolor": "#2a3f5f",
             "gridcolor": "white",
             "linecolor": "white",
             "minorgridcolor": "white",
             "startlinecolor": "#2a3f5f"
            },
            "baxis": {
             "endlinecolor": "#2a3f5f",
             "gridcolor": "white",
             "linecolor": "white",
             "minorgridcolor": "white",
             "startlinecolor": "#2a3f5f"
            },
            "type": "carpet"
           }
          ],
          "choropleth": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "choropleth"
           }
          ],
          "contour": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "contour"
           }
          ],
          "contourcarpet": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "contourcarpet"
           }
          ],
          "heatmap": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "heatmap"
           }
          ],
          "heatmapgl": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "heatmapgl"
           }
          ],
          "histogram": [
           {
            "marker": {
             "pattern": {
              "fillmode": "overlay",
              "size": 10,
              "solidity": 0.2
             }
            },
            "type": "histogram"
           }
          ],
          "histogram2d": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "histogram2d"
           }
          ],
          "histogram2dcontour": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "histogram2dcontour"
           }
          ],
          "mesh3d": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "type": "mesh3d"
           }
          ],
          "parcoords": [
           {
            "line": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "parcoords"
           }
          ],
          "pie": [
           {
            "automargin": true,
            "type": "pie"
           }
          ],
          "scatter": [
           {
            "fillpattern": {
             "fillmode": "overlay",
             "size": 10,
             "solidity": 0.2
            },
            "type": "scatter"
           }
          ],
          "scatter3d": [
           {
            "line": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatter3d"
           }
          ],
          "scattercarpet": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattercarpet"
           }
          ],
          "scattergeo": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattergeo"
           }
          ],
          "scattergl": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattergl"
           }
          ],
          "scattermapbox": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scattermapbox"
           }
          ],
          "scatterpolar": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterpolar"
           }
          ],
          "scatterpolargl": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterpolargl"
           }
          ],
          "scatterternary": [
           {
            "marker": {
             "colorbar": {
              "outlinewidth": 0,
              "ticks": ""
             }
            },
            "type": "scatterternary"
           }
          ],
          "surface": [
           {
            "colorbar": {
             "outlinewidth": 0,
             "ticks": ""
            },
            "colorscale": [
             [
              0,
              "#0d0887"
             ],
             [
              0.1111111111111111,
              "#46039f"
             ],
             [
              0.2222222222222222,
              "#7201a8"
             ],
             [
              0.3333333333333333,
              "#9c179e"
             ],
             [
              0.4444444444444444,
              "#bd3786"
             ],
             [
              0.5555555555555556,
              "#d8576b"
             ],
             [
              0.6666666666666666,
              "#ed7953"
             ],
             [
              0.7777777777777778,
              "#fb9f3a"
             ],
             [
              0.8888888888888888,
              "#fdca26"
             ],
             [
              1,
              "#f0f921"
             ]
            ],
            "type": "surface"
           }
          ],
          "table": [
           {
            "cells": {
             "fill": {
              "color": "#EBF0F8"
             },
             "line": {
              "color": "white"
             }
            },
            "header": {
             "fill": {
              "color": "#C8D4E3"
             },
             "line": {
              "color": "white"
             }
            },
            "type": "table"
           }
          ]
         },
         "layout": {
          "annotationdefaults": {
           "arrowcolor": "#2a3f5f",
           "arrowhead": 0,
           "arrowwidth": 1
          },
          "autotypenumbers": "strict",
          "coloraxis": {
           "colorbar": {
            "outlinewidth": 0,
            "ticks": ""
           }
          },
          "colorscale": {
           "diverging": [
            [
             0,
             "#8e0152"
            ],
            [
             0.1,
             "#c51b7d"
            ],
            [
             0.2,
             "#de77ae"
            ],
            [
             0.3,
             "#f1b6da"
            ],
            [
             0.4,
             "#fde0ef"
            ],
            [
             0.5,
             "#f7f7f7"
            ],
            [
             0.6,
             "#e6f5d0"
            ],
            [
             0.7,
             "#b8e186"
            ],
            [
             0.8,
             "#7fbc41"
            ],
            [
             0.9,
             "#4d9221"
            ],
            [
             1,
             "#276419"
            ]
           ],
           "sequential": [
            [
             0,
             "#0d0887"
            ],
            [
             0.1111111111111111,
             "#46039f"
            ],
            [
             0.2222222222222222,
             "#7201a8"
            ],
            [
             0.3333333333333333,
             "#9c179e"
            ],
            [
             0.4444444444444444,
             "#bd3786"
            ],
            [
             0.5555555555555556,
             "#d8576b"
            ],
            [
             0.6666666666666666,
             "#ed7953"
            ],
            [
             0.7777777777777778,
             "#fb9f3a"
            ],
            [
             0.8888888888888888,
             "#fdca26"
            ],
            [
             1,
             "#f0f921"
            ]
           ],
           "sequentialminus": [
            [
             0,
             "#0d0887"
            ],
            [
             0.1111111111111111,
             "#46039f"
            ],
            [
             0.2222222222222222,
             "#7201a8"
            ],
            [
             0.3333333333333333,
             "#9c179e"
            ],
            [
             0.4444444444444444,
             "#bd3786"
            ],
            [
             0.5555555555555556,
             "#d8576b"
            ],
            [
             0.6666666666666666,
             "#ed7953"
            ],
            [
             0.7777777777777778,
             "#fb9f3a"
            ],
            [
             0.8888888888888888,
             "#fdca26"
            ],
            [
             1,
             "#f0f921"
            ]
           ]
          },
          "colorway": [
           "#636efa",
           "#EF553B",
           "#00cc96",
           "#ab63fa",
           "#FFA15A",
           "#19d3f3",
           "#FF6692",
           "#B6E880",
           "#FF97FF",
           "#FECB52"
          ],
          "font": {
           "color": "#2a3f5f"
          },
          "geo": {
           "bgcolor": "white",
           "lakecolor": "white",
           "landcolor": "#E5ECF6",
           "showlakes": true,
           "showland": true,
           "subunitcolor": "white"
          },
          "hoverlabel": {
           "align": "left"
          },
          "hovermode": "closest",
          "mapbox": {
           "style": "light"
          },
          "paper_bgcolor": "white",
          "plot_bgcolor": "#E5ECF6",
          "polar": {
           "angularaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "bgcolor": "#E5ECF6",
           "radialaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           }
          },
          "scene": {
           "xaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           },
           "yaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           },
           "zaxis": {
            "backgroundcolor": "#E5ECF6",
            "gridcolor": "white",
            "gridwidth": 2,
            "linecolor": "white",
            "showbackground": true,
            "ticks": "",
            "zerolinecolor": "white"
           }
          },
          "shapedefaults": {
           "line": {
            "color": "#2a3f5f"
           }
          },
          "ternary": {
           "aaxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "baxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           },
           "bgcolor": "#E5ECF6",
           "caxis": {
            "gridcolor": "white",
            "linecolor": "white",
            "ticks": ""
           }
          },
          "title": {
           "x": 0.05
          },
          "xaxis": {
           "automargin": true,
           "gridcolor": "white",
           "linecolor": "white",
           "ticks": "",
           "title": {
            "standoff": 15
           },
           "zerolinecolor": "white",
           "zerolinewidth": 2
          },
          "yaxis": {
           "automargin": true,
           "gridcolor": "white",
           "linecolor": "white",
           "ticks": "",
           "title": {
            "standoff": 15
           },
           "zerolinecolor": "white",
           "zerolinewidth": 2
          }
         }
        },
        "title": {
         "text": "<b>Patients with a minor allele frequency > 25%</b><br><sup>Gene Name: deep_pileup / Position: pcawg / Displaying 35 of 44 cohorts"
        },
        "xaxis": {
         "linecolor": "black",
         "linewidth": 1,
         "showline": true,
         "tickangle": 45,
         "title": {
          "text": "Cohorts"
         }
        },
        "yaxis": {
         "linecolor": "black",
         "linewidth": 1,
         "showline": true,
         "tickcolor": "black",
         "ticks": "outside",
         "tickwidth": 2,
         "title": {
          "text": "Percent of Patients"
         }
        }
       }
      }
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "af_greater_than_25_scatterplot(\"/Users/nicholasabad/Desktop/workspace/data/deep_pileup/pcawg/DLX2/chr2:172967805/Overview.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "ename": "EmptyDataError",
     "evalue": "No columns to parse from file",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mEmptyDataError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[40], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mat_least_two_variant_alleles_scatterplot\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43m/Users/nicholasabad/Desktop/workspace/data/deep_pileup/pcawg/DLX2/chr2:172967805/Overview.tsv\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[0;32mIn[38], line 102\u001b[0m, in \u001b[0;36mat_least_two_variant_alleles_scatterplot\u001b[0;34m(path_to_overview_file, only_relevant, ref, alt)\u001b[0m\n\u001b[1;32m     96\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mat_least_two_variant_alleles_scatterplot\u001b[39m(\n\u001b[1;32m     97\u001b[0m     path_to_overview_file: \u001b[38;5;28mstr\u001b[39m, \n\u001b[1;32m     98\u001b[0m     only_relevant: \u001b[38;5;28mbool\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m     99\u001b[0m     ref: \u001b[38;5;28mstr\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m    100\u001b[0m     alt: \u001b[38;5;28mstr\u001b[39m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m    101\u001b[0m ):\n\u001b[0;32m--> 102\u001b[0m     data \u001b[38;5;241m=\u001b[39m \u001b[43mpd\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mread_csv\u001b[49m\u001b[43m(\u001b[49m\u001b[43mpath_to_overview_file\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mdelimiter\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;130;43;01m\\t\u001b[39;49;00m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m)\u001b[49m\n\u001b[1;32m    104\u001b[0m     gene_name \u001b[38;5;241m=\u001b[39m path_to_overview_file\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/\u001b[39m\u001b[38;5;124m\"\u001b[39m)[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m5\u001b[39m]\n\u001b[1;32m    105\u001b[0m     position \u001b[38;5;241m=\u001b[39m path_to_overview_file\u001b[38;5;241m.\u001b[39msplit(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m/\u001b[39m\u001b[38;5;124m\"\u001b[39m)[\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m4\u001b[39m]\n",
      "File \u001b[0;32m~/miniconda3/envs/psnv-hunter/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1026\u001b[0m, in \u001b[0;36mread_csv\u001b[0;34m(filepath_or_buffer, sep, delimiter, header, names, index_col, usecols, dtype, engine, converters, true_values, false_values, skipinitialspace, skiprows, skipfooter, nrows, na_values, keep_default_na, na_filter, verbose, skip_blank_lines, parse_dates, infer_datetime_format, keep_date_col, date_parser, date_format, dayfirst, cache_dates, iterator, chunksize, compression, thousands, decimal, lineterminator, quotechar, quoting, doublequote, escapechar, comment, encoding, encoding_errors, dialect, on_bad_lines, delim_whitespace, low_memory, memory_map, float_precision, storage_options, dtype_backend)\u001b[0m\n\u001b[1;32m   1013\u001b[0m kwds_defaults \u001b[38;5;241m=\u001b[39m _refine_defaults_read(\n\u001b[1;32m   1014\u001b[0m     dialect,\n\u001b[1;32m   1015\u001b[0m     delimiter,\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1022\u001b[0m     dtype_backend\u001b[38;5;241m=\u001b[39mdtype_backend,\n\u001b[1;32m   1023\u001b[0m )\n\u001b[1;32m   1024\u001b[0m kwds\u001b[38;5;241m.\u001b[39mupdate(kwds_defaults)\n\u001b[0;32m-> 1026\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43m_read\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/psnv-hunter/lib/python3.11/site-packages/pandas/io/parsers/readers.py:620\u001b[0m, in \u001b[0;36m_read\u001b[0;34m(filepath_or_buffer, kwds)\u001b[0m\n\u001b[1;32m    617\u001b[0m _validate_names(kwds\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnames\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m))\n\u001b[1;32m    619\u001b[0m \u001b[38;5;66;03m# Create the parser.\u001b[39;00m\n\u001b[0;32m--> 620\u001b[0m parser \u001b[38;5;241m=\u001b[39m \u001b[43mTextFileReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfilepath_or_buffer\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    622\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m chunksize \u001b[38;5;129;01mor\u001b[39;00m iterator:\n\u001b[1;32m    623\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m parser\n",
      "File \u001b[0;32m~/miniconda3/envs/psnv-hunter/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1620\u001b[0m, in \u001b[0;36mTextFileReader.__init__\u001b[0;34m(self, f, engine, **kwds)\u001b[0m\n\u001b[1;32m   1617\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39moptions[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhas_index_names\u001b[39m\u001b[38;5;124m\"\u001b[39m]\n\u001b[1;32m   1619\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles: IOHandles \u001b[38;5;241m|\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m-> 1620\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_engine\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mengine\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~/miniconda3/envs/psnv-hunter/lib/python3.11/site-packages/pandas/io/parsers/readers.py:1898\u001b[0m, in \u001b[0;36mTextFileReader._make_engine\u001b[0;34m(self, f, engine)\u001b[0m\n\u001b[1;32m   1895\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(msg)\n\u001b[1;32m   1897\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1898\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mmapping\u001b[49m\u001b[43m[\u001b[49m\u001b[43mengine\u001b[49m\u001b[43m]\u001b[49m\u001b[43m(\u001b[49m\u001b[43mf\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43moptions\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1899\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mException\u001b[39;00m:\n\u001b[1;32m   1900\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhandles \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n",
      "File \u001b[0;32m~/miniconda3/envs/psnv-hunter/lib/python3.11/site-packages/pandas/io/parsers/c_parser_wrapper.py:93\u001b[0m, in \u001b[0;36mCParserWrapper.__init__\u001b[0;34m(self, src, **kwds)\u001b[0m\n\u001b[1;32m     90\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m kwds[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdtype_backend\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpyarrow\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n\u001b[1;32m     91\u001b[0m     \u001b[38;5;66;03m# Fail here loudly instead of in cython after reading\u001b[39;00m\n\u001b[1;32m     92\u001b[0m     import_optional_dependency(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpyarrow\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m---> 93\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reader \u001b[38;5;241m=\u001b[39m \u001b[43mparsers\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mTextReader\u001b[49m\u001b[43m(\u001b[49m\u001b[43msrc\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwds\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     95\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39munnamed_cols \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_reader\u001b[38;5;241m.\u001b[39munnamed_cols\n\u001b[1;32m     97\u001b[0m \u001b[38;5;66;03m# error: Cannot determine type of 'names'\u001b[39;00m\n",
      "File \u001b[0;32mparsers.pyx:581\u001b[0m, in \u001b[0;36mpandas._libs.parsers.TextReader.__cinit__\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mEmptyDataError\u001b[0m: No columns to parse from file"
     ]
    }
   ],
   "source": [
    "at_least_two_variant_alleles_scatterplot(\"/Users/nicholasabad/Desktop/workspace/data/deep_pileup/pcawg/DLX2/chr2:172967805/Overview.tsv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "psnv-hunter",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
